// @tss/sdk/esm2015/lib/esri/base-esri.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LrsWhereClauseBuilder } from '../lrs/lrs-where-clause-builder';
/**
 * @abstract
 */
export class BaseEsriService {
    /**
     * @param {?} http
     * @param {?} config
     * @param {?} dataConfig
     * @param {?} log
     */
    constructor(http, config, dataConfig, log) {
        this.http = http;
        this.config = config;
        this.dataConfig = dataConfig;
        this.log = log;
    }
    /**
     * @protected
     * @param {?} payload
     * @return {?}
     */
    shouldUsePostMethod(payload) {
        return this.config.arcgisQueryMethodStrategy === 'force-post'
            || JSON.stringify(payload || '').length > this.config.arcgisMaxWhereLengthGetThreshold;
    }
    /**
     * @protected
     * @template T
     * @param {?} response
     * @param {?} url
     * @return {?}
     */
    handleEsriResponseError(response, url) {
        /** @type {?} */
        const error = (/** @type {?} */ (response));
        if (error && error.error) {
            /** @type {?} */
            const returnError = new Error(`Failed to query map service: ${error.error.message}`);
            this.log.error(`Failed to execute map service request: ${url}`, returnError);
            throw returnError;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    getWhereBuilder() {
        return new LrsWhereClauseBuilder({ quoteFields: !this.dataConfig.skipColumnQuotes });
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.http;
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.dataConfig;
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.log;
}
