// @tss/sdk/esm2015/lib/esri/esri-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { AutoConfig } from '../core/auto-config.decorator';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing global ESRI-related settings.
 */
let EsriConfig = /**
 * Configuration class for storing global ESRI-related settings.
 */
class EsriConfig {
    constructor() {
        this.arcgis = {};
        this.arcgisQueryMethodStrategy = 'prefer-get';
        this.arcgisMaxWhereLengthGetThreshold = 3500;
    }
    /**
     * @return {?}
     */
    get hasGeometryUrl() {
        return Boolean(this.arcgis && this.arcgis.geometryUrl);
    }
    /**
     * @return {?}
     */
    get arcGisGeometryUrl() {
        return this.hasGeometryUrl ? this.arcgis.geometryUrl : null;
    }
};
EsriConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ EsriConfig.ngInjectableDef = i0.defineInjectable({ factory: function EsriConfig_Factory() { return new EsriConfig(); }, token: EsriConfig, providedIn: "root" });
/**
 * Configuration class for storing global ESRI-related settings.
 */
EsriConfig = tslib_1.__decorate([
    AutoConfig('EsriConfig')
], EsriConfig);
export { EsriConfig };
if (false) {
    /** @type {?} */
    EsriConfig.prototype.arcgis;
    /** @type {?} */
    EsriConfig.prototype.arcgisQueryMethodStrategy;
    /** @type {?} */
    EsriConfig.prototype.arcgisMaxWhereLengthGetThreshold;
    /** @type {?} */
    EsriConfig.prototype.arcgisPortalUrl;
    /** @type {?} */
    EsriConfig.prototype.arcgisPortalClientId;
}
