// @tss/sdk/esm2015/lib/esri/esri.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Server identifiers.
/** @type {?} */
export const MAP_SERVER_SERVICE_MATCH_IDENTIFIER = /(\/MapServer$|\/MapServer\/)/;
/** @type {?} */
export const MAP_SERVER_SERVICE_IDENTIFIER = /(MapServer|MapServer\/)$/;
/** @type {?} */
export const FEATURE_SERVER_SERVICE_IDENTIFIER = /(FeatureServer|FeatureServer\/)$/;
/** @type {?} */
export const IMAGE_SERVER_SERVICE_IDENTIFIER = /(ImageServer|ImageServer\/)$'/;
/** @type {?} */
export const FEATURE_SERVER_LAYER_SERVICE_IDENTIFIER = /FeatureServer\/(\d+|\d+\/)$/;
/** @type {?} */
export const MAP_SERVER_LAYER_SERVICE_MATCH_IDENTIFIER = /(\/MapServer\/(\d+|\d+\/)|\/MapServer\/(\d+|\d+\/)\/)/;
/** @type {?} */
export const MAP_SERVER_LAYER_SERVICE_IDENTIFIER = /MapServer\/(\d+|\d+\/)$/;
/** @type {?} */
export const LRS_SERVER_SERVICE_IDENTIFIER = /(MapServer\/exts\/LRSServer|MapServer\/exts\/LRSServer\/)$/;
/** @type {?} */
export const LRS_SERVER_LAYERS_SERVICE_IDENTIFIER = /(MapServer\/exts\/LRSServer\/layers|MapServer\/exts\/LRSServer\/layers\/)$/;
// Map endpoints.
/** @type {?} */
export const MAP_LAYERS_SERVICE_ENDPOINT = 'layers';
/** @type {?} */
export const MAP_LAYER_SERVICE_ENDPOINT = '{layerId}';
/** @type {?} */
export const MAP_LAYER_QUERY_SERVICE_ENDPOINT = '{layerId}/query';
/** @type {?} */
export const MAP_LAYER_QUERY_RELATED_RECORDS_SERVICE_ENDPOINT = 'layers/{layerId}/queryRelatedRecords';
// LRS endpoints.
/** @type {?} */
export const LRS_SERVER_META_ENDPOINT = 'exts/LRSServer';
/** @type {?} */
export const LRS_LAYERS_SERVICE_ENDPOINT = 'exts/LRSServer/layers';
/** @type {?} */
export const LRS_SERVER_APPLY_EDITS_ENDPOINT = 'exts/LRSServer/applyEdits';
// Network endpoints.
/** @type {?} */
export const LRS_GEOMETRY_TO_MEASURE = 'exts/LRSServer/networkLayers/{layerId}/geometryToMeasure';
// Portal.
/** @type {?} */
export const PORTAL_SHARING_ENDPOINT = 'sharing';
/** @type {?} */
export const PORTAL_OAUTH_AUTHORIZE_ENDPOINT = 'sharing/oauth2/authorize';
/** @type {?} */
export const PORTAL_OAUTH_TOKEN_ENDPOINT = 'sharing/rest/oauth2/token';
/** @type {?} */
export const PORTAL_GENERATE_TOKEN_ENDPOINT = 'sharing/generateToken';
/** @type {?} */
export const PORTAL_REDIRECT_URL_PARAMS_BREAK_CHARACTER = '#';
/** @type {?} */
export const DEFAULT_PORTAL_OAUTH_REDIRECT_OPTIONS = {
    responseType: 'token',
    urlParamsBreak: PORTAL_REDIRECT_URL_PARAMS_BREAK_CHARACTER,
    redirectUrl: window.location.href
};
