// @tss/sdk/esm5/lib/maps/map.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TssCoreModule } from '../core/core.module';
import { LayerRendererTypePipe } from './drawing/layer-renderer-type.pipe';
var TssMapsModule = /** @class */ (function () {
    function TssMapsModule() {
    }
    TssMapsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        TssCoreModule,
                    ],
                    declarations: [
                        LayerRendererTypePipe
                    ],
                    exports: [
                        LayerRendererTypePipe
                    ],
                    providers: [],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                },] }
    ];
    return TssMapsModule;
}());
export { TssMapsModule };
