// @tss/sdk/esm2015/lib/security/users/user.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { SecurityConfig } from '../security-config';
import { join } from '../../core/url/join.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../security-config";
/**
 * \@example
 * ## Get Users that belong to a given role (by ID).
 * service.usersByRoleId('1428a342-8a95-4b6c-b3e9-da135e3ced04')
 *
 * // ...which is just a shortcut for:
 * service.query({}, new HttpParams().set('filterByRoleId', roleName));
 *
 * ## Get Users that belong to a given role (by name).
 * service.usersByRoleName('Admin')
 *
 * // ...which is just a shortcut for:
 * service.query({}, new HttpParams().set('filterByRoleName', roleName));
 */
export class UserService extends BaseResourceService {
    /**
     * @param {?} http
     * @param {?} config
     */
    constructor(http, config) {
        super(http, {
            url: join(config.serviceUrl, config.usersEndpoint)
        });
        this.http = http;
    }
    /**
     * @param {?} roleId
     * @param {?=} params
     * @return {?}
     */
    usersByRoleId(roleId, params) {
        params = (params || new HttpParams()).set('filterByRoleId', roleId);
        return this.query({}, params);
    }
    /**
     * @param {?} roleName
     * @param {?=} params
     * @return {?}
     */
    usersByRoleName(roleName, params) {
        params = (params || new HttpParams()).set('filterByRoleName', roleName);
        return this.query({}, params);
    }
    /**
     * @param {?} searchText
     * @param {?=} params
     * @return {?}
     */
    search(searchText, params) {
        params = (params || new HttpParams()).set('search', searchText);
        return this.query({}, params);
    }
}
UserService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
UserService.ctorParameters = () => [
    { type: HttpClient },
    { type: SecurityConfig }
];
/** @nocollapse */ UserService.ngInjectableDef = i0.defineInjectable({ factory: function UserService_Factory() { return new UserService(i0.inject(i1.HttpClient), i0.inject(i2.SecurityConfig)); }, token: UserService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserService.prototype.http;
}
