// @tss/sdk/esm2015/lib/core/media/media.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const BASE_64_METADATA_REGEX = /data:([a-zA-Z0-9]+\/[a-zA-Z0-9-.+]+).*,.*/;
/** @type {?} */
export const JSON_MIME_TYPE = 'application/json';
// Text files.
/** @type {?} */
export const TEXT_MIME_TYPE = 'text/plain';
/** @type {?} */
export const TEXT_FILE_EXTENSION = '.txt';
// Audio.
/** @type {?} */
export const AUDIO_WEBM_MIME_TYPE = 'audio/webm';
/** @type {?} */
export const AUDIO_MP3_MIME_TYPE = 'audio/mpeg-3';
/** @type {?} */
export const AUDIO_MP3_FILE_EXTENSION = '.mp3';
// Video.
/** @type {?} */
export const VIDEO_WEBM_MIME_TYPE = 'video/webm;codecs=h264';
/** @type {?} */
export const VIDEO_MP4_MIME_TYPE = 'video/mp4';
/** @type {?} */
export const VIDEO_MP4_FILE_EXTENSION = '.mp4';
// Images.
/** @type {?} */
export const IMAGE_PNG_FILE_EXTENSION = '.png';
/** @type {?} */
export const IMAGE_PNG_MIME_TYPE = 'image/png';
/** @type {?} */
export const IMAGE_JPEG_FILE_EXTENSION = '.jpeg';
/** @type {?} */
export const IMAGE_JPEG_MIME_TYPE = 'image/jpeg';
