// @tss/sdk/esm2015/lib/core/validation/validation-check.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ValidationResult } from './validation-results';
/** @type {?} */
let BASE_ID = 1;
/**
 * @template T
 */
export class ValidationCheck {
    /**
     * @param {?} description
     * @param {?=} passCheck
     */
    constructor(description, passCheck) {
        this.subChecks = [];
        this.description = description;
        this.passCheck = passCheck;
        this.id = BASE_ID++;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    hasId(id) {
        this.id = id;
        return this;
    }
    /**
     * @param {?} description
     * @return {?}
     */
    message(description) {
        this.description = description;
        return this;
    }
    /**
     * @param {?=} passCheck
     * @return {?}
     */
    pass(passCheck) {
        this.passCheck = passCheck;
        return this;
    }
    /**
     * @param {?=} failCheck
     * @return {?}
     */
    fail(failCheck) {
        this.failCheck = failCheck;
        return this;
    }
    /**
     * @param {?=} warnCheck
     * @return {?}
     */
    warn(warnCheck) {
        this.warnCheck = warnCheck;
        return this;
    }
    /**
     * @param {?=} onlyIfCheck
     * @return {?}
     */
    onlyIf(onlyIfCheck) {
        this.onlyIfCheck = onlyIfCheck;
        return this;
    }
    /**
     * @param {...?} innerChecks
     * @return {?}
     */
    add(...innerChecks) {
        innerChecks.forEach((/**
         * @param {?} inner
         * @return {?}
         */
        inner => this.subChecks.push(inner)));
        return this;
    }
    /**
     * @param {?} target
     * @return {?}
     */
    check(target) {
        /** @type {?} */
        const result = new ValidationResult();
        if (this.onlyIfCheck && !this.onlyIfCheck(target)) {
            result.skipped = true;
            return result;
        }
        if (this.passCheck) {
            result.passed = this.passCheck(target);
        }
        if (this.failCheck) {
            result.passed = !this.failCheck(target);
        }
        if (this.warnCheck) {
            result.warned = !this.warnCheck(target);
        }
        if (this.subChecks) {
            this.subChecks.forEach((/**
             * @param {?} c
             * @return {?}
             */
            c => {
                /** @type {?} */
                const innerResult = c.check(target);
                result.add(innerResult);
            }));
        }
        return result;
    }
}
if (false) {
    /** @type {?} */
    ValidationCheck.prototype.id;
    /** @type {?} */
    ValidationCheck.prototype.description;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.subChecks;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.passCheck;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.failCheck;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.warnCheck;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.onlyIfCheck;
}
