// @tss/sdk/esm2015/lib/core/validation/validation-results.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ValidationResult {
    constructor() {
        this.violaters = [];
    }
    /**
     * @return {?}
     */
    get failed() {
        return this.ran && !this.passed;
    }
    /**
     * @return {?}
     */
    get skipped() {
        return this.validationSkipped;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set skipped(value) {
        this.validationSkipped = value;
        this.ran = true;
    }
    /**
     * @param {?} inner
     * @return {?}
     */
    add(inner) {
        if (inner.failed) {
            this.passed = false;
        }
        if (inner.warned) {
            this.warned = true;
        }
        return this;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidationResult.prototype.validationSkipped;
    /** @type {?} */
    ValidationResult.prototype.passed;
    /** @type {?} */
    ValidationResult.prototype.warned;
    /** @type {?} */
    ValidationResult.prototype.ran;
    /** @type {?} */
    ValidationResult.prototype.violaters;
}
