// @tss/sdk/esm5/lib/lrs/redlines/redline-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// TSS.
import { LrsFeature } from '../lrs-feature';
import { toFeature } from '../feature';
import { isPolygon } from '../../geometry/geometry/is-polygon.function';
import { clone } from '../../core/clone/clone.function';
var RedlineFeature = /** @class */ (function (_super) {
    tslib_1.__extends(RedlineFeature, _super);
    function RedlineFeature(feature, layer) {
        var _this = _super.call(this, feature, layer) || this;
        _this.routeId = _this.getValue(layer.routeIdFieldName);
        _this.fromMeasure = _this.getValue(layer.fromMeasureFieldName);
        _this.toMeasure = _this.getValue(layer.toMeasureFieldName);
        _this.isLinear = Boolean(layer.toMeasureFieldName);
        return _this;
    }
    /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    RedlineFeature.create = /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    function (layer, properties, geometry) {
        if (properties === void 0) { properties = {}; }
        if (geometry === void 0) { geometry = null; }
        /** @type {?} */
        var feature = toFeature(properties, (/** @type {?} */ (geometry)));
        return new RedlineFeature(feature, layer);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    RedlineFeature.prototype.setFromMeasure = /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    function (measure) {
        (/** @type {?} */ (this)).fromMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.fromMeasureFieldName, measure);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    RedlineFeature.prototype.setToMeasure = /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    function (measure) {
        if (!(/** @type {?} */ (this)).isLinear) {
            throw new Error('Cannot set to measure for a point-based feature');
        }
        (/** @type {?} */ (this)).toMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toMeasureFieldName, measure);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @return {THIS}
     */
    RedlineFeature.prototype.setMeasureExtent = /**
     * @template THIS
     * @this {THIS}
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @return {THIS}
     */
    function (fromMeasure, toMeasure) {
        return (/** @type {?} */ (this)).setFromMeasure(fromMeasure)
            .setToMeasure(toMeasure);
    };
    /**
     * @return {?}
     */
    RedlineFeature.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var copyFeature = clone(this.toFeature());
        return new RedlineFeature(copyFeature, this.layer);
    };
    /**
     * @protected
     * @return {?}
     */
    RedlineFeature.prototype.isGeometryTypeValid = /**
     * @protected
     * @return {?}
     */
    function () {
        return isPolygon(this.geometry) || _super.prototype.isGeometryTypeValid.call(this);
    };
    return RedlineFeature;
}(LrsFeature));
export { RedlineFeature };
if (false) {
    /** @type {?} */
    RedlineFeature.prototype.layer;
}
