// @tss/sdk/esm2015/lib/geolocation/position-path-builder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import destination from '@turf/destination';
import bearing from '@turf/bearing';
import { isPositionAcceptable } from './is-position-acceptable.function';
import { setMeasuresByGeometricLength } from '../geometry/geometry/set-measures-by-geometric-length.function';
import { isMultiPartLine } from '../geometry/line/is-multi-part.function';
import { getMeasureFromPoint } from '../geometry/point/get-measure-from-point.function';
import { PATH_BUILDER_EMPTY_CALIBRATION_MEASURE_VALUE, PATH_BUILDER_CALCULATED_M_DEVICE_SOURCE } from './geolocation.constants';
import { getDistanceBetweenLatLng } from '../geometry/point/get-distance-between-points.function';
import { convertLengthUnits } from '../core/units/convert-length-units.function';
import { LengthUnit } from '../core/units/length-unit';
import { sumValues } from '../core/stats/sum-values.function';
import { getLastItem } from '../core/array/get-last-item.function';
import { isNumber } from '../core/type-check/is-number.function';
import { pointsAlmostEqual } from '../core/equality/points-almost-equal.function';
/** @type {?} */
const UNPLACED_CALIBRATION_MEASURE_KEY = 'pending';
/** @type {?} */
const DEFAULT_OPTIONS = {
    gpsRequiredAccuracyMeters: 10,
    requireAccuracy: true,
    pointEquivalencyPrecision: 8,
    returnZValues: true,
    returnMValues: true,
    mValueUnit: 'miles',
    storeUnfilteredPath: false,
    calibrationMeasureSnappingDistanceDelta: 2,
    calibrationMeasureSnappingDistanceDeltaUnit: LengthUnit.Foot
};
export class PositionPathBuilder {
    /**
     * @return {?}
     */
    get positionCount() {
        /** @type {?} */
        const counts = this.parts.map((/**
         * @param {?} part
         * @return {?}
         */
        part => part.length));
        return sumValues(counts);
    }
    /**
     * @return {?}
     */
    get currentPartIndex() {
        return this.parts.length - 1;
    }
    /**
     * @return {?}
     */
    get currentPart() {
        return this.parts[this.currentPartIndex];
    }
    /**
     * @return {?}
     */
    get lastPosition() {
        return getLastItem(this.currentPart);
    }
    /**
     * @return {?}
     */
    get hasPosition() {
        return Boolean(this.lastPosition);
    }
    /**
     * @return {?}
     */
    get isEmpty() {
        return this.currentPart.length === 0;
    }
    /**
     * @return {?}
     */
    get isValidPath() {
        return this.currentPart.length > 1;
    }
    /**
     * @return {?}
     */
    get isSinglePoint() {
        return this.parts.length === 1 && this.currentPart.length === 1;
    }
    /**
     * @param {?=} options
     */
    constructor(options) {
        this.parts = [[]];
        this.unfilteredPath = [];
        this.calibratedMeasures = new Map();
        this.options = Object.assign({}, DEFAULT_OPTIONS, options);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} positionStream
     * @return {THIS}
     */
    trackPosition(positionStream) {
        (/** @type {?} */ (this)).trackSubscription = positionStream.subscribe((/**
         * @template THIS
         * @this {THIS}
         * @param {?} position
         * @return {THIS}
         */
        position => (/** @type {?} */ (this)).addPosition(position)));
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} position
     * @param {?=} calibrationMeasure
     * @return {THIS}
     */
    addPosition(position, calibrationMeasure) {
        /** @type {?} */
        const accuracyCriteriaMet = !(/** @type {?} */ (this)).options.requireAccuracy
            || isPositionAcceptable(position, (/** @type {?} */ (this)).options.gpsRequiredAccuracyMeters);
        /** @type {?} */
        const positionNotAcceptable = !position || !accuracyCriteriaMet;
        if (positionNotAcceptable) {
            return (/** @type {?} */ (this)).tryStoreUnfilteredPath(position);
        }
        if ((/** @type {?} */ (this)).isEmpty) {
            return (/** @type {?} */ (this)).forceAddPosition(position);
        }
        /** @type {?} */
        const pointsAtSameLocation = (/** @type {?} */ (this)).arePositionsEquivalent((/** @type {?} */ (this)).lastPosition, position);
        /** @type {?} */
        const bothPointsAtStandStill = (/** @type {?} */ (this)).lastPosition.speed === 0 && position.speed === 0 || pointsAtSameLocation;
        if (bothPointsAtStandStill) {
            (/** @type {?} */ (this)).tryStoreUnfilteredPath(position)
                .replaceLastPositionIfBetter(position);
            return (/** @type {?} */ (this));
        }
        (/** @type {?} */ (this)).forceAddPosition(position);
        if (isNumber(calibrationMeasure)) {
            (/** @type {?} */ (this)).addCalibrationMeasureToLastPosition(calibrationMeasure);
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    startNewPart() {
        (/** @type {?} */ (this)).parts.push([]);
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} vertexIndex
     * @param {?=} partIndex
     * @return {THIS}
     */
    removePositionByIndex(vertexIndex, partIndex = 0) {
        /** @type {?} */
        const part = (/** @type {?} */ (this)).parts[partIndex];
        if (!part) {
            throw new Error(`Invalid part index" ${partIndex}`);
        }
        part.splice(vertexIndex, 1);
        /** @type {?} */
        const key = (/** @type {?} */ (this)).getCalibrationMeasureLookupKey(partIndex, vertexIndex);
        if ((/** @type {?} */ (this)).calibratedMeasures.has(key)) {
            (/** @type {?} */ (this)).calibratedMeasures.delete(key);
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} position
     * @return {THIS}
     */
    removePosition(position) {
        /** @type {?} */
        let partIndex = -1;
        /** @type {?} */
        let vertexIndex = -1;
        (/** @type {?} */ (this)).parts.forEach((/**
         * @param {?} part
         * @param {?} index
         * @return {?}
         */
        (part, index) => {
            /** @type {?} */
            const foundPositionIndex = part.indexOf(position);
            if (foundPositionIndex > -1) {
                partIndex = index;
                vertexIndex = foundPositionIndex;
            }
        }));
        if (partIndex > -1 && vertexIndex > -1) {
            return (/** @type {?} */ (this)).removePositionByIndex(vertexIndex, partIndex);
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    removeCalibrationMeasure(measure) {
        Array.from((/** @type {?} */ (this)).calibratedMeasures.values())
            .filter((/**
         * @param {?} calibration
         * @return {?}
         */
        calibration => calibration.measure === measure))
            .forEach((/**
         * @template THIS
         * @this {THIS}
         * @param {?} calibration
         * @return {THIS}
         */
        calibration => (/** @type {?} */ (this)).removeCalibrationMeasureByIndex(calibration.vertexIndex, calibration.partIndex)));
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} vertexIndex
     * @param {?=} partIndex
     * @return {THIS}
     */
    removeCalibrationMeasureByIndex(vertexIndex, partIndex = 0) {
        /** @type {?} */
        const key = (/** @type {?} */ (this)).getCalibrationMeasureLookupKey(partIndex, vertexIndex);
        if ((/** @type {?} */ (this)).calibratedMeasures.has(key)) {
            (/** @type {?} */ (this)).calibratedMeasures.delete(key);
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    addCalibrationMeasure(measure) {
        /** @type {?} */
        const timestamp = new Date().getTime();
        const [lastVertex] = (/** @type {?} */ (this)).currentPart.slice(-1);
        if (lastVertex && lastVertex.timestamp && lastVertex.timestamp === timestamp) {
            return (/** @type {?} */ (this)).addCalibrationMeasureToLastPosition(measure);
        }
        /** @type {?} */
        const partIndex = (/** @type {?} */ (this)).currentPartIndex;
        /** @type {?} */
        const key = UNPLACED_CALIBRATION_MEASURE_KEY;
        (/** @type {?} */ (this)).calibratedMeasures.set(key, {
            measure,
            partIndex,
            vertexIndex: null,
            timestamp
        });
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    addCalibrationMeasureToLastPosition(measure) {
        /** @type {?} */
        const partIndex = (/** @type {?} */ (this)).currentPartIndex;
        /** @type {?} */
        const lastVertexIndex = (/** @type {?} */ (this)).currentPart.length - 1;
        return (/** @type {?} */ (this)).addCalibrationMeasureAtIndex(measure, lastVertexIndex, partIndex);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @param {?} vertexIndex
     * @param {?=} partIndex
     * @return {THIS}
     */
    addCalibrationMeasureAtIndex(measure, vertexIndex, partIndex = 0) {
        /** @type {?} */
        const key = (/** @type {?} */ (this)).getCalibrationMeasureLookupKey(partIndex, vertexIndex);
        (/** @type {?} */ (this)).calibratedMeasures.set(key, {
            measure,
            partIndex,
            vertexIndex,
            timestamp: new Date().getTime()
        });
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} position
     * @return {THIS}
     */
    forceAddPosition(position) {
        (/** @type {?} */ (this)).tryStoreUnfilteredPath(position);
        (/** @type {?} */ (this)).addPositionToPath(position);
        return (/** @type {?} */ (this));
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toPointFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toPointGeometry(),
            properties
        };
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toLineFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toLineGeometry(),
            properties
        };
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toSingleLineFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toSingleLineGeometry(),
            properties
        };
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toMultiLineFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toMultiLineGeometry(),
            properties
        };
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toPolygonFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toPolygonGeometry(),
            properties
        };
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toMultiPolygonFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toMultiPolygonGeometry(),
            properties
        };
    }
    /**
     * @param {?=} properties
     * @return {?}
     */
    toSinglePolygonFeature(properties = {}) {
        return {
            type: 'Feature',
            geometry: this.toSinglePolygonGeometry(),
            properties
        };
    }
    /**
     * @return {?}
     */
    toFeature() {
        return this.isSinglePoint
            ? (/** @type {?} */ (this.toPointFeature()))
            : this.toLineFeature();
    }
    /**
     * @return {?}
     */
    toGeometry() {
        return this.isSinglePoint
            ? (/** @type {?} */ (this.toPointGeometry()))
            : this.toLineGeometry();
    }
    /**
     * @param {?=} calibrateMeasures
     * @return {?}
     */
    toSingleLineGeometry(calibrateMeasures = true) {
        /** @type {?} */
        const line = this.toLineGeometry(calibrateMeasures);
        if (isMultiPartLine(line.coordinates)) {
            throw new Error('Cannot convert MultiLineString to LineString. Data would be lost');
        }
        return (/** @type {?} */ (line));
    }
    /**
     * @param {?=} calibrateMeasures
     * @return {?}
     */
    toMultiLineGeometry(calibrateMeasures = true) {
        /** @type {?} */
        let line = this.toLineGeometry(calibrateMeasures);
        /** @type {?} */
        const isSinglePartLine = !isMultiPartLine(line.coordinates);
        if (isSinglePartLine) {
            line = {
                type: 'MultiLineString',
                bbox: line.bbox,
                coordinates: [(/** @type {?} */ (line.coordinates))]
            };
        }
        return (/** @type {?} */ (line));
    }
    /**
     * @param {?=} calibrateMeasures
     * @return {?}
     */
    toLineGeometry(calibrateMeasures = true) {
        /** @type {?} */
        const parts = this.toMultiPartArray();
        /** @type {?} */
        const line = parts.length > 1
            ? {
                type: 'MultiLineString',
                coordinates: parts
            }
            : {
                type: 'LineString',
                coordinates: parts[0]
            };
        if (this.options.returnMValues) {
            /** @type {?} */
            const calibrationPositions = calibrateMeasures
                ? Array.from(this.calibratedMeasures.values())
                : [];
            setMeasuresByGeometricLength(line, {
                units: this.options.mValueUnit,
                hasZValues: this.options.returnZValues,
                calibrationPositions
            });
        }
        return line;
    }
    /**
     * @return {?}
     */
    toMultiPolygonGeometry() {
        /** @type {?} */
        let poly = this.toPolygonGeometry();
        /** @type {?} */
        const isSinglePolygon = poly.type === 'Polygon';
        if (isSinglePolygon) {
            poly = {
                type: 'MultiPolygon',
                bbox: poly.bbox,
                coordinates: [(/** @type {?} */ (poly.coordinates))]
            };
        }
        return (/** @type {?} */ (poly));
    }
    /**
     * @return {?}
     */
    toSinglePolygonGeometry() {
        /** @type {?} */
        const poly = this.toPolygonGeometry();
        /** @type {?} */
        const isMultiPart = poly.type === 'MultiPolygon';
        if (isMultiPart) {
            throw new Error('Cannot convert MultiPolygon to Polygon. Data would be lost');
        }
        return (/** @type {?} */ (poly));
    }
    /**
     * @return {?}
     */
    toPolygonGeometry() {
        /** @type {?} */
        const parts = this.toMultiPartArray();
        parts.forEach((/**
         * @param {?} part
         * @return {?}
         */
        part => this.ensurePolygonIsClosed(part)));
        return parts.length > 1
            ? {
                type: 'MultiPolygon',
                coordinates: [parts]
            }
            : {
                type: 'Polygon',
                coordinates: parts
            };
    }
    /**
     * @return {?}
     */
    toPointGeometry() {
        /** @type {?} */
        const parts = this.toMultiPartArray();
        return parts.length > 1
            ? { type: 'MultiPoint', coordinates: parts.map((/**
                 * @param {?} part
                 * @return {?}
                 */
                part => part[0])) }
            : { type: 'Point', coordinates: parts[0][0] };
    }
    /**
     * @param {?} position
     * @return {?}
     */
    replaceLastPositionIfBetter(position) {
        /** @type {?} */
        let wasPositionAdded = true;
        if (this.isEmpty) {
            this.addPositionToPath(position);
            return wasPositionAdded;
        }
        /** @type {?} */
        const newPositionHasBetterAccuracy = this.lastPosition.accuracy > position.accuracy;
        if (newPositionHasBetterAccuracy) {
            this.replaceLastPosition(position);
        }
        else {
            wasPositionAdded = false;
        }
        return wasPositionAdded;
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} position
     * @return {THIS}
     */
    replaceLastPosition(position) {
        (/** @type {?} */ (this)).currentPart[(/** @type {?} */ (this)).currentPart.length - 1] = position;
        return (/** @type {?} */ (this));
    }
    /**
     * @return {?}
     */
    getBestPosition() {
        throw new Error('Not implemented yet');
    }
    /**
     * @return {?}
     */
    getUnfilteredPath() {
        return this.unfilteredPath.map((/**
         * @param {?} position
         * @return {?}
         */
        position => {
            /** @type {?} */
            const unfilteredGpsPosition = Object.assign({
                included: this.pathContains(position)
            }, position);
            return unfilteredGpsPosition;
        }));
    }
    /**
     * @return {?}
     */
    getGeometryPathSummary() {
        /** @type {?} */
        const line = this.toMultiLineGeometry(false);
        return line.coordinates.map((/**
         * @param {?} part
         * @param {?} partIndex
         * @return {?}
         */
        (part, partIndex) => this.getGeometryLinePartSummary(part, partIndex)));
    }
    /**
     * @param {?} position
     * @return {?}
     */
    pathContains(position) {
        /** @type {?} */
        const found = this.parts.find((/**
         * @param {?} part
         * @return {?}
         */
        part => part.indexOf(position) > -1));
        return Boolean(found);
    }
    /**
     * @param {?} partIndex
     * @param {?} vertexIndex
     * @return {?}
     */
    getPositionByIndex(partIndex, vertexIndex) {
        return this.parts[partIndex][vertexIndex];
    }
    /**
     * @return {?}
     */
    dispose() {
        if (this.trackSubscription) {
            this.trackSubscription.unsubscribe();
        }
        this.parts.length = 0;
    }
    /**
     * @private
     * @param {?} linePart
     * @param {?} partIndex
     * @return {?}
     */
    getGeometryLinePartSummary(linePart, partIndex) {
        return linePart.map((/**
         * @param {?} vertex
         * @param {?} vertexIndex
         * @return {?}
         */
        (vertex, vertexIndex) => {
            /** @type {?} */
            const position = this.getPositionByIndex(partIndex, vertexIndex);
            /** @type {?} */
            const calibrationKey = this.getCalibrationMeasureLookupKey(partIndex, vertexIndex);
            /** @type {?} */
            const calibrationPointMeasureValue = this.calibratedMeasures.has(calibrationKey)
                ? this.calibratedMeasures.get(calibrationKey).measure
                : PATH_BUILDER_EMPTY_CALIBRATION_MEASURE_VALUE;
            return Object.assign({
                geometricMeasure: getMeasureFromPoint(vertex),
                partIndex,
                vertexIndex,
                calibrationPointMeasureValue
            }, position);
        }));
    }
    /**
     * @private
     * @param {?} partIndex
     * @param {?} vertexIndex
     * @return {?}
     */
    getCalibrationMeasureLookupKey(partIndex, vertexIndex) {
        return `${partIndex}_${vertexIndex}`;
    }
    /**
     * @private
     * @return {?}
     */
    toMultiPartArray() {
        /** @type {?} */
        const parts = [];
        this.parts.forEach((/**
         * @param {?} part
         * @return {?}
         */
        part => {
            /** @type {?} */
            const coordinates = part.map((/**
             * @param {?} position
             * @return {?}
             */
            position => this.options.returnZValues
                ? [position.longitude, position.latitude, position.altitude || 0]
                : [position.longitude, position.latitude]));
            parts.push(coordinates);
        }));
        return parts;
    }
    /**
     * @private
     * @param {?} polygonPoints
     * @return {?}
     */
    ensurePolygonIsClosed(polygonPoints) {
        if (polygonPoints.length > 1) {
            const [first] = polygonPoints;
            /** @type {?} */
            const last = getLastItem(polygonPoints);
            const [firstLng, firstLat, firstZ] = first;
            const [lastLng, lastLat] = last;
            /** @type {?} */
            const doesPolygonNotClose = !this.arePointsEquivalent(firstLng, firstLat, lastLng, lastLat);
            if (doesPolygonNotClose) {
                /** @type {?} */
                const closingPoint = [firstLng, firstLat, firstZ || 0];
                polygonPoints.push(closingPoint);
            }
        }
    }
    /**
     * @private
     * @param {?} position
     * @return {?}
     */
    addPositionToPath(position) {
        this.currentPart.push(position);
        if (this.calibratedMeasures.has(UNPLACED_CALIBRATION_MEASURE_KEY)) {
            this.addPendingCalibrationPoint();
        }
    }
    /**
     * @private
     * @return {?}
     */
    addPendingCalibrationPoint() {
        /** @type {?} */
        const calibratonPosition = this.calibratedMeasures.get(UNPLACED_CALIBRATION_MEASURE_KEY);
        /** @type {?} */
        const calibrationTimestamp = calibratonPosition.timestamp;
        const [secondToLastPosition, lastPosition] = this.currentPart.slice(-2);
        /** @type {?} */
        const isTimestampBetweenVertices = calibrationTimestamp > secondToLastPosition.timestamp
            && calibrationTimestamp < lastPosition.timestamp;
        if (isTimestampBetweenVertices) {
            /** @type {?} */
            const totalTimeDelta = Math.abs(lastPosition.timestamp - secondToLastPosition.timestamp);
            /** @type {?} */
            const offsetTimeDelta = Math.abs(calibrationTimestamp - secondToLastPosition.timestamp);
            /** @type {?} */
            const percentOfDistance = offsetTimeDelta / totalTimeDelta;
            /** @type {?} */
            const distanceBetweenVerticesKilometers = getDistanceBetweenLatLng(secondToLastPosition.latitude, secondToLastPosition.longitude, lastPosition.latitude, lastPosition.longitude, LengthUnit.Mile);
            /** @type {?} */
            const distanceToPreviousVertexKilometers = percentOfDistance * distanceBetweenVerticesKilometers;
            /** @type {?} */
            const didSnapCalibrationMeasureToVertes = this.trySnapCalibrationMeasureToVertex(calibratonPosition.measure, distanceBetweenVerticesKilometers, distanceToPreviousVertexKilometers);
            if (didSnapCalibrationMeasureToVertes) {
                return;
            }
            /** @type {?} */
            const heading = bearing([secondToLastPosition.longitude, secondToLastPosition.latitude], [lastPosition.longitude, lastPosition.latitude]);
            /** @type {?} */
            const newPoint = destination([secondToLastPosition.longitude, secondToLastPosition.latitude], distanceToPreviousVertexKilometers, heading, { units: 'kilometers' });
            /** @type {?} */
            const spliceIndexForSecondToLastPosition = this.currentPart.length - 1;
            /** @type {?} */
            const newPosition = Object.assign({}, lastPosition, {
                longitude: newPoint.geometry.coordinates[0],
                latitude: newPoint.geometry.coordinates[1],
                deviceSource: PATH_BUILDER_CALCULATED_M_DEVICE_SOURCE
            });
            this.calibratedMeasures.delete(UNPLACED_CALIBRATION_MEASURE_KEY);
            this.addCalibrationMeasureAtIndex(calibratonPosition.measure, spliceIndexForSecondToLastPosition, this.currentPartIndex);
            this.currentPart.splice(spliceIndexForSecondToLastPosition, 0, newPosition);
        }
        else if (calibrationTimestamp >= lastPosition.timestamp) {
            this.addCalibrationMeasureToLastPosition(calibratonPosition.measure);
        }
        else {
            throw new Error(`Calibration measure timestamp '${calibratonPosition.timestamp}' can not be located between`
                + ` '${secondToLastPosition.timestamp}' and '${lastPosition.timestamp}'`);
        }
    }
    /**
     * @private
     * @param {?} calibrationMeasure
     * @param {?} totalVerticeDistanceKilometers
     * @param {?} distanceToPreviousVertexKilometers
     * @return {?}
     */
    trySnapCalibrationMeasureToVertex(calibrationMeasure, totalVerticeDistanceKilometers, distanceToPreviousVertexKilometers) {
        /** @type {?} */
        const distanceToNextVertexKilometers = totalVerticeDistanceKilometers - distanceToPreviousVertexKilometers;
        /** @type {?} */
        const distanceThresholdKilometers = convertLengthUnits(this.options.calibrationMeasureSnappingDistanceDelta, this.options.calibrationMeasureSnappingDistanceDeltaUnit, LengthUnit.Kilometer);
        /** @type {?} */
        const isPointCloseEnoughToPrevious = distanceToPreviousVertexKilometers < distanceThresholdKilometers;
        /** @type {?} */
        const isPointCloseEnoughToNext = distanceToNextVertexKilometers < distanceThresholdKilometers;
        if (isPointCloseEnoughToPrevious || isPointCloseEnoughToNext) {
            /** @type {?} */
            const previousVertexIndex = this.currentPart.length - 2;
            /** @type {?} */
            const nextVertexIndex = this.currentPart.length - 1;
            /** @type {?} */
            const vertexIndex = distanceToPreviousVertexKilometers < distanceToNextVertexKilometers
                ? previousVertexIndex
                : nextVertexIndex;
            this.addCalibrationMeasureAtIndex(calibrationMeasure, vertexIndex, this.currentPartIndex);
            this.calibratedMeasures.delete(UNPLACED_CALIBRATION_MEASURE_KEY);
            return true;
        }
        return false;
    }
    /**
     * @private
     * @param {?} position1
     * @param {?} position2
     * @return {?}
     */
    arePositionsEquivalent(position1, position2) {
        return this.arePointsEquivalent(position1.longitude, position1.latitude, position2.longitude, position2.latitude);
    }
    /**
     * @private
     * @param {?} longitude1
     * @param {?} latitude1
     * @param {?} longitude2
     * @param {?} latitude2
     * @return {?}
     */
    arePointsEquivalent(longitude1, latitude1, longitude2, latitude2) {
        return pointsAlmostEqual(longitude1, latitude1, longitude2, latitude2, this.options.pointEquivalencyPrecision);
    }
    /**
     * @private
     * @template THIS
     * @this {THIS}
     * @param {?} position
     * @return {THIS}
     */
    tryStoreUnfilteredPath(position) {
        if ((/** @type {?} */ (this)).options.storeUnfilteredPath) {
            (/** @type {?} */ (this)).unfilteredPath.push(position);
        }
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    PositionPathBuilder.prototype.parts;
    /** @type {?} */
    PositionPathBuilder.prototype.unfilteredPath;
    /** @type {?} */
    PositionPathBuilder.prototype.pathBreaks;
    /** @type {?} */
    PositionPathBuilder.prototype.calibratedMeasures;
    /**
     * @type {?}
     * @private
     */
    PositionPathBuilder.prototype.options;
    /**
     * @type {?}
     * @private
     */
    PositionPathBuilder.prototype.trackSubscription;
}
