// @tss/sdk/esm5/lib/app/api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { AppConfig } from './app.config';
import { BaseEmptyResourceService } from '../core/resource/base-empty-resource.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./app.config";
/**
 * @record
 */
export function ApiStatus() { }
if (false) {
    /** @type {?} */
    ApiStatus.prototype.up;
    /** @type {?} */
    ApiStatus.prototype.down;
    /** @type {?} */
    ApiStatus.prototype.configStoreUp;
    /** @type {?} */
    ApiStatus.prototype.configStoreDown;
    /** @type {?} */
    ApiStatus.prototype.upSince;
    /** @type {?} */
    ApiStatus.prototype.daysUp;
    /** @type {?} */
    ApiStatus.prototype.appCount;
}
var ApiService = /** @class */ (function (_super) {
    tslib_1.__extends(ApiService, _super);
    function ApiService(http, appConfig) {
        var _this = _super.call(this, http, appConfig.serviceUrl) || this;
        _this.appConfig = appConfig;
        return _this;
    }
    /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    ApiService.prototype.ping = /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    function (routeTokens, params) {
        /** @type {?} */
        var url = this.getUrl(this.appConfig.pingEndpoint, routeTokens);
        return this.http.get(url, { params: params });
    };
    /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    ApiService.prototype.status = /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    function (routeTokens, params) {
        /** @type {?} */
        var url = this.getUrl(this.appConfig.pingEndpoint, routeTokens);
        return this.http.get(url, { params: params });
    };
    ApiService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ApiService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: AppConfig }
    ]; };
    /** @nocollapse */ ApiService.ngInjectableDef = i0.defineInjectable({ factory: function ApiService_Factory() { return new ApiService(i0.inject(i1.HttpClient), i0.inject(i2.AppConfig)); }, token: ApiService, providedIn: "root" });
    return ApiService;
}(BaseEmptyResourceService));
export { ApiService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ApiService.prototype.appConfig;
}
