// @tss/sdk/esm5/lib/app/app-profile.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { AppConfig } from './app.config';
import { BaseResourceService } from '../core/resource/base-resource.service';
import { join } from '../core/url/join.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./app.config";
var AppProfileService = /** @class */ (function (_super) {
    tslib_1.__extends(AppProfileService, _super);
    function AppProfileService(http, config) {
        var _this = _super.call(this, http, {
            url: join(config.serviceUrl, config.appProfileEndpoint),
            defaultRouteTokens: {
                appId: config.appId
            }
        }) || this;
        _this.http = http;
        return _this;
    }
    AppProfileService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AppProfileService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: AppConfig }
    ]; };
    /** @nocollapse */ AppProfileService.ngInjectableDef = i0.defineInjectable({ factory: function AppProfileService_Factory() { return new AppProfileService(i0.inject(i1.HttpClient), i0.inject(i2.AppConfig)); }, token: AppProfileService, providedIn: "root" });
    return AppProfileService;
}(BaseResourceService));
export { AppProfileService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AppProfileService.prototype.http;
}
