// @tss/sdk/esm5/lib/app/app-spatial.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
// TSS.
import { TssGeometryModule } from '../geometry/geometry.module';
import { TssEsriModule } from '../esri/esri.module';
import { TssMapsModule } from '../maps/map.module';
import { TssLrsModule } from '../lrs/lrs.module';
import { TssAppModule } from './app.module';
var TssAppSpatialModule = /** @class */ (function () {
    function TssAppSpatialModule() {
    }
    TssAppSpatialModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        TssAppModule,
                        TssEsriModule,
                        TssGeometryModule,
                        TssMapsModule,
                        TssLrsModule,
                    ],
                    declarations: [],
                    exports: [],
                    providers: [],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                },] }
    ];
    return TssAppSpatialModule;
}());
export { TssAppSpatialModule };
