// @tss/sdk/esm2015/lib/maps/drawing/layer-drawing-info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LayerDrawingInfoRenderer } from './layer-drawing-info-renderer';
import { transparencyToOpacity } from '../../core/color/transparency-to-opacity.function';
/**
 * @record
 */
export function DrawingInfoProperties() { }
if (false) {
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.renderer;
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.transparency;
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.positionRenderer;
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.labelingInfo;
}
export class LayerDrawingInfo {
    /**
     * @return {?}
     */
    get isEmpty() {
        return !this.transparency
            && !this.zIndex
            && this.renderer.isEmpty
            && this.positionRenderer.isEmpty
            && !Object.keys(this.labelingInfo || {}).length;
    }
    /**
     * @return {?}
     */
    get opacity() {
        return transparencyToOpacity(this.transparency);
    }
    /**
     * @param {...?} baseDrawingInfos
     */
    constructor(...baseDrawingInfos) {
        this.merge(...baseDrawingInfos);
        this.transparency = this.transparency || 0;
        this.zIndex = this.zIndex || null;
        this.renderer = new LayerDrawingInfoRenderer(this.renderer);
        this.positionRenderer = new LayerDrawingInfoRenderer(this.positionRenderer);
        this.labelingInfo = this.labelingInfo || null;
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {...?} targetDrawingInfos
     * @return {THIS}
     */
    merge(...targetDrawingInfos) {
        /** @type {?} */
        const properties = targetDrawingInfos.map((/**
         * @param {?} info
         * @return {?}
         */
        info => info && ((/** @type {?} */ (info))).toProperties
            ? ((/** @type {?} */ (info))).toProperties()
            : info));
        Object.assign((/** @type {?} */ (this)), ...properties);
        return (/** @type {?} */ (this));
    }
    /**
     * @param {?=} value
     * @return {?}
     */
    getRendererSymbol(value) {
        return this.renderer.getSymbol(value);
    }
    /**
     * @param {?=} value
     * @return {?}
     */
    getPositionRendererSymbol(value) {
        return this.positionRenderer.getSymbol(value);
    }
    /**
     * @param {?=} value
     * @return {?}
     */
    getBestRepresentativeColor(value) {
        return this.renderer.getBestRepresentativeColor(value);
    }
    /**
     * @return {?}
     */
    toProperties() {
        /** @type {?} */
        const infoProperties = {};
        if (this.labelingInfo) {
            infoProperties.labelingInfo = this.labelingInfo;
        }
        if (this.positionRenderer && !this.positionRenderer.isEmpty) {
            infoProperties.positionRenderer = this.positionRenderer;
        }
        if (this.renderer && !this.renderer.isEmpty) {
            infoProperties.renderer = this.renderer;
        }
        if (this.transparency) {
            infoProperties.transparency = this.transparency;
        }
        return infoProperties;
    }
}
if (false) {
    /** @type {?} */
    LayerDrawingInfo.prototype.renderer;
    /** @type {?} */
    LayerDrawingInfo.prototype.transparency;
    /** @type {?} */
    LayerDrawingInfo.prototype.zIndex;
    /** @type {?} */
    LayerDrawingInfo.prototype.positionRenderer;
    /** @type {?} */
    LayerDrawingInfo.prototype.labelingInfo;
}
