// @tss/sdk/esm5/lib/lrs/data-source-lrs-metadata.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { LrsConfig } from './lrs.config';
import { join } from '../core/url/join.function';
import { BaseResourceService } from '../core/resource/base-resource.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./lrs.config";
var DataSourceLrsMetadataService = /** @class */ (function (_super) {
    tslib_1.__extends(DataSourceLrsMetadataService, _super);
    function DataSourceLrsMetadataService(http, config) {
        var _this = _super.call(this, http, {
            url: join(config.serviceUrl, config.dataSourceLrsMetadataEndpoint),
            idFieldName: 'dataSourceId'
        }) || this;
        _this.http = http;
        return _this;
    }
    /**
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    DataSourceLrsMetadataService.prototype.create = /**
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    function (entity, routeTokens, params) {
        routeTokens = routeTokens || {};
        routeTokens.id = entity.dataSourceId;
        return _super.prototype.create.call(this, entity, routeTokens, params);
    };
    DataSourceLrsMetadataService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DataSourceLrsMetadataService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: LrsConfig }
    ]; };
    /** @nocollapse */ DataSourceLrsMetadataService.ngInjectableDef = i0.defineInjectable({ factory: function DataSourceLrsMetadataService_Factory() { return new DataSourceLrsMetadataService(i0.inject(i1.HttpClient), i0.inject(i2.LrsConfig)); }, token: DataSourceLrsMetadataService, providedIn: "root" });
    return DataSourceLrsMetadataService;
}(BaseResourceService));
export { DataSourceLrsMetadataService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    DataSourceLrsMetadataService.prototype.http;
}
