// @tss/sdk/esm5/lib/lrs/lrs-metadata.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
// 3rd Party.
import { of } from 'rxjs';
import { tap } from 'rxjs/operators';
// TSS.
import { DataSourceService } from '../data-source/data-source.service';
import { DataSourceLrsMetadataService } from './data-source-lrs-metadata.service';
import { Logger } from '../core/logging/logger.service';
import { NotifyService } from '../core/notify/notify.service';
import { LrsMetadataRetriever } from './lrs-metadata-retriever';
import { LrsConfig } from './lrs.config';
import { EsriMapServerService } from '../esri/esri-map-server.service';
import { getDataSourceId } from '../data-source/get-data-source-id.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../esri/esri-map-server.service";
import * as i3 from "../data-source/data-source.service";
import * as i4 from "./data-source-lrs-metadata.service";
import * as i5 from "../core/notify/notify.service";
import * as i6 from "./lrs.config";
import * as i7 from "../core/logging/logger.service";
var LrsMetadataService = /** @class */ (function () {
    function LrsMetadataService(http, mapServer, dataSourceService, dataSourceLrsMetadataService, notify, config, log) {
        this.http = http;
        this.mapServer = mapServer;
        this.dataSourceService = dataSourceService;
        this.dataSourceLrsMetadataService = dataSourceLrsMetadataService;
        this.notify = notify;
        this.config = config;
        this.log = log;
        this.lrsCache = [];
    }
    /**
     * @param {?} dataSource
     * @param {?=} lrsMetadata
     * @return {?}
     */
    LrsMetadataService.prototype.get = /**
     * @param {?} dataSource
     * @param {?=} lrsMetadata
     * @return {?}
     */
    function (dataSource, lrsMetadata) {
        var _this = this;
        /** @type {?} */
        var dataSourceId = getDataSourceId(dataSource);
        /** @type {?} */
        var cachedMeta = this.lrsCache.find((/**
         * @param {?} cache
         * @return {?}
         */
        function (cache) { return cache.dataSourceId === dataSourceId; }));
        if (cachedMeta) {
            return of(cachedMeta.metaData);
        }
        return new LrsMetadataRetriever(this.http, this.mapServer, this.dataSourceService, this.dataSourceLrsMetadataService, this.notify, this.log)
            .withDataSource(dataSource)
            .withLrsMetadata(lrsMetadata)
            .retrieve()
            .pipe(tap((/**
         * @param {?} meta
         * @return {?}
         */
        function (meta) { return _this.tryCacheLrsMetadata(meta); })));
    };
    /**
     * @private
     * @param {?} meta
     * @return {?}
     */
    LrsMetadataService.prototype.tryCacheLrsMetadata = /**
     * @private
     * @param {?} meta
     * @return {?}
     */
    function (meta) {
        if (this.config.cacheInSessionLrsMetadata) {
            this.lrsCache.push({
                dataSourceId: meta.dataSourceId,
                metaData: meta
            });
        }
    };
    LrsMetadataService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    LrsMetadataService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: EsriMapServerService },
        { type: DataSourceService },
        { type: DataSourceLrsMetadataService },
        { type: NotifyService },
        { type: LrsConfig },
        { type: Logger }
    ]; };
    /** @nocollapse */ LrsMetadataService.ngInjectableDef = i0.defineInjectable({ factory: function LrsMetadataService_Factory() { return new LrsMetadataService(i0.inject(i1.HttpClient), i0.inject(i2.EsriMapServerService), i0.inject(i3.DataSourceService), i0.inject(i4.DataSourceLrsMetadataService), i0.inject(i5.NotifyService), i0.inject(i6.LrsConfig), i0.inject(i7.Logger)); }, token: LrsMetadataService, providedIn: "root" });
    return LrsMetadataService;
}());
export { LrsMetadataService };
if (false) {
    /** @type {?} */
    LrsMetadataService.prototype.lrsCache;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.http;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.mapServer;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.dataSourceService;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.dataSourceLrsMetadataService;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.notify;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    LrsMetadataService.prototype.log;
}
