// @tss/sdk/esm5/lib/lrs/lrs-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { MapFeature } from '../maps/map-feature';
import { isLine } from '../geometry/geometry/is-line.function';
import { isPoint } from '../geometry/geometry/is-point.function';
import { WGS_84_SRID } from '../geometry/geometry.constants';
import { clone } from '../core/clone/clone.function';
var LrsFeature = /** @class */ (function (_super) {
    tslib_1.__extends(LrsFeature, _super);
    function LrsFeature(feature, layer, spatialReferenceId) {
        if (spatialReferenceId === void 0) { spatialReferenceId = WGS_84_SRID; }
        var _this = _super.call(this, feature, layer, spatialReferenceId) || this;
        if (_this.geometry && _this.isGeometryTypeInvalid()) {
            throw new Error("Unsupported geometry type of " + _this.geometry.type + " detected.");
        }
        return _this;
    }
    Object.defineProperty(LrsFeature.prototype, "layerId", {
        get: /**
         * @return {?}
         */
        function () {
            return this.layer ? this.layer.id : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LrsFeature.prototype, "fromDate", {
        get: /**
         * @return {?}
         */
        function () {
            return this.getValue(this.layer.fromDateFieldName);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LrsFeature.prototype, "toDate", {
        get: /**
         * @return {?}
         */
        function () {
            return this.getValue(this.layer.toDateFieldName);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LrsFeature.prototype, "isRetired", {
        get: /**
         * @return {?}
         */
        function () {
            return Boolean(this.toDate);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LrsFeature.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.isRetired;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} date
     * @return {THIS}
     */
    LrsFeature.prototype.setFromDate = /**
     * @template THIS
     * @this {THIS}
     * @param {?} date
     * @return {THIS}
     */
    function (date) {
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.fromDateFieldName, date);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} date
     * @return {THIS}
     */
    LrsFeature.prototype.setToDate = /**
     * @template THIS
     * @this {THIS}
     * @param {?} date
     * @return {THIS}
     */
    function (date) {
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toDateFieldName, date);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    LrsFeature.prototype.retire = /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        if ((/** @type {?} */ (this)).layer.toDateFieldName) {
            (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toDateFieldName, new Date());
        }
        return (/** @type {?} */ (this));
    };
    /**
     * @return {?}
     */
    LrsFeature.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var copyFeature = clone(this.toFeature());
        return new LrsFeature(copyFeature, this.layer);
    };
    /**
     * @protected
     * @return {?}
     */
    LrsFeature.prototype.isGeometryTypeValid = /**
     * @protected
     * @return {?}
     */
    function () {
        return isLine(this.geometry) || isPoint(this.geometry);
    };
    /**
     * @private
     * @return {?}
     */
    LrsFeature.prototype.isGeometryTypeInvalid = /**
     * @private
     * @return {?}
     */
    function () {
        return !this.isGeometryTypeValid();
    };
    return LrsFeature;
}(MapFeature));
export { LrsFeature };
if (false) {
    /** @type {?} */
    LrsFeature.prototype.routeId;
    /** @type {?} */
    LrsFeature.prototype.fromMeasure;
    /** @type {?} */
    LrsFeature.prototype.toMeasure;
    /** @type {?} */
    LrsFeature.prototype.layer;
    /** @type {?} */
    LrsFeature.prototype.displayValue;
    /** @type {?} */
    LrsFeature.prototype.styleValue;
    /** @type {?} */
    LrsFeature.prototype.viewDate;
    /** @type {?} */
    LrsFeature.prototype.gdbVersion;
}
