// @tss/sdk/esm5/lib/geometry/line/get-distance-to-line.function.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { toSinglePartPointArray } from '../point/to-single-part-point.function';
/**
 * @record
 */
function Point() { }
if (false) {
    /** @type {?} */
    Point.prototype.x;
    /** @type {?} */
    Point.prototype.y;
}
/**
 * @param {?} x
 * @return {?}
 */
function squareRoot(x) {
    return x * x;
}
/**
 * @param {?} v
 * @param {?} w
 * @return {?}
 */
function squareRootDistance(v, w) {
    return squareRoot(v.x - w.x) + squareRoot(v.y - w.y);
}
/**
 * @param {?} testPoint
 * @param {?} lineStartPoint
 * @param {?} lineEndPoint
 * @return {?}
 */
function distanceToSegmentSquared(testPoint, lineStartPoint, lineEndPoint) {
    /** @type {?} */
    var lengthSquared = squareRootDistance(lineStartPoint, lineEndPoint);
    if (lengthSquared === 0) {
        return squareRootDistance(testPoint, lineStartPoint);
    }
    /** @type {?} */
    var part1 = (testPoint.x - lineStartPoint.x) * (lineEndPoint.x - lineStartPoint.x);
    /** @type {?} */
    var part2 = (testPoint.y - lineStartPoint.y) * (lineEndPoint.y - lineStartPoint.y);
    /** @type {?} */
    var formulaPart1 = part1 + part2;
    /** @type {?} */
    var delta = formulaPart1 / lengthSquared;
    if (delta < 0) {
        return squareRootDistance(testPoint, lineStartPoint);
    }
    if (delta > 1) {
        return squareRootDistance(testPoint, lineEndPoint);
    }
    /** @type {?} */
    var xValue = lineStartPoint.x + delta * (lineEndPoint.x - lineStartPoint.x);
    /** @type {?} */
    var yValue = lineStartPoint.y + delta * (lineEndPoint.y - lineStartPoint.y);
    return squareRootDistance(testPoint, { x: xValue, y: yValue });
}
/**
 * @param {?} testPoint
 * @param {?} lineStartPoint
 * @param {?} lineEndPoint
 * @return {?}
 */
function distToSegment(testPoint, lineStartPoint, lineEndPoint) {
    return Math.sqrt(distanceToSegmentSquared(testPoint, lineStartPoint, lineEndPoint));
}
/**
 * @param {?} testPoint
 * @param {?} lineStartPoint
 * @param {?} lineEndPoint
 * @return {?}
 */
export function getDistanceToLine(testPoint, lineStartPoint, lineEndPoint) {
    /** @type {?} */
    var testPointArray = toSinglePartPointArray(testPoint);
    /** @type {?} */
    var lineStartPointArray = toSinglePartPointArray(lineStartPoint);
    /** @type {?} */
    var lineEndPointArray = toSinglePartPointArray(lineEndPoint);
    var _a = tslib_1.__read(testPointArray, 2), testPointX = _a[0], testPointY = _a[1];
    var _b = tslib_1.__read(lineStartPointArray, 2), lineStartX = _b[0], lineStartY = _b[1];
    var _c = tslib_1.__read(lineEndPointArray, 2), lineEndX = _c[0], lineEndY = _c[1];
    return distToSegment({ x: testPointX, y: testPointY }, { x: lineStartX, y: lineStartY }, { x: lineEndX, y: lineEndY });
}
