// @tss/sdk/esm2015/lib/geometry/line/get-measure.function.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { lineString } from '@turf/helpers';
import nearestPointOnLine from '@turf/nearest-point-on-line';
import lineSplit from '@turf/line-split';
import length from '@turf/length';
import { getMeasureFromPoint } from '../point/get-measure-from-point.function';
import { toSinglePartPointArray } from '../point/to-single-part-point.function';
import { isMultiPartLine } from './is-multi-part.function';
import { toLineOrMultiLineString } from './to-line-or-multi-line-string.function';
import { getDistanceBetweenPoints } from '../point/get-distance-between-points.function';
import { getLastItem } from '../../core/array/get-last-item.function';
/**
 * @param {?} line
 * @param {?} point
 * @param {?=} options
 * @return {?}
 */
export function getMeasure(line, point, options) {
    options = Object.assign({
        lengthUnit: 'miles',
        coordinateSystem: 'wgs84'
    }, options);
    /** @type {?} */
    const closestPointFeature = nearestPointOnLine(line, point);
    if (!closestPointFeature || !closestPointFeature.geometry) {
        throw new Error('Unable to get measure for geometry');
    }
    /** @type {?} */
    const foundPointMeasure = getMeasureFromPoint(closestPointFeature.geometry);
    // If the point happens to land on a verticee, then it will already have a measure.
    if (foundPointMeasure !== null) {
        return foundPointMeasure;
    }
    /** @type {?} */
    const lineGeometry = toLineOrMultiLineString(line);
    if (isMultiPartLine(lineGeometry.coordinates)) {
        throw new Error('MultiLineString is not supported in getMeasure function yet.');
    }
    return interpolateMeasure((/** @type {?} */ (lineGeometry)), closestPointFeature.geometry, options);
}
/**
 * @param {?} line
 * @param {?} pointOnLine
 * @param {?} options
 * @return {?}
 */
function interpolateMeasure(line, pointOnLine, options) {
    /** @type {?} */
    const pointCoordinates = toSinglePartPointArray(pointOnLine);
    /** @type {?} */
    const pointGeometry = { type: 'Point', coordinates: pointCoordinates };
    /** @type {?} */
    const pointFeature = { type: 'Feature', geometry: pointGeometry, properties: {} };
    /** @type {?} */
    const lineFeature = {
        type: 'Feature',
        geometry: line,
        properties: {}
    };
    /** @type {?} */
    const splitResult = lineSplit(lineFeature, pointFeature);
    /** @type {?} */
    const firstFeature = splitResult.features[0];
    /** @type {?} */
    const splitVertex = getLastItem(firstFeature.geometry.coordinates);
    /** @type {?} */
    const lastVertexWithMeasureIndex = firstFeature.geometry.coordinates.length - 2;
    /** @type {?} */
    const lastVertexWithMeasure = firstFeature.geometry.coordinates[lastVertexWithMeasureIndex];
    /** @type {?} */
    const newLineSegment = lineString([lastVertexWithMeasure, splitVertex]);
    /** @type {?} */
    const measureOffset = options.coordinateSystem === 'flat'
        ? getDistanceBetweenPoints(lastVertexWithMeasure, splitVertex)
        : length(newLineSegment, { units: (/** @type {?} */ (options.lengthUnit)) });
    /** @type {?} */
    const lastVertexMeasure = getMeasureFromPoint(lastVertexWithMeasure);
    return lastVertexMeasure + measureOffset;
}
