// @tss/sdk/esm2015/lib/geometry/line/get-distance-to-line.function.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { toSinglePartPointArray } from '../point/to-single-part-point.function';
/**
 * @record
 */
function Point() { }
if (false) {
    /** @type {?} */
    Point.prototype.x;
    /** @type {?} */
    Point.prototype.y;
}
/**
 * @param {?} x
 * @return {?}
 */
function squareRoot(x) {
    return x * x;
}
/**
 * @param {?} v
 * @param {?} w
 * @return {?}
 */
function squareRootDistance(v, w) {
    return squareRoot(v.x - w.x) + squareRoot(v.y - w.y);
}
/**
 * @param {?} testPoint
 * @param {?} lineStartPoint
 * @param {?} lineEndPoint
 * @return {?}
 */
function distanceToSegmentSquared(testPoint, lineStartPoint, lineEndPoint) {
    /** @type {?} */
    const lengthSquared = squareRootDistance(lineStartPoint, lineEndPoint);
    if (lengthSquared === 0) {
        return squareRootDistance(testPoint, lineStartPoint);
    }
    /** @type {?} */
    const part1 = (testPoint.x - lineStartPoint.x) * (lineEndPoint.x - lineStartPoint.x);
    /** @type {?} */
    const part2 = (testPoint.y - lineStartPoint.y) * (lineEndPoint.y - lineStartPoint.y);
    /** @type {?} */
    const formulaPart1 = part1 + part2;
    /** @type {?} */
    const delta = formulaPart1 / lengthSquared;
    if (delta < 0) {
        return squareRootDistance(testPoint, lineStartPoint);
    }
    if (delta > 1) {
        return squareRootDistance(testPoint, lineEndPoint);
    }
    /** @type {?} */
    const xValue = lineStartPoint.x + delta * (lineEndPoint.x - lineStartPoint.x);
    /** @type {?} */
    const yValue = lineStartPoint.y + delta * (lineEndPoint.y - lineStartPoint.y);
    return squareRootDistance(testPoint, { x: xValue, y: yValue });
}
/**
 * @param {?} testPoint
 * @param {?} lineStartPoint
 * @param {?} lineEndPoint
 * @return {?}
 */
function distToSegment(testPoint, lineStartPoint, lineEndPoint) {
    return Math.sqrt(distanceToSegmentSquared(testPoint, lineStartPoint, lineEndPoint));
}
/**
 * @param {?} testPoint
 * @param {?} lineStartPoint
 * @param {?} lineEndPoint
 * @return {?}
 */
export function getDistanceToLine(testPoint, lineStartPoint, lineEndPoint) {
    /** @type {?} */
    const testPointArray = toSinglePartPointArray(testPoint);
    /** @type {?} */
    const lineStartPointArray = toSinglePartPointArray(lineStartPoint);
    /** @type {?} */
    const lineEndPointArray = toSinglePartPointArray(lineEndPoint);
    const [testPointX, testPointY] = testPointArray;
    const [lineStartX, lineStartY] = lineStartPointArray;
    const [lineEndX, lineEndY] = lineEndPointArray;
    return distToSegment({ x: testPointX, y: testPointY }, { x: lineStartX, y: lineStartY }, { x: lineEndX, y: lineEndY });
}
