// @tss/sdk/esm5/lib/core/logging/log-level.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var LogLevel = {
    Trace: 0,
    Debug: 1,
    Info: 2,
    Warn: 3,
    Error: 4,
    Fatal: 5,
};
export { LogLevel };
LogLevel[LogLevel.Trace] = 'Trace';
LogLevel[LogLevel.Debug] = 'Debug';
LogLevel[LogLevel.Info] = 'Info';
LogLevel[LogLevel.Warn] = 'Warn';
LogLevel[LogLevel.Error] = 'Error';
LogLevel[LogLevel.Fatal] = 'Fatal';
/**
 * @param {?} type
 * @return {?}
 */
export function typeToLevel(type) {
    if (type === 'trace') {
        return LogLevel.Trace;
    }
    else if (type === 'debug') {
        return LogLevel.Debug;
    }
    else if (type === 'info') {
        return LogLevel.Info;
    }
    else if (type === 'warn') {
        return LogLevel.Warn;
    }
    else if (type === 'error') {
        return LogLevel.Error;
    }
    else if (type === 'fatal') {
        return LogLevel.Fatal;
    }
    else {
        throw new Error("Unknown log type: '" + type + "'");
    }
}
/**
 * @param {?} level
 * @return {?}
 */
export function levelToType(level) {
    if (level === LogLevel.Trace) {
        return 'trace';
    }
    else if (level === LogLevel.Debug) {
        return 'debug';
    }
    else if (level === LogLevel.Info) {
        return 'info';
    }
    else if (level === LogLevel.Warn) {
        return 'warn';
    }
    else if (level === LogLevel.Error) {
        return 'error';
    }
    else if (level === LogLevel.Fatal) {
        return 'fatal';
    }
    else {
        throw new Error("Unknown log level: '" + level + "'");
    }
}
