// @tss/sdk/esm5/lib/lrs/events/event-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// TSS.
import { LrsFeature } from '../lrs-feature';
import { toFeature } from '../feature';
import { WGS_84_SRID } from '../../geometry/geometry.constants';
import { clone } from '../../core/clone/clone.function';
var EventFeature = /** @class */ (function (_super) {
    tslib_1.__extends(EventFeature, _super);
    function EventFeature(feature, layer, spatialReferenceId) {
        if (spatialReferenceId === void 0) { spatialReferenceId = WGS_84_SRID; }
        var _this = _super.call(this, feature, layer, spatialReferenceId) || this;
        _this.routeId = _this.getValue(layer.routeIdFieldName);
        _this.fromMeasure = _this.getValue(layer.fromMeasureFieldName);
        _this.toMeasure = _this.getValue(layer.toMeasureFieldName);
        return _this;
    }
    /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    EventFeature.create = /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    function (layer, properties, geometry) {
        if (properties === void 0) { properties = {}; }
        if (geometry === void 0) { geometry = null; }
        /** @type {?} */
        var feature = toFeature(properties, (/** @type {?} */ (geometry)));
        return new EventFeature(feature, layer);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} routeId
     * @return {THIS}
     */
    EventFeature.prototype.setRouteId = /**
     * @template THIS
     * @this {THIS}
     * @param {?} routeId
     * @return {THIS}
     */
    function (routeId) {
        (/** @type {?} */ (this)).routeId = routeId;
        return (/** @type {?} */ (this)).setValue(routeId, (/** @type {?} */ (this)).layer.routeIdFieldName);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    EventFeature.prototype.setFromMeasure = /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    function (measure) {
        (/** @type {?} */ (this)).fromMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.fromMeasureFieldName, measure);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    EventFeature.prototype.setToMeasure = /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    function (measure) {
        if (!(/** @type {?} */ (this)).isLinear) {
            throw new Error('Cannot set to measure for a point-based feature');
        }
        (/** @type {?} */ (this)).toMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toMeasureFieldName, measure);
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @return {THIS}
     */
    EventFeature.prototype.setMeasureExtent = /**
     * @template THIS
     * @this {THIS}
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @return {THIS}
     */
    function (fromMeasure, toMeasure) {
        return (/** @type {?} */ (this)).setFromMeasure(fromMeasure)
            .setToMeasure(toMeasure);
    };
    /**
     * @return {?}
     */
    EventFeature.prototype.clone = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var copyFeature = clone(this.toFeature());
        return new EventFeature(copyFeature, this.layer);
    };
    /**
     * @protected
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    EventFeature.prototype.setGeometryTypeProperties = /**
     * @protected
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        (/** @type {?} */ (this)).isLinear = Boolean((/** @type {?} */ (this)).layer.toMeasureFieldName);
        (/** @type {?} */ (this)).isPoint = !(/** @type {?} */ (this)).isLinear;
        (/** @type {?} */ (this)).isPolygon = false;
        return (/** @type {?} */ (this));
    };
    return EventFeature;
}(LrsFeature));
export { EventFeature };
if (false) {
    /** @type {?} */
    EventFeature.prototype.layer;
}
