// @tss/sdk/esm2015/lib/core/storage/in-memory-storage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { BaseStorage } from './base-storage.service';
import * as i0 from "@angular/core";
export class InMemoryStorage extends BaseStorage {
    constructor() {
        super(...arguments);
        this.cache = new Map();
    }
    /**
     * @return {?}
     */
    get length() {
        return this.cache.size;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    get(key) {
        return this.cache.get(this.formatKey(key));
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    set(key, value) {
        this.cache.set(this.formatKey(key), value);
    }
    /**
     * @param {?} key
     * @return {?}
     */
    remove(key) {
        this.cache.delete(this.formatKey(key));
    }
    /**
     * @return {?}
     */
    clear() {
        this.cache.clear();
    }
}
InMemoryStorage.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */ InMemoryStorage.ngInjectableDef = i0.defineInjectable({ factory: function InMemoryStorage_Factory() { return new InMemoryStorage(); }, token: InMemoryStorage, providedIn: "root" });
if (false) {
    /** @type {?} */
    InMemoryStorage.prototype.cache;
}
