// @tss/sdk/esm2015/lib/queuing/queuable-work-queue-manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// 3rd Party.
import { forkJoin } from 'rxjs';
import { map } from 'rxjs/operators';
// TSS.
import { WorkQueueManager, DEFAULT_WORK_QUEUE_OPTIONS } from './work-queue-manager';
/**
 * @abstract
 * @template TInput, TOutput
 */
export class QueuableWorkQueueManager extends WorkQueueManager {
    constructor() {
        super(...arguments);
        this.options = Object.assign({
            autoDequeu: true
        }, DEFAULT_WORK_QUEUE_OPTIONS);
    }
    /**
     * Enqueues multiple items into an array.
     * @param {...?} items Items to enqueue.
     *
     * TODO: We want to change this to use "mergeAll" once we
     *  upgrade to rxjs 6, as that is the desired behavior here.
     *
     * // The following works in rxjs 6, but not in 5:
     * of(...sources)
     * .pipe(
     *   mergeAll(),
     *   // NOTE: in rxjs 5
     *   tap((x: TInput) => x)
     * )
     * @return {?}
     */
    enqueuMany(...items) {
        /** @type {?} */
        const sources = items
            .map((/**
         * @param {?} i
         * @return {?}
         */
        i => this.enqueue(i)));
        return forkJoin(sources)
            .pipe(map((/**
         * @param {?} results
         * @return {?}
         */
        (results) => results.pop())));
    }
    /**
     * Deque many items from the queue.
     * @param {...?} items Items do deque.
     * @return {?}
     */
    dequeuMany(...items) {
        /** @type {?} */
        const sources = items
            .map((/**
         * @param {?} i
         * @return {?}
         */
        i => this.dequeue(i)));
        return forkJoin(sources)
            .pipe(map((/**
         * @param {?} results
         * @return {?}
         */
        (results) => results.pop())));
    }
    /**
     * @protected
     * @param {?} payload
     * @return {?}
     */
    onWorkComplete(payload) {
        return forkJoin(super.onWorkComplete(payload), this.dequeue(payload.input))
            .pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        (result) => result[0])));
    }
}
if (false) {
    /** @type {?} */
    QueuableWorkQueueManager.prototype.options;
    /**
     * @abstract
     * @param {?} item
     * @return {?}
     */
    QueuableWorkQueueManager.prototype.enqueue = function (item) { };
    /**
     * @abstract
     * @param {?} item
     * @return {?}
     */
    QueuableWorkQueueManager.prototype.dequeue = function (item) { };
}
