// @tss/sdk/esm2015/lib/queuing/queue-worker-token.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Guid } from '../core/uuid/guid';
import { getElapsedSeconds } from '../core/date-time/get-elapsed-seconds.function';
import { calculateRemainingSeconds } from '../core/stats/calculate-remaining-seconds.function';
/**
 * Unique token created for each queue worker
 */
export class QueueWorkerToken {
    constructor() {
        this.isWorking = false;
        this.isCanceled = false;
        this.id = Guid.newGuid().value;
    }
    /**
     * @return {?}
     */
    get isComplete() {
        return !this.isWorking;
    }
    /**
     * @return {?}
     */
    get ellapsedSeconds() {
        return getElapsedSeconds(new Date(), this.startTime);
    }
    /**
     * @return {?}
     */
    get totalProcessed() {
        return this.totalComplete + this.totalFailed;
    }
    /**
     * @return {?}
     */
    get recordsProcessedPerSecond() {
        return this.totalComplete / this.ellapsedSeconds;
    }
    /**
     * @return {?}
     */
    get recordsProcessedPerMinute() {
        return this.recordsProcessedPerSecond * 60;
    }
    /**
     * @return {?}
     */
    start() {
        this.reset();
        this.startTime = new Date();
        this.isWorking = true;
        return this;
    }
    /**
     * @return {?}
     */
    reset() {
        this.isWorking = false;
        this.isCanceled = false;
        this.totalComplete = 0;
        this.totalFailed = 0;
        this.startTime = null;
        this.endTime = null;
        return this;
    }
    /**
     * @param {?} remainingCount
     * @return {?}
     */
    calculateRemainingSeconds(remainingCount) {
        return this.isComplete
            ? 0
            : calculateRemainingSeconds(this.totalProcessed, this.ellapsedSeconds, remainingCount);
    }
    /**
     * @return {?}
     */
    complete() {
        this.endTime = new Date();
        this.isWorking = false;
        return this;
    }
    /**
     * @return {?}
     */
    cancel() {
        this.isCanceled = true;
        this.isWorking = false;
        return this;
    }
}
if (false) {
    /** @type {?} */
    QueueWorkerToken.prototype.id;
    /** @type {?} */
    QueueWorkerToken.prototype.isWorking;
    /** @type {?} */
    QueueWorkerToken.prototype.isCanceled;
    /** @type {?} */
    QueueWorkerToken.prototype.totalComplete;
    /** @type {?} */
    QueueWorkerToken.prototype.totalFailed;
    /** @type {?} */
    QueueWorkerToken.prototype.startTime;
    /** @type {?} */
    QueueWorkerToken.prototype.endTime;
}
