// @tss/sdk/esm2015/lib/queuing/work-queue-status-tracker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { average } from '../core/stats/average.function';
export class WorkQueueStatusTracker {
    constructor() {
        this.totalQueueSize = 0;
        this.workers = [];
    }
    /**
     * @return {?}
     */
    get workerCount() {
        return this.workers.length;
    }
    /**
     * @return {?}
     */
    get percentComplete() {
        return this.totalQueueSize >= 0
            ? Math.round(this.totalProcessed / this.totalQueueSize * 100) || 0
            : 0;
    }
    /**
     * @return {?}
     */
    get totalComplete() {
        return this.workers.reduce((/**
         * @param {?} accumulator
         * @param {?} current
         * @return {?}
         */
        (accumulator, current) => accumulator + current.totalComplete), 0) || 0;
    }
    /**
     * @return {?}
     */
    get totalFailed() {
        return this.workers.reduce((/**
         * @param {?} accumulator
         * @param {?} current
         * @return {?}
         */
        (accumulator, current) => accumulator + current.totalFailed), 0) || 0;
    }
    /**
     * @return {?}
     */
    get totalProcessed() {
        return this.totalComplete + this.totalFailed;
    }
    /**
     * @return {?}
     */
    get isWorking() {
        return Boolean(this.runningWorkerCount);
    }
    /**
     * @return {?}
     */
    get isComplete() {
        return this.startTime && !this.isWorking;
    }
    /**
     * @return {?}
     */
    get isCanceled() {
        /** @type {?} */
        const canceledWorkersCount = this.workers.filter((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.isCanceled)).length;
        return Boolean(canceledWorkersCount && canceledWorkersCount === this.workers.length);
    }
    /**
     * @return {?}
     */
    get runningWorkerCount() {
        return this.workers.filter((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.isWorking)).length;
    }
    /**
     * @return {?}
     */
    get startTime() {
        /** @type {?} */
        const startTimes = this.workers
            .filter((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.startTime))
            .map((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.startTime
            ? worker.startTime.getTime()
            : null));
        return new Date(Math.min(...startTimes));
    }
    /**
     * @return {?}
     */
    get endTime() {
        /** @type {?} */
        const startTimes = this.workers
            .filter((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.startTime))
            .map((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.startTime
            ? worker.startTime.getTime()
            : null));
        return new Date(Math.max(...startTimes));
    }
    /**
     * @return {?}
     */
    get remainingQueueSize() {
        return this.totalQueueSize - this.totalProcessed;
    }
    /**
     * @return {?}
     */
    get estimatedSecondsLeft() {
        /** @type {?} */
        const remainingCount = this.remainingQueueSize;
        return remainingCount
            ? average(this.workers
                .map((/**
             * @param {?} worker
             * @return {?}
             */
            worker => worker.calculateRemainingSeconds(remainingCount))))
            : 0;
    }
    /**
     * @return {?}
     */
    get averageRecordsProcessedPerMinute() {
        return average(this.workers
            .map((/**
         * @param {?} worker
         * @return {?}
         */
        worker => worker.recordsProcessedPerMinute)));
    }
    /**
     * @param {?} totalQueueSize
     * @return {?}
     */
    setQueueSize(totalQueueSize) {
        this.totalQueueSize = totalQueueSize;
        return this;
    }
    /**
     * @return {?}
     */
    complete() {
        this.workers.splice(0, this.workers.length);
        this.totalQueueSize = 0;
        return this;
    }
    /**
     * @param {...?} workers
     * @return {?}
     */
    addWorker(...workers) {
        workers.forEach((/**
         * @param {?} worker
         * @return {?}
         */
        worker => {
            this.workers.push(worker);
        }));
        return this;
    }
    /**
     * @return {?}
     */
    getStatus() {
        return {
            isWorking: this.isWorking,
            isCanceled: this.isCanceled,
            isComplete: this.isComplete,
            totalProcessed: this.totalProcessed,
            startTime: this.startTime,
            endTime: this.endTime,
            percentComplete: this.percentComplete,
            totalQueueSize: this.totalQueueSize,
            remainingQueueSize: this.remainingQueueSize,
            totalComplete: this.totalComplete,
            totalFailed: this.totalFailed,
            estimatedSecondsLeft: this.estimatedSecondsLeft,
            workerCount: this.workers.length,
        };
    }
}
if (false) {
    /** @type {?} */
    WorkQueueStatusTracker.prototype.totalQueueSize;
    /**
     * @type {?}
     * @protected
     */
    WorkQueueStatusTracker.prototype.workers;
}
