// @tss/sdk/esm2015/lib/security/security-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ResourceConfig } from '../core/resource.config';
import { AutoConfig } from '../core/auto-config.decorator';
import * as i0 from "@angular/core";
/**
 * Security related configuration options. These settings are used to drive all
 * security related behaviour throughout the application. The settings can be
 * overridden directly in the code, application module, or from the TDS server.
 */
let SecurityConfig = /**
 * Security related configuration options. These settings are used to drive all
 * security related behaviour throughout the application. The settings can be
 * overridden directly in the code, application module, or from the TDS server.
 */
class SecurityConfig extends ResourceConfig {
    constructor() {
        super(...arguments);
        this.rolesEnabled = true;
        this.storeToken = true;
        this.autoSignIn = true;
        this.authorizationType = 'Bearer';
        this.tokenEndpoint = 'token';
        this.signOutEndpoint = 'api/account/signout';
        this.userInfoEndpoint = 'api/account/userinfo';
        this.usersEndpoint = 'api/users/{id}';
        this.rolesEndpoint = 'api/roles/{id}';
        this.privilegesEndpoint = 'api/privileges/{id}';
        this.userRolesEndpoint = 'api/users/{userId}/roles/{roleName}';
        this.rolesPrivilegesEndpoint = 'api/roles/{roleId}/privileges/{privilegeId}';
        this.tokenStorageKey = '__tss_token';
        this.tokenInvalidationDateLookup = '__tss_token_invalidate_date';
        this.tdsTokenInclusionExpression = '/api/';
        this.withCredentialsInclusionExpression = '(.*?)';
        this.setWithCredentials = false;
        this.accessToken = null;
        this.defaultUserName = null;
        this.defaultPassword = null;
        this.roleSplitterExp = '[,|]+';
        this.tokenBody = (/**
         * @param {?} userName
         * @param {?} password
         * @return {?}
         */
        (userName, password) => `grant_type=password&username=${userName}&password=${password}`);
    }
};
SecurityConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ SecurityConfig.ngInjectableDef = i0.defineInjectable({ factory: function SecurityConfig_Factory() { return new SecurityConfig(); }, token: SecurityConfig, providedIn: "root" });
/**
 * Security related configuration options. These settings are used to drive all
 * security related behaviour throughout the application. The settings can be
 * overridden directly in the code, application module, or from the TDS server.
 */
SecurityConfig = tslib_1.__decorate([
    AutoConfig('SecurityConfig')
], SecurityConfig);
export { SecurityConfig };
if (false) {
    /** @type {?} */
    SecurityConfig.prototype.rolesEnabled;
    /** @type {?} */
    SecurityConfig.prototype.storeToken;
    /** @type {?} */
    SecurityConfig.prototype.autoSignIn;
    /** @type {?} */
    SecurityConfig.prototype.authorizationType;
    /** @type {?} */
    SecurityConfig.prototype.tokenEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.signOutEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.userInfoEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.usersEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.rolesEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.privilegesEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.userRolesEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.rolesPrivilegesEndpoint;
    /** @type {?} */
    SecurityConfig.prototype.tokenStorageKey;
    /** @type {?} */
    SecurityConfig.prototype.tokenInvalidationDateLookup;
    /** @type {?} */
    SecurityConfig.prototype.tdsTokenInclusionExpression;
    /** @type {?} */
    SecurityConfig.prototype.withCredentialsInclusionExpression;
    /** @type {?} */
    SecurityConfig.prototype.setWithCredentials;
    /** @type {?} */
    SecurityConfig.prototype.accessToken;
    /** @type {?} */
    SecurityConfig.prototype.defaultUserName;
    /** @type {?} */
    SecurityConfig.prototype.defaultPassword;
    /** @type {?} */
    SecurityConfig.prototype.roleSplitterExp;
    /** @type {?} */
    SecurityConfig.prototype.tokenBody;
}
