// @tss/sdk/esm2015/lib/export/export.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { JSON_MIME_TYPE, TEXT_MIME_TYPE } from '../core/media/media.constants';
import { base64MimeType } from '../core/media/base64-mime-type.function';
import { base64toBlob } from '../core/media/base64-to-blob.function';
import { DEFAULT_EXPORT_GEOJSON_FILE_NAME } from './export.constants';
import { format } from '../core/formatting/format.function';
import { isString } from '../core/type-check/is-string.function';
import { formatDate } from '../core/date-time/format/format-date.function';
import * as i0 from "@angular/core";
export class ExportService {
    constructor() {
    }
    /**
     * @param {?} fileNameTemplate
     * @param {?=} tokens
     * @param {?=} removeEmptyTokens
     * @return {?}
     */
    replaceTokens(fileNameTemplate, tokens = {}, removeEmptyTokens) {
        return format(fileNameTemplate, Object.assign(this.getReplaceTokens(), tokens), removeEmptyTokens);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?=} fileNameTemplate
     * @param {...?} features
     * @return {THIS}
     */
    exportFeatureCollection(fileNameTemplate = DEFAULT_EXPORT_GEOJSON_FILE_NAME, ...features) {
        /** @type {?} */
        const fileName = (/** @type {?} */ (this)).replaceTokens(fileNameTemplate, {
            featureLength: features.length,
        });
        /** @type {?} */
        const featureCollection = {
            type: 'FeatureCollection',
            features
        };
        return (/** @type {?} */ (this)).exportFile((/** @type {?} */ (this)).stringify(featureCollection), fileName, JSON_MIME_TYPE);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?=} fileNameTemplate
     * @param {?=} feature
     * @return {THIS}
     */
    exportFeature(fileNameTemplate = DEFAULT_EXPORT_GEOJSON_FILE_NAME, feature) {
        return (/** @type {?} */ (this)).exportFile((/** @type {?} */ (this)).stringify(feature), fileNameTemplate, JSON_MIME_TYPE);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    stringify(data) {
        return JSON.stringify(data, (/**
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
        (key, value) => {
            return key.startsWith('$$') ? undefined : value;
        }), 2);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} data
     * @param {?} fileNameTemplate
     * @return {THIS}
     */
    exportJson(data, fileNameTemplate) {
        /** @type {?} */
        const jsonString = isString(data)
            ? (/** @type {?} */ (data))
            : (/** @type {?} */ (this)).stringify(data);
        return (/** @type {?} */ (this)).exportFile(jsonString, fileNameTemplate, JSON_MIME_TYPE);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} base64Data
     * @param {?} fileNameTemplate
     * @param {?=} contentType
     * @return {THIS}
     */
    exportBase64(base64Data, fileNameTemplate, contentType) {
        /** @type {?} */
        const fileName = (/** @type {?} */ (this)).replaceTokens(fileNameTemplate);
        contentType = contentType || base64MimeType(base64Data);
        /** @type {?} */
        const blob = base64toBlob(base64Data, contentType);
        return (/** @type {?} */ (this)).exportFromUrl(URL.createObjectURL(blob), fileName);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} data
     * @param {?} fileNameTemplate
     * @param {?=} contentType
     * @return {THIS}
     */
    exportFile(data, fileNameTemplate, contentType = TEXT_MIME_TYPE) {
        /** @type {?} */
        const blob = new Blob([data], { type: contentType });
        /** @type {?} */
        const fileName = (/** @type {?} */ (this)).replaceTokens(fileNameTemplate);
        return (/** @type {?} */ (this)).exportBlob(blob, fileName);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} blob
     * @param {?} fileNameTemplate
     * @return {THIS}
     */
    exportBlob(blob, fileNameTemplate) {
        /** @type {?} */
        const fileName = (/** @type {?} */ (this)).replaceTokens(fileNameTemplate);
        return (/** @type {?} */ (this)).exportFromUrl(window.URL.createObjectURL(blob), fileName);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} url
     * @param {?} fileName
     * @return {THIS}
     */
    exportFromUrl(url, fileName) {
        /** @type {?} */
        const link = document.createElement('a');
        link.href = url;
        link.download = fileName;
        link.click();
        return (/** @type {?} */ (this));
    }
    /**
     * @protected
     * @return {?}
     */
    getReplaceTokens() {
        return {
            date: new Date(),
            shortDate: formatDate(new Date(), 'yyyy-MM-dd'),
            year: new Date().getFullYear(),
        };
    }
}
ExportService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ExportService.ctorParameters = () => [];
/** @nocollapse */ ExportService.ngInjectableDef = i0.defineInjectable({ factory: function ExportService_Factory() { return new ExportService(); }, token: ExportService, providedIn: "root" });
