// @tss/sdk/esm5/lib/security/has-roles.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { RoleManagerService } from './role-manager.service';
import { Logger } from '../core/logging/logger.service';
var HasRolesDirective = /** @class */ (function () {
    function HasRolesDirective(templateRef, viewContainer, roleMgr, log) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.roleMgr = roleMgr;
        this.log = log;
    }
    Object.defineProperty(HasRolesDirective.prototype, "hasRoles", {
        set: /**
         * @param {?} roles
         * @return {?}
         */
        function (roles) {
            if (this.roleMgr.hasRoles(roles)) {
                // If condition is true add template to DOM
                this.viewContainer.createEmbeddedView(this.templateRef);
                this.log.trace("Granted access: " + roles);
            }
            else {
                // Else remove template from DOM
                this.viewContainer.clear();
                this.log.trace("Denied access: " + roles);
            }
        },
        enumerable: true,
        configurable: true
    });
    HasRolesDirective.decorators = [
        { type: Directive, args: [{
                    // TODO: remove this once this is moved to it's own package - RG.
                    // tslint:disable-next-line
                    selector: '[hasRoles]'
                },] }
    ];
    /** @nocollapse */
    HasRolesDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef },
        { type: RoleManagerService },
        { type: Logger }
    ]; };
    HasRolesDirective.propDecorators = {
        hasRoles: [{ type: Input }]
    };
    return HasRolesDirective;
}());
export { HasRolesDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HasRolesDirective.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    HasRolesDirective.prototype.viewContainer;
    /**
     * @type {?}
     * @private
     */
    HasRolesDirective.prototype.roleMgr;
    /**
     * @type {?}
     * @private
     */
    HasRolesDirective.prototype.log;
}
