// @tss/sdk/esm5/lib/security/security.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { TssCoreModule } from '../core/core.module';
import { TokenInterceptor } from './token.interceptor';
import { HasRolesDirective } from './has-roles.directive';
var TssSecurityModule = /** @class */ (function () {
    function TssSecurityModule() {
    }
    TssSecurityModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        FormsModule,
                        TssCoreModule
                    ],
                    declarations: [
                        HasRolesDirective,
                    ],
                    exports: [
                        HasRolesDirective,
                    ],
                    providers: [
                        {
                            provide: HTTP_INTERCEPTORS,
                            useClass: TokenInterceptor,
                            multi: true
                        }
                    ],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                },] }
    ];
    return TssSecurityModule;
}());
export { TssSecurityModule };
