// @tss/sdk/esm5/lib/security/token.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { SecurityConfig } from './security-config';
import * as i0 from "@angular/core";
import * as i1 from "./security-config";
var TokenInterceptor = /** @class */ (function () {
    function TokenInterceptor(config) {
        this.config = config;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    TokenInterceptor.prototype.intercept = /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    function (request, next) {
        /** @type {?} */
        var withCredsExpr = new RegExp(this.config.withCredentialsInclusionExpression);
        if (this.config.accessToken && new RegExp(this.config.tdsTokenInclusionExpression).test(request.url)) {
            request = request.clone({
                withCredentials: this.config.setWithCredentials && withCredsExpr.test(request.url),
                setHeaders: {
                    Authorization: this.config.authorizationType + " " + this.config.accessToken
                }
            });
        }
        else if (this.config.setWithCredentials && withCredsExpr.test(request.url)) {
            request = request.clone({
                withCredentials: true
            });
        }
        return next.handle(request);
    };
    TokenInterceptor.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TokenInterceptor.ctorParameters = function () { return [
        { type: SecurityConfig }
    ]; };
    /** @nocollapse */ TokenInterceptor.ngInjectableDef = i0.defineInjectable({ factory: function TokenInterceptor_Factory() { return new TokenInterceptor(i0.inject(i1.SecurityConfig)); }, token: TokenInterceptor, providedIn: "root" });
    return TokenInterceptor;
}());
export { TokenInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    TokenInterceptor.prototype.config;
}
