// @tss/sdk/esm5/lib/motion/device-motion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// 3rd Party.
import { Subject, fromEvent } from 'rxjs';
import { MotionConfig } from './motion.config';
import * as i0 from "@angular/core";
import * as i1 from "./motion.config";
/**
 * Geolocator service. Provides a single point of motion API.
 */
var DeviceMotionService = /** @class */ (function () {
    function DeviceMotionService(config) {
        var _this = this;
        this.config = config;
        this.motionChangeSource = new Subject();
        this.motionChange = this.motionChangeSource.asObservable();
        fromEvent(window, 'ondevicemotion').subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return _this.emitMotionChange(event); }));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} motion
     * @return {THIS}
     */
    DeviceMotionService.prototype.mockMotion = /**
     * @template THIS
     * @this {THIS}
     * @param {?} motion
     * @return {THIS}
     */
    function (motion) {
        if (!(/** @type {?} */ (this)).config.allowMotionMocking) {
            throw new Error('Permission denied to mock device motion');
        }
        (/** @type {?} */ (this)).motionChangeSource.next(motion);
        return (/** @type {?} */ (this));
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    DeviceMotionService.prototype.emitMotionChange = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.motionChangeSource.next({
            x: event.acceleration.x,
            y: event.acceleration.y,
            z: event.acceleration.z,
            withGravityX: event.accelerationIncludingGravity.x,
            withGravityY: event.accelerationIncludingGravity.y,
            withGravityZ: event.accelerationIncludingGravity.z,
            rotationRateAlpha: event.rotationRate.alpha,
            rotationRateBeta: event.rotationRate.beta,
            rotationRateGamma: event.rotationRate.gamma,
        });
    };
    DeviceMotionService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DeviceMotionService.ctorParameters = function () { return [
        { type: MotionConfig }
    ]; };
    /** @nocollapse */ DeviceMotionService.ngInjectableDef = i0.defineInjectable({ factory: function DeviceMotionService_Factory() { return new DeviceMotionService(i0.inject(i1.MotionConfig)); }, token: DeviceMotionService, providedIn: "root" });
    return DeviceMotionService;
}());
export { DeviceMotionService };
if (false) {
    /** @type {?} */
    DeviceMotionService.prototype.motionChange;
    /** @type {?} */
    DeviceMotionService.prototype.motionChangeSource;
    /**
     * @type {?}
     * @private
     */
    DeviceMotionService.prototype.config;
}
