// @tss/sdk/esm2015/lib/core/sort/sort.function.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SortDirection } from './sort-direction';
import { isFunction } from '../type-check/is-function.function';
import { isString } from '../type-check/is-string.function';
/**
 * @template T
 * @param {?} list
 * @param {...?} sortProperties
 * @return {?}
 */
export function sortAscending(list, ...sortProperties) {
    return sort(list, SortDirection.Ascending, ...sortProperties);
}
/**
 * @template T
 * @param {?} list
 * @param {...?} sortProperties
 * @return {?}
 */
export function sortDescending(list, ...sortProperties) {
    return sort(list, SortDirection.Descending, ...sortProperties);
}
/**
 * @template T
 * @param {?} list
 * @param {?=} direction
 * @param {...?} sortProperties
 * @return {?}
 */
export function sort(list, direction = SortDirection.Ascending, ...sortProperties) {
    list.sort((/**
     * @param {?} item1
     * @param {?} item2
     * @return {?}
     */
    (item1, item2) => {
        /** @type {?} */
        let sortResponse = 0;
        if (sortProperties && sortProperties.length) {
            sortProperties.forEach((/**
             * @param {?} property
             * @return {?}
             */
            property => {
                if (sortResponse === 0) {
                    /** @type {?} */
                    const value1 = getPropertyValue(item1, property);
                    /** @type {?} */
                    const value2 = getPropertyValue(item2, property);
                    sortResponse = getSortResultForString(value1, value2, direction);
                }
            }));
        }
        else {
            sortResponse = getSortResultForString(item1, item2, direction);
        }
        return sortResponse;
    }));
    return list;
}
/**
 * @template T
 * @param {?} item
 * @param {?} property
 * @return {?}
 */
function getPropertyValue(item, property) {
    return isFunction(property)
        ? ((/** @type {?} */ (property)))(item)
        : item[(/** @type {?} */ (property))];
}
/**
 * @param {?} value1
 * @param {?} value2
 * @param {?} direction
 * @return {?}
 */
function getSortResultForString(value1, value2, direction) {
    return direction === SortDirection.Ascending
        ? getSortResultForStringAscending(value1, value2)
        : getSortResultForStringDescending(value1, value2);
}
/**
 * @param {?} value
 * @return {?}
 */
function ensureValue(value) {
    return value === null || value === undefined ? '' : value;
}
/**
 * @param {?} value1
 * @param {?} value2
 * @return {?}
 */
function getSortResultForStringAscending(value1, value2) {
    value1 = ensureValue(value1);
    value2 = ensureValue(value2);
    /** @type {?} */
    const compare1 = isString(value1) ? value1.toUpperCase() : value1;
    /** @type {?} */
    const compare2 = isString(value2) ? value2.toUpperCase() : value2;
    /** @type {?} */
    let compareResult = 0;
    if (compare1 < compare2) {
        compareResult = -1;
    }
    else if (compare1 > compare2) {
        compareResult = 1;
    }
    // NOTE: 0 = equals, 1 = greater, -1 less than.
    return compareResult;
}
/**
 * @param {?} value1
 * @param {?} value2
 * @return {?}
 */
function getSortResultForStringDescending(value1, value2) {
    /** @type {?} */
    const compare1 = isString(value1) ? value1.toUpperCase() : value1;
    /** @type {?} */
    const compare2 = isString(value2) ? value2.toUpperCase() : value2;
    /** @type {?} */
    let compareResult = 0;
    if (compare1 > compare2) {
        compareResult = -1;
    }
    else if (compare1 < compare2) {
        compareResult = 1;
    }
    // NOTE: 0 = equals, 1 = greater, -1 less than.
    return compareResult;
}
