// @tss/sdk/esm2015/lib/lrs/redlines/redline-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TSS.
import { LrsFeature } from '../lrs-feature';
import { toFeature } from '../feature';
import { isPolygon } from '../../geometry/geometry/is-polygon.function';
import { clone } from '../../core/clone/clone.function';
export class RedlineFeature extends LrsFeature {
    /**
     * @param {?} feature
     * @param {?} layer
     */
    constructor(feature, layer) {
        super(feature, layer);
        this.routeId = this.getValue(layer.routeIdFieldName);
        this.fromMeasure = this.getValue(layer.fromMeasureFieldName);
        this.toMeasure = this.getValue(layer.toMeasureFieldName);
        this.isLinear = Boolean(layer.toMeasureFieldName);
    }
    /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    static create(layer, properties = {}, geometry = null) {
        /** @type {?} */
        const feature = toFeature(properties, (/** @type {?} */ (geometry)));
        return new RedlineFeature(feature, layer);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    setFromMeasure(measure) {
        (/** @type {?} */ (this)).fromMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.fromMeasureFieldName, measure);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    setToMeasure(measure) {
        if (!(/** @type {?} */ (this)).isLinear) {
            throw new Error('Cannot set to measure for a point-based feature');
        }
        (/** @type {?} */ (this)).toMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toMeasureFieldName, measure);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @return {THIS}
     */
    setMeasureExtent(fromMeasure, toMeasure) {
        return (/** @type {?} */ (this)).setFromMeasure(fromMeasure)
            .setToMeasure(toMeasure);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const copyFeature = clone(this.toFeature());
        return new RedlineFeature(copyFeature, this.layer);
    }
    /**
     * @protected
     * @return {?}
     */
    isGeometryTypeValid() {
        return isPolygon(this.geometry) || super.isGeometryTypeValid();
    }
}
if (false) {
    /** @type {?} */
    RedlineFeature.prototype.layer;
}
