// @tss/sdk/esm5/lib/core/date-time/date-time.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var STANDARDIZED_QUERY_DATE_FORMAT = "timestamp '{date}'";
/** @type {?} */
export var ORACLE_DATE_FORMAT = "TO_DATE('{date}','YYYY-MM-DD HH24:MI:SS')";
/** @type {?} */
export var DEFAULT_DATE_FORMAT = 'yyyy-MM-dd HH:mm:ss';
/** @type {?} */
export var HOURS_PER_DAY = 24;
/** @type {?} */
export var MINUTES_PER_HOUR = 60;
/** @type {?} */
export var SECONDS_PER_MINUTE = 60;
/** @type {?} */
export var MILLISECONDS_PER_SECOND = 1000;
/** @type {?} */
export var MILLISECONDS_PER_MINUTE = 60000;
/** @type {?} */
export var MILLISECONDS_PER_HOUR = 3600000;
/** @type {?} */
export var MILLISECONDS_PER_DAY = 86400000;
/** @type {?} */
export var READABLE_TIME_SUFFIX_MAP = {
    ms: 'ms',
    msDecimalPlaces: 0,
    sec: 'sec',
    secDecimalPlaces: 0,
    min: 'min',
    minDecimalPlaces: 2,
    hr: 'hr',
    hrDecimalPlaces: 2,
    days: 'days',
    daysDecimalPlaces: 2,
};
