// @tss/sdk/esm5/lib/core/validation/validation-check.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ValidationResult } from './validation-results';
/** @type {?} */
var BASE_ID = 1;
/**
 * @template T
 */
var /**
 * @template T
 */
ValidationCheck = /** @class */ (function () {
    function ValidationCheck(description, passCheck) {
        this.subChecks = [];
        this.description = description;
        this.passCheck = passCheck;
        this.id = BASE_ID++;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    ValidationCheck.prototype.hasId = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.id = id;
        return this;
    };
    /**
     * @param {?} description
     * @return {?}
     */
    ValidationCheck.prototype.message = /**
     * @param {?} description
     * @return {?}
     */
    function (description) {
        this.description = description;
        return this;
    };
    /**
     * @param {?=} passCheck
     * @return {?}
     */
    ValidationCheck.prototype.pass = /**
     * @param {?=} passCheck
     * @return {?}
     */
    function (passCheck) {
        this.passCheck = passCheck;
        return this;
    };
    /**
     * @param {?=} failCheck
     * @return {?}
     */
    ValidationCheck.prototype.fail = /**
     * @param {?=} failCheck
     * @return {?}
     */
    function (failCheck) {
        this.failCheck = failCheck;
        return this;
    };
    /**
     * @param {?=} warnCheck
     * @return {?}
     */
    ValidationCheck.prototype.warn = /**
     * @param {?=} warnCheck
     * @return {?}
     */
    function (warnCheck) {
        this.warnCheck = warnCheck;
        return this;
    };
    /**
     * @param {?=} onlyIfCheck
     * @return {?}
     */
    ValidationCheck.prototype.onlyIf = /**
     * @param {?=} onlyIfCheck
     * @return {?}
     */
    function (onlyIfCheck) {
        this.onlyIfCheck = onlyIfCheck;
        return this;
    };
    /**
     * @param {...?} innerChecks
     * @return {?}
     */
    ValidationCheck.prototype.add = /**
     * @param {...?} innerChecks
     * @return {?}
     */
    function () {
        var _this = this;
        var innerChecks = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            innerChecks[_i] = arguments[_i];
        }
        innerChecks.forEach((/**
         * @param {?} inner
         * @return {?}
         */
        function (inner) { return _this.subChecks.push(inner); }));
        return this;
    };
    /**
     * @param {?} target
     * @return {?}
     */
    ValidationCheck.prototype.check = /**
     * @param {?} target
     * @return {?}
     */
    function (target) {
        /** @type {?} */
        var result = new ValidationResult();
        if (this.onlyIfCheck && !this.onlyIfCheck(target)) {
            result.skipped = true;
            return result;
        }
        if (this.passCheck) {
            result.passed = this.passCheck(target);
        }
        if (this.failCheck) {
            result.passed = !this.failCheck(target);
        }
        if (this.warnCheck) {
            result.warned = !this.warnCheck(target);
        }
        if (this.subChecks) {
            this.subChecks.forEach((/**
             * @param {?} c
             * @return {?}
             */
            function (c) {
                /** @type {?} */
                var innerResult = c.check(target);
                result.add(innerResult);
            }));
        }
        return result;
    };
    return ValidationCheck;
}());
/**
 * @template T
 */
export { ValidationCheck };
if (false) {
    /** @type {?} */
    ValidationCheck.prototype.id;
    /** @type {?} */
    ValidationCheck.prototype.description;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.subChecks;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.passCheck;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.failCheck;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.warnCheck;
    /**
     * @type {?}
     * @private
     */
    ValidationCheck.prototype.onlyIfCheck;
}
