// @tss/sdk/esm5/lib/core/validation/validation-results.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ValidationResult = /** @class */ (function () {
    function ValidationResult() {
        this.violaters = [];
    }
    Object.defineProperty(ValidationResult.prototype, "failed", {
        get: /**
         * @return {?}
         */
        function () {
            return this.ran && !this.passed;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ValidationResult.prototype, "skipped", {
        get: /**
         * @return {?}
         */
        function () {
            return this.validationSkipped;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.validationSkipped = value;
            this.ran = true;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} inner
     * @return {?}
     */
    ValidationResult.prototype.add = /**
     * @param {?} inner
     * @return {?}
     */
    function (inner) {
        if (inner.failed) {
            this.passed = false;
        }
        if (inner.warned) {
            this.warned = true;
        }
        return this;
    };
    return ValidationResult;
}());
export { ValidationResult };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidationResult.prototype.validationSkipped;
    /** @type {?} */
    ValidationResult.prototype.passed;
    /** @type {?} */
    ValidationResult.prototype.warned;
    /** @type {?} */
    ValidationResult.prototype.ran;
    /** @type {?} */
    ValidationResult.prototype.violaters;
}
