// @tss/sdk/esm5/lib/esri/query/esri-query.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { WGS_84_SRID } from '../../geometry/geometry.constants';
/** @type {?} */
export var DEFAULT_MAP_QUERY_OPTIONS = {
    // mapServiceUrl: '',
    // layerId: '',
    where: '1=1',
    text: '',
    objectIds: '',
    time: '',
    geometry: '',
    geometryType: 'esriGeometryEnvelope',
    spatialRel: 'esriSpatialRelIntersects',
    relationParam: '',
    outFields: '*',
    returnGeometry: true,
    returnTrueCurves: false,
    maxAllowableOffset: '',
    geometryPrecision: '',
    inSR: WGS_84_SRID,
    outSR: WGS_84_SRID,
    returnIdsOnly: false,
    returnCountOnly: false,
    orderByFields: '',
    groupByFieldsForStatistics: '',
    outStatistics: '',
    returnZ: false,
    returnM: true,
    gdbVersion: '',
    returnDistinctValues: false,
    resultOffset: '',
    resultRecordCount: '',
    queryByDistance: '',
    returnExtentsOnly: false,
    datumTransformation: '',
    parameterValues: '',
    rangeValues: '',
    f: 'geojson'
};
