// @tss/sdk/esm2015/lib/motion/device-motion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// 3rd Party.
import { Subject, fromEvent } from 'rxjs';
import { MotionConfig } from './motion.config';
import * as i0 from "@angular/core";
import * as i1 from "./motion.config";
/**
 * Geolocator service. Provides a single point of motion API.
 */
export class DeviceMotionService {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.motionChangeSource = new Subject();
        this.motionChange = this.motionChangeSource.asObservable();
        fromEvent(window, 'ondevicemotion').subscribe((/**
         * @param {?} event
         * @return {?}
         */
        (event) => this.emitMotionChange(event)));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} motion
     * @return {THIS}
     */
    mockMotion(motion) {
        if (!(/** @type {?} */ (this)).config.allowMotionMocking) {
            throw new Error('Permission denied to mock device motion');
        }
        (/** @type {?} */ (this)).motionChangeSource.next(motion);
        return (/** @type {?} */ (this));
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    emitMotionChange(event) {
        this.motionChangeSource.next({
            x: event.acceleration.x,
            y: event.acceleration.y,
            z: event.acceleration.z,
            withGravityX: event.accelerationIncludingGravity.x,
            withGravityY: event.accelerationIncludingGravity.y,
            withGravityZ: event.accelerationIncludingGravity.z,
            rotationRateAlpha: event.rotationRate.alpha,
            rotationRateBeta: event.rotationRate.beta,
            rotationRateGamma: event.rotationRate.gamma,
        });
    }
}
DeviceMotionService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DeviceMotionService.ctorParameters = () => [
    { type: MotionConfig }
];
/** @nocollapse */ DeviceMotionService.ngInjectableDef = i0.defineInjectable({ factory: function DeviceMotionService_Factory() { return new DeviceMotionService(i0.inject(i1.MotionConfig)); }, token: DeviceMotionService, providedIn: "root" });
if (false) {
    /** @type {?} */
    DeviceMotionService.prototype.motionChange;
    /** @type {?} */
    DeviceMotionService.prototype.motionChangeSource;
    /**
     * @type {?}
     * @private
     */
    DeviceMotionService.prototype.config;
}
