// @tss/sdk/esm2015/lib/gps-gate/gps-gate.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { AutoConfig } from '../core/auto-config.decorator';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing Gps Gate-related settings.
 */
let GpsGateConfig = /**
 * Configuration class for storing Gps Gate-related settings.
 */
class GpsGateConfig {
    constructor() {
        this.refreshInterval = 1000;
        this.urlTemplate = 'http://localhost:12175/gps/{endpoint}?noCache={cacheBuster}';
        this.getInfoEndpoint = 'getGpsInfo';
        this.getVersionEndpoint = 'getVersion';
        this.jsonpParameterName = 'jsonp';
        this.geolocatorName = 'GpsGate Client';
        this.geolocatorDescription = 'GpsGate client that can connect to a variety of GPS data sources.';
    }
};
GpsGateConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ GpsGateConfig.ngInjectableDef = i0.defineInjectable({ factory: function GpsGateConfig_Factory() { return new GpsGateConfig(); }, token: GpsGateConfig, providedIn: "root" });
/**
 * Configuration class for storing Gps Gate-related settings.
 */
GpsGateConfig = tslib_1.__decorate([
    AutoConfig('GpsGateConfig')
], GpsGateConfig);
export { GpsGateConfig };
if (false) {
    /** @type {?} */
    GpsGateConfig.prototype.refreshInterval;
    /** @type {?} */
    GpsGateConfig.prototype.urlTemplate;
    /** @type {?} */
    GpsGateConfig.prototype.getInfoEndpoint;
    /** @type {?} */
    GpsGateConfig.prototype.getVersionEndpoint;
    /** @type {?} */
    GpsGateConfig.prototype.jsonpParameterName;
    /** @type {?} */
    GpsGateConfig.prototype.geolocatorName;
    /** @type {?} */
    GpsGateConfig.prototype.geolocatorDescription;
}
