// @tss/sdk/esm5/lib/security/users/user.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { SecurityConfig } from '../security-config';
import { join } from '../../core/url/join.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../security-config";
/**
 * \@example
 * ## Get Users that belong to a given role (by ID).
 * service.usersByRoleId('1428a342-8a95-4b6c-b3e9-da135e3ced04')
 *
 * // ...which is just a shortcut for:
 * service.query({}, new HttpParams().set('filterByRoleId', roleName));
 *
 * ## Get Users that belong to a given role (by name).
 * service.usersByRoleName('Admin')
 *
 * // ...which is just a shortcut for:
 * service.query({}, new HttpParams().set('filterByRoleName', roleName));
 */
var UserService = /** @class */ (function (_super) {
    tslib_1.__extends(UserService, _super);
    function UserService(http, config) {
        var _this = _super.call(this, http, {
            url: join(config.serviceUrl, config.usersEndpoint)
        }) || this;
        _this.http = http;
        return _this;
    }
    /**
     * @param {?} roleId
     * @param {?=} params
     * @return {?}
     */
    UserService.prototype.usersByRoleId = /**
     * @param {?} roleId
     * @param {?=} params
     * @return {?}
     */
    function (roleId, params) {
        params = (params || new HttpParams()).set('filterByRoleId', roleId);
        return this.query({}, params);
    };
    /**
     * @param {?} roleName
     * @param {?=} params
     * @return {?}
     */
    UserService.prototype.usersByRoleName = /**
     * @param {?} roleName
     * @param {?=} params
     * @return {?}
     */
    function (roleName, params) {
        params = (params || new HttpParams()).set('filterByRoleName', roleName);
        return this.query({}, params);
    };
    /**
     * @param {?} searchText
     * @param {?=} params
     * @return {?}
     */
    UserService.prototype.search = /**
     * @param {?} searchText
     * @param {?=} params
     * @return {?}
     */
    function (searchText, params) {
        params = (params || new HttpParams()).set('search', searchText);
        return this.query({}, params);
    };
    UserService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    UserService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: SecurityConfig }
    ]; };
    /** @nocollapse */ UserService.ngInjectableDef = i0.defineInjectable({ factory: function UserService_Factory() { return new UserService(i0.inject(i1.HttpClient), i0.inject(i2.SecurityConfig)); }, token: UserService, providedIn: "root" });
    return UserService;
}(BaseResourceService));
export { UserService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    UserService.prototype.http;
}
