// @tss/sdk/esm2015/lib/lrs/data-source-lrs-metadata.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { LrsConfig } from './lrs.config';
import { join } from '../core/url/join.function';
import { BaseResourceService } from '../core/resource/base-resource.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./lrs.config";
export class DataSourceLrsMetadataService extends BaseResourceService {
    /**
     * @param {?} http
     * @param {?} config
     */
    constructor(http, config) {
        super(http, {
            url: join(config.serviceUrl, config.dataSourceLrsMetadataEndpoint),
            idFieldName: 'dataSourceId'
        });
        this.http = http;
    }
    /**
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    create(entity, routeTokens, params) {
        routeTokens = routeTokens || {};
        routeTokens.id = entity.dataSourceId;
        return super.create(entity, routeTokens, params);
    }
}
DataSourceLrsMetadataService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DataSourceLrsMetadataService.ctorParameters = () => [
    { type: HttpClient },
    { type: LrsConfig }
];
/** @nocollapse */ DataSourceLrsMetadataService.ngInjectableDef = i0.defineInjectable({ factory: function DataSourceLrsMetadataService_Factory() { return new DataSourceLrsMetadataService(i0.inject(i1.HttpClient), i0.inject(i2.LrsConfig)); }, token: DataSourceLrsMetadataService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    DataSourceLrsMetadataService.prototype.http;
}
