// @tss/sdk/esm2015/lib/lrs/lrs-metadata.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { IntersectionLayer } from './intersections/intersection-layer';
import { EventLayer } from './events/event-layer';
import { NetworkLayer } from './networks/network-layer';
import { RedlineLayer } from './redlines/redline-layer';
import { CenterlineLayer } from './centerlines/centerline-layer';
import { CalibrationPointLayer } from './calibration-points/calibration-point-layer';
import { MapServerMetadata } from '../maps/map-server-metadata';
import { getGeometryField, getIdField } from '../maps/fields/layer-field';
import { findLayer } from './layer-search';
import { mergeMapMetadata, mergeLrsMetadata, deepMergeMapMetadata } from './lrs-metadata-merge';
import { LRS_METADATA_VALIDATIONS } from './lrs-metadata-validations';
import { NonLrsLayer } from './non-lrs/non-lrs-layer';
import { isNotValue } from '../core/type-check/is-not-value.function';
import { clone } from '../core/clone/clone.function';
/**
 * @param {?} layerId
 * @param {...?} layerList
 * @return {?}
 */
function find(layerId, ...layerList) {
    return findLayer(layerId, ...layerList);
}
export class LrsMetadata extends MapServerMetadata {
    /**
     * @param {?=} dataSource
     */
    constructor(dataSource) {
        super();
        this.networkLayers = [];
        this.eventLayers = [];
        this.redlineLayers = [];
        this.centerlineLayers = [];
        this.calibrationPointLayers = [];
        this.intersectionLayers = [];
        this.nonLRSLayers = [];
        this.lrs = [];
        this.domains = [];
        if (dataSource) {
            this.dataSource = dataSource;
            this.dataSourceId = dataSource.id;
        }
    }
    /**
     * Gets a layer by ID - searching across all types (event layers, network layers, etc).
     * @param {?} layerId
     * @return {?}
     */
    getLayer(layerId) {
        return find(layerId, this.getAllLayers());
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    getLrsLayer(layerId) {
        return (/** @type {?} */ ((find(layerId, this.networkLayers, this.eventLayers, this.redlineLayers))));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    getLayerDetail(layerId) {
        // TODO: Come up with a better way to check these layer types.
        // NOTE: We are doing all of these boolean checks instead of just simply calling
        // isEventLayer = this.isEventLayer(layerId) because of performance reasons.
        /** @type {?} */
        const isLrsLayer = this.isLrsLayer(layerId);
        /** @type {?} */
        const isEventLayer = this.isEventLayer(layerId);
        /** @type {?} */
        const isNetworkLayer = !isEventLayer && this.isNetworkLayer(layerId);
        /** @type {?} */
        const isCalibrationPointLayer = !isEventLayer && !isNetworkLayer && this.isCalibrationPointLayer(layerId);
        /** @type {?} */
        const isCenterlineLayer = !isEventLayer && !isNetworkLayer && !isCalibrationPointLayer && this.isCenterlineLayer(layerId);
        /** @type {?} */
        const isIntersectionLayer = !isEventLayer && !isNetworkLayer && !isCalibrationPointLayer && !isCenterlineLayer
            && this.isIntersectionLayer(layerId);
        /** @type {?} */
        const isRedlineLayer = !isEventLayer && !isNetworkLayer && !isCalibrationPointLayer && !isCenterlineLayer && !isIntersectionLayer
            && this.isRedlineLayer(layerId);
        /** @type {?} */
        const isTable = !isEventLayer && !isNetworkLayer && !isCalibrationPointLayer && !isCenterlineLayer && !isIntersectionLayer
            && !isRedlineLayer && this.isTable(layerId);
        /** @type {?} */
        const isNonLRSLayer = !isEventLayer && !isNetworkLayer && !isCalibrationPointLayer && !isCenterlineLayer && !isIntersectionLayer
            && !isRedlineLayer && !isTable && this.isNonLRSLayer(layerId);
        /** @type {?} */
        const layer = this.getLayer(layerId);
        /** @type {?} */
        const shapeField = layer ? getGeometryField(layer.fields) : null;
        /** @type {?} */
        const idField = layer ? getIdField(layer.fields) : null;
        return {
            layer,
            layerId,
            isEventLayer,
            isNetworkLayer,
            isCalibrationPointLayer,
            isCenterlineLayer,
            isIntersectionLayer,
            isRedlineLayer,
            isTable,
            isNonLRSLayer,
            isLrsLayer,
            shapeField,
            idField,
        };
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isLrsLayer(layerId) {
        return Boolean(this.getLrsLayer(layerId));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isNetworkLayer(layerId) {
        return Boolean(find(layerId, this.networkLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isEventLayer(layerId) {
        return Boolean(find(layerId, this.eventLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isIntersectionLayer(layerId) {
        return Boolean(find(layerId, this.intersectionLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isRedlineLayer(layerId) {
        return Boolean(find(layerId, this.redlineLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isCenterlineLayer(layerId) {
        return Boolean(find(layerId, this.centerlineLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isCalibrationPointLayer(layerId) {
        return Boolean(find(layerId, this.calibrationPointLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isNonLRSLayer(layerId) {
        return Boolean(find(layerId, this.nonLRSLayers));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    isTable(layerId) {
        return Boolean(find(layerId, this.tables));
    }
    /**
     * Merges map server general information into this LRS metadata instance.
     * @param {?} mapMetadata
     * @return {?}
     */
    mergeMapMetadata(mapMetadata) {
        mergeMapMetadata(this, mapMetadata);
        return this;
    }
    /**
     * @param {?} mapMetadata
     * @return {?}
     */
    deepMergeMapMetadata(mapMetadata) {
        deepMergeMapMetadata(this, mapMetadata);
        return this;
    }
    /**
     * Merges other LRS metadata into this LRS metdata instance.
     * @param {?} lrsMetadata
     * @return {?}
     */
    mergeLrsMetadata(lrsMetadata) {
        mergeLrsMetadata(this, lrsMetadata);
        return this;
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    getNetworkLayer(layerId) {
        return (/** @type {?} */ (find(layerId, this.networkLayers)));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    getEventLayer(layerId) {
        return (/** @type {?} */ (find(layerId, this.eventLayers)));
    }
    /**
     * @param {?} layerId
     * @return {?}
     */
    getIntersectionLayer(layerId) {
        return (/** @type {?} */ (find(layerId, this.intersectionLayers)));
    }
    /**
     * @return {?}
     */
    getLrsLayers() {
        return ((/** @type {?} */ (this.eventLayers)))
            .concat(this.networkLayers)
            .concat(this.redlineLayers);
    }
    /**
     * @return {?}
     */
    getFeatureClassLayers() {
        return ((/** @type {?} */ (this.eventLayers)))
            .concat(this.networkLayers)
            .concat(this.redlineLayers)
            .concat(this.calibrationPointLayers)
            .concat(this.centerlineLayers)
            .concat(this.nonLRSLayers);
    }
    /**
     * @return {?}
     */
    getAllLayers() {
        return ((/** @type {?} */ (this.getFeatureClassLayers())))
            .concat(this.tables)
            .concat(this.layers);
    }
    /**
     * @param {?} layer
     * @return {?}
     */
    getNetworkIdForLayer(layer) {
        /** @type {?} */
        let networkId = null;
        if (this.isEventLayer(layer.id)) {
            /** @type {?} */
            const eventLayer = (/** @type {?} */ (layer));
            networkId = eventLayer.parentNetwork ? eventLayer.parentNetwork.id : null;
        }
        if (isNotValue(networkId)) {
            this.networkLayers.forEach((/**
             * @param {?} network
             * @return {?}
             */
            network => {
                if (network.eventLayers.find((/**
                 * @param {?} l
                 * @return {?}
                 */
                l => l.id === layer.id))) {
                    networkId = network.id;
                }
            }));
        }
        return networkId;
    }
    /**
     * @param {?} networkLayerId
     * @return {?}
     */
    getEventLayersByNetwork(networkLayerId) {
        /** @type {?} */
        const networkLayer = this.getNetworkLayer(networkLayerId);
        return this.eventLayers.filter((/**
         * @param {?} layer
         * @return {?}
         */
        layer => layer.parentNetwork
            && layer.parentNetwork.id === networkLayerId
            || networkLayer.eventLayers
                && networkLayer.eventLayers.find((/**
                 * @param {?} l
                 * @return {?}
                 */
                l => l.id === layer.id))));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    rebuildLayers() {
        (/** @type {?} */ (this)).eventLayers = (/** @type {?} */ (this)).eventLayers.map((/**
         * @param {?} eventLayer
         * @return {?}
         */
        eventLayer => new EventLayer(find(eventLayer.id, (/** @type {?} */ (this)).layers), eventLayer)));
        (/** @type {?} */ (this)).networkLayers = (/** @type {?} */ (this)).networkLayers.map((/**
         * @param {?} networkLayer
         * @return {?}
         */
        networkLayer => new NetworkLayer(find(networkLayer.id, (/** @type {?} */ (this)).layers), networkLayer)));
        (/** @type {?} */ (this)).redlineLayers = (/** @type {?} */ (this)).redlineLayers.map((/**
         * @param {?} redlineLayer
         * @return {?}
         */
        redlineLayer => new RedlineLayer(find(redlineLayer.id, (/** @type {?} */ (this)).layers), redlineLayer)));
        (/** @type {?} */ (this)).centerlineLayers = (/** @type {?} */ (this)).centerlineLayers.map((/**
         * @param {?} centerlineLayer
         * @return {?}
         */
        centerlineLayer => new CenterlineLayer(centerlineLayer)));
        (/** @type {?} */ (this)).calibrationPointLayers = (/** @type {?} */ (this)).calibrationPointLayers.map((/**
         * @param {?} calibrationPointLayer
         * @return {?}
         */
        calibrationPointLayer => new CalibrationPointLayer(calibrationPointLayer)));
        (/** @type {?} */ (this)).intersectionLayers = (/** @type {?} */ (this)).intersectionLayers.map((/**
         * @param {?} intersectionLayer
         * @return {?}
         */
        intersectionLayer => new IntersectionLayer(intersectionLayer)));
        (/** @type {?} */ (this)).nonLRSLayers = (/** @type {?} */ (this)).nonLRSLayers.map((/**
         * @param {?} nonLrsLayer
         * @return {?}
         */
        nonLrsLayer => new NonLrsLayer(find(nonLrsLayer.id, (/** @type {?} */ (this)).layers), nonLrsLayer)));
        return (/** @type {?} */ (this)).assignCodedValueDomainsToFields();
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    assignCodedValueDomainsToFields() {
        if ((/** @type {?} */ (this)).domains.length) {
            (/** @type {?} */ (this)).getFeatureClassLayers()
                .forEach((/**
             * @param {?} layer
             * @return {?}
             */
            layer => layer.applyFieldDomains((/** @type {?} */ (this)).domains)));
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} source
     * @return {THIS}
     */
    assign(source) {
        Object.assign((/** @type {?} */ (this)), source);
        return (/** @type {?} */ (this));
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const copy = clone(this);
        return new LrsMetadata()
            .assign(copy)
            .rebuildLayers();
    }
    /**
     * @return {?}
     */
    validate() {
        return LRS_METADATA_VALIDATIONS
            .check(this);
    }
}
if (false) {
    /** @type {?} */
    LrsMetadata.prototype.dataSource;
    /** @type {?} */
    LrsMetadata.prototype.dataSourceId;
    /** @type {?} */
    LrsMetadata.prototype.name;
    /** @type {?} */
    LrsMetadata.prototype.serviceHasLrsMeta;
    /** @type {?} */
    LrsMetadata.prototype.networkLayers;
    /** @type {?} */
    LrsMetadata.prototype.eventLayers;
    /** @type {?} */
    LrsMetadata.prototype.redlineLayers;
    /** @type {?} */
    LrsMetadata.prototype.centerlineLayers;
    /** @type {?} */
    LrsMetadata.prototype.calibrationPointLayers;
    /** @type {?} */
    LrsMetadata.prototype.intersectionLayers;
    /** @type {?} */
    LrsMetadata.prototype.nonLRSLayers;
    /** @type {?} */
    LrsMetadata.prototype.lrs;
    /** @type {?} */
    LrsMetadata.prototype.domains;
}
