// @tss/sdk/esm2015/lib/lrs/lrs-feature-collection.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class LrsFeatureCollection {
    /**
     * @param {?} collection
     */
    constructor(collection) {
        this.bbox = collection.bbox;
        this.type = collection.type;
    }
    /**
     * @param {?} features
     * @param {?=} bbox
     * @return {?}
     */
    static create(features, bbox) {
        return new LrsFeatureCollection({
            type: 'FeatureCollection',
            features,
            bbox
        });
    }
    /**
     * @return {?}
     */
    toGeoJson() {
        return {
            type: 'FeatureCollection',
            bbox: this.bbox,
            features: this.features.map((/**
             * @param {?} feature
             * @return {?}
             */
            feature => feature.toGeoJson()))
        };
    }
}
if (false) {
    /** @type {?} */
    LrsFeatureCollection.prototype.type;
    /** @type {?} */
    LrsFeatureCollection.prototype.features;
    /** @type {?} */
    LrsFeatureCollection.prototype.bbox;
}
