// @tss/sdk/esm2015/lib/lrs/lrs-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { MapFeature } from '../maps/map-feature';
import { isLine } from '../geometry/geometry/is-line.function';
import { isPoint } from '../geometry/geometry/is-point.function';
import { WGS_84_SRID } from '../geometry/geometry.constants';
import { clone } from '../core/clone/clone.function';
export class LrsFeature extends MapFeature {
    /**
     * @return {?}
     */
    get layerId() {
        return this.layer ? this.layer.id : null;
    }
    /**
     * @return {?}
     */
    get fromDate() {
        return this.getValue(this.layer.fromDateFieldName);
    }
    /**
     * @return {?}
     */
    get toDate() {
        return this.getValue(this.layer.toDateFieldName);
    }
    /**
     * @return {?}
     */
    get isRetired() {
        return Boolean(this.toDate);
    }
    /**
     * @return {?}
     */
    get isActive() {
        return !this.isRetired;
    }
    /**
     * @param {?} feature
     * @param {?} layer
     * @param {?=} spatialReferenceId
     */
    constructor(feature, layer, spatialReferenceId = WGS_84_SRID) {
        super(feature, layer, spatialReferenceId);
        if (this.geometry && this.isGeometryTypeInvalid()) {
            throw new Error(`Unsupported geometry type of ${this.geometry.type} detected.`);
        }
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} date
     * @return {THIS}
     */
    setFromDate(date) {
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.fromDateFieldName, date);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} date
     * @return {THIS}
     */
    setToDate(date) {
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toDateFieldName, date);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    retire() {
        if ((/** @type {?} */ (this)).layer.toDateFieldName) {
            (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toDateFieldName, new Date());
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const copyFeature = clone(this.toFeature());
        return new LrsFeature(copyFeature, this.layer);
    }
    /**
     * @protected
     * @return {?}
     */
    isGeometryTypeValid() {
        return isLine(this.geometry) || isPoint(this.geometry);
    }
    /**
     * @private
     * @return {?}
     */
    isGeometryTypeInvalid() {
        return !this.isGeometryTypeValid();
    }
}
if (false) {
    /** @type {?} */
    LrsFeature.prototype.routeId;
    /** @type {?} */
    LrsFeature.prototype.fromMeasure;
    /** @type {?} */
    LrsFeature.prototype.toMeasure;
    /** @type {?} */
    LrsFeature.prototype.layer;
    /** @type {?} */
    LrsFeature.prototype.displayValue;
    /** @type {?} */
    LrsFeature.prototype.styleValue;
    /** @type {?} */
    LrsFeature.prototype.viewDate;
    /** @type {?} */
    LrsFeature.prototype.gdbVersion;
}
