// @tss/sdk/esm2015/lib/security/roles/role-privilege.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { SecurityConfig } from '../security-config';
import { MISSING_ROLE_PARAM_EXCEPTION_MSG } from './role';
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { join } from '../../core/url/join.function';
import { UNSUPPORTED_GET_EXCEPTION_MSG, UNSUPPORTED_CREATE_EXCEPTION_MSG } from '../../core/resource/resource.constants';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../security-config";
/** @type {?} */
export const MISSING_PRIVILEGE_PARAM_EXCEPTION_MSG = 'Missing privilege identifier parameter';
/**
 * Service to provide all CRUD operations for privilege/role assignment.
 *
 * \@example
 * ## Gets privileges for a given role:
 * service.privilegesByRole('879e27e1-b139-4348-86ca-ec4d5651475f')
 *
 * // Which is just a shortcut for:
 * service.query({ roleId: 'admin' });
 *
 */
export class RolePrivilegeService extends BaseResourceService {
    /**
     * @param {?} http
     * @param {?} config
     */
    constructor(http, config) {
        super(http, {
            url: join(config.serviceUrl, config.rolesPrivilegesEndpoint),
            idFieldName: 'roleId'
        });
        this.http = http;
    }
    /**
     * Retrieves a list of privileges for a given role ID.
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    query(routeTokens, params) {
        if (!routeTokens || !routeTokens.roleId) {
            throw new Error(MISSING_ROLE_PARAM_EXCEPTION_MSG);
        }
        return super.query(routeTokens, params);
    }
    /**
     * This method is not supported for this resource. Please use "privilegesByRole".
     * @param {?} id
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    get(id, routeTokens, params) {
        throw new Error(UNSUPPORTED_GET_EXCEPTION_MSG);
    }
    /**
     * This method is not supported for this resource. Please use "addPrivilegeToRole".
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    create(entity, routeTokens, params) {
        throw new Error(UNSUPPORTED_CREATE_EXCEPTION_MSG);
    }
    /**
     * Updates a privilege assignment (adds a privilege to a role).
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    update(entity, routeTokens, params) {
        if (!routeTokens || !routeTokens.roleId) {
            throw new Error(MISSING_ROLE_PARAM_EXCEPTION_MSG);
        }
        else if (!routeTokens.privilegeId) {
            throw new Error(MISSING_PRIVILEGE_PARAM_EXCEPTION_MSG);
        }
        return super.update(null, routeTokens, params);
    }
    /**
     * Removes a privilege from a role assignment.
     * @param {?} id
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    delete(id, routeTokens, params) {
        if (!routeTokens || !routeTokens.roleId) {
            throw new Error(MISSING_ROLE_PARAM_EXCEPTION_MSG);
        }
        else if (!routeTokens.privilegeId) {
            throw new Error(MISSING_PRIVILEGE_PARAM_EXCEPTION_MSG);
        }
        return super.delete(id, routeTokens, params);
    }
    /**
     * Retrieves all privileges assigned to the given role ID.
     * @param {?} roleId
     * @param {?=} params
     * @return {?}
     */
    privilegesByRole(roleId, params) {
        return this.query({ roleId }, params);
    }
    /**
     * Adds a privilege to role.
     * @param {?} roleId
     * @param {?} privilegeId
     * @return {?}
     */
    addPrivilegeToRole(roleId, privilegeId) {
        return this.update(null, { roleId, privilegeId });
    }
    /**
     * Removes a privilege from the given role.
     * @param {?} roleId
     * @param {?} privilegeId
     * @return {?}
     */
    removePrivilegeFromRole(roleId, privilegeId) {
        return this.delete(roleId, { roleId, privilegeId });
    }
}
RolePrivilegeService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
RolePrivilegeService.ctorParameters = () => [
    { type: HttpClient },
    { type: SecurityConfig }
];
/** @nocollapse */ RolePrivilegeService.ngInjectableDef = i0.defineInjectable({ factory: function RolePrivilegeService_Factory() { return new RolePrivilegeService(i0.inject(i1.HttpClient), i0.inject(i2.SecurityConfig)); }, token: RolePrivilegeService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    RolePrivilegeService.prototype.http;
}
