// @tss/sdk/esm2015/lib/lrs/networks/network-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TSS.
import { LrsFeature } from '../lrs-feature';
import { toFeature } from '../feature';
import { getPoint } from '../../geometry/line/get-point.function';
import { getMeasure } from '../../geometry/line/get-measure.function';
import { WGS_84_SRID } from '../../geometry/geometry.constants';
import { clone } from '../../core/clone/clone.function';
export class NetworkFeature extends LrsFeature {
    /**
     * @param {?} feature
     * @param {?} layer
     * @param {?=} spatialReferenceId
     */
    constructor(feature, layer, spatialReferenceId = WGS_84_SRID) {
        super(feature, layer, spatialReferenceId);
        this.routeId = this.getValue(layer.compositeRouteIdFieldName);
        this.displayValue = this.displayValue || this.routeId;
        this.setMeasureFromGeometry();
    }
    /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    static create(layer, properties = {}, geometry = null) {
        /** @type {?} */
        const feature = toFeature(properties, (/** @type {?} */ (geometry)));
        return new NetworkFeature(feature, layer);
    }
    /**
     * @param {?} point
     * @return {?}
     */
    getMeasure(point) {
        return getMeasure((/** @type {?} */ (this.geometry)), point);
    }
    /**
     * @param {?} measure
     * @return {?}
     */
    getPoint(measure) {
        return getPoint((/** @type {?} */ (this.geometry)), measure);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const copyFeature = clone(this.toFeature());
        return new NetworkFeature(copyFeature, this.layer);
    }
}
if (false) {
    /** @type {?} */
    NetworkFeature.prototype.layer;
}
