// @tss/sdk/esm5/lib/cache/cache.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { AutoConfig } from '../core/auto-config.decorator';
import { isRouteMatch } from '../http/is-route-match.function';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing global data caching settings.
 */
var CacheConfig = /** @class */ (function () {
    function CacheConfig() {
        /**
         * Determines whether caching should be used or not. When disabled, all
         * caching-related functionality is ignored (not used).
         */
        this.cacheEnabled = false;
        /**
         * List of configurations that describe what to cache and how to cache it.
         * The main parts of this configuration include the "route" (how to match what
         * to cache) and the "strategy" (whether to look for content in the cache first
         * or try to make the HTTP request first).
         */
        this.cacheDefitions = [];
        /**
         * When set, this strategy will trump all strategies defined in the "Cache Defintions".
         * WARNING: This is a global setting and can have adverse affects when used improperly.
         * The main use case for this setting is to allow for the global cache strategy of
         * 'cacheOnly', which will allow an application to force an "offline" state (forcing
         * requests to pull from cache only).
         */
        this.forcedGlobalCacheStrategy = null;
    }
    CacheConfig_1 = CacheConfig;
    /**
     * @param {?} config
     * @return {?}
     */
    CacheConfig.create = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return Object.assign(new CacheConfig_1(), config);
    };
    /**
     * @param {...?} defs
     * @return {?}
     */
    CacheConfig.enable = /**
     * @param {...?} defs
     * @return {?}
     */
    function () {
        var defs = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            defs[_i] = arguments[_i];
        }
        return CacheConfig_1.create({
            cacheEnabled: true,
            cacheDefitions: defs,
        });
    };
    /**
     * @param {?} route
     * @return {?}
     */
    CacheConfig.prototype.getCacheDefinition = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return this.cacheDefitions.find((/**
         * @param {?} def
         * @return {?}
         */
        function (def) { return isRouteMatch(route, def.route); }));
    };
    var CacheConfig_1;
    CacheConfig.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ CacheConfig.ngInjectableDef = i0.defineInjectable({ factory: function CacheConfig_Factory() { return new CacheConfig(); }, token: CacheConfig, providedIn: "root" });
    /**
     * Configuration class for storing global data caching settings.
     */
    CacheConfig = CacheConfig_1 = tslib_1.__decorate([
        AutoConfig('CacheConfig')
    ], CacheConfig);
    return CacheConfig;
}());
export { CacheConfig };
if (false) {
    /**
     * Determines whether caching should be used or not. When disabled, all
     * caching-related functionality is ignored (not used).
     * @type {?}
     */
    CacheConfig.prototype.cacheEnabled;
    /**
     * List of configurations that describe what to cache and how to cache it.
     * The main parts of this configuration include the "route" (how to match what
     * to cache) and the "strategy" (whether to look for content in the cache first
     * or try to make the HTTP request first).
     * @type {?}
     */
    CacheConfig.prototype.cacheDefitions;
    /**
     * When set, this strategy will trump all strategies defined in the "Cache Defintions".
     * WARNING: This is a global setting and can have adverse affects when used improperly.
     * The main use case for this setting is to allow for the global cache strategy of
     * 'cacheOnly', which will allow an application to force an "offline" state (forcing
     * requests to pull from cache only).
     * @type {?}
     */
    CacheConfig.prototype.forcedGlobalCacheStrategy;
}
