// @tss/sdk/esm5/lib/cache/cache.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { HttpResponse } from '@angular/common/http';
// 3rd Party.
import { PersistentCacheStorage } from './cache.storage';
import * as i0 from "@angular/core";
import * as i1 from "./cache.storage";
/**
 * Cache Service is an observables based in-memory cache implementation
 * Keeps track of in-flight observables and sets a default expiry for cached values
 */
var CacheService = /** @class */ (function () {
    function CacheService(cache) {
        this.cache = cache;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype.get = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var cacheItem = this.cache.getObject(key);
        /** @type {?} */
        var now = new Date();
        /** @type {?} */
        var cacheData;
        if (cacheItem && cacheItem.expires && cacheItem.expires.getTime() < now.getTime()) {
            this.cache.remove(key);
        }
        else if (cacheItem) {
            // NOTE: We can't just pull the serialized data, because
            // we need an actual instance of the HttpResponse class.
            cacheData = Object.assign(new HttpResponse(), cacheItem.value);
        }
        else {
            cacheData = null;
        }
        return cacheData;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @param {?=} timeToLive
     * @return {?}
     */
    CacheService.prototype.set = /**
     * @param {?} key
     * @param {?} value
     * @param {?=} timeToLive
     * @return {?}
     */
    function (key, value, timeToLive) {
        if (timeToLive === void 0) { timeToLive = null; }
        /** @type {?} */
        var expires = timeToLive ? this.getExpireDate(timeToLive) : null;
        /** @type {?} */
        var cacheData = { expires: expires, value: value };
        this.cache.setObject(key, cacheData);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    CacheService.prototype.has = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.cache.has(key);
    };
    /**
     * @private
     * @param {?=} timeToLive
     * @return {?}
     */
    CacheService.prototype.getExpireDate = /**
     * @private
     * @param {?=} timeToLive
     * @return {?}
     */
    function (timeToLive) {
        if (timeToLive === void 0) { timeToLive = null; }
        /** @type {?} */
        var expires = new Date();
        expires.setSeconds(expires.getSeconds() + timeToLive);
        return expires;
    };
    CacheService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    CacheService.ctorParameters = function () { return [
        { type: PersistentCacheStorage }
    ]; };
    /** @nocollapse */ CacheService.ngInjectableDef = i0.defineInjectable({ factory: function CacheService_Factory() { return new CacheService(i0.inject(i1.PersistentCacheStorage)); }, token: CacheService, providedIn: "root" });
    return CacheService;
}());
export { CacheService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CacheService.prototype.cache;
}
