// @tss/sdk/esm5/lib/geolocation/geolocator/base-geolocator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { GeolocatorAccuracyUnit } from './geolocator-accuracy-unit';
import { Logger } from '../../core/logging/logger.service';
import * as i0 from "@angular/core";
import * as i1 from "../../core/logging/logger.service";
/**
 * Wrapper for geolocation service. The wrapper primarily exists to abstract the direct interaction with
 * the "navigator" object and to allow for injecting other implementations of geolocation functionality.
 * @abstract
 */
var BaseGeolocator = /** @class */ (function () {
    function BaseGeolocator(log) {
        this.log = log;
        this.capabilities = [];
        this.accuracyUnit = GeolocatorAccuracyUnit.Meter;
    }
    /**
     * @param {?} capability
     * @return {?}
     */
    BaseGeolocator.prototype.hasCapability = /**
     * @param {?} capability
     * @return {?}
     */
    function (capability) {
        return this.capabilities.indexOf(capability) > -1;
    };
    BaseGeolocator.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    BaseGeolocator.ctorParameters = function () { return [
        { type: Logger }
    ]; };
    /** @nocollapse */ BaseGeolocator.ngInjectableDef = i0.defineInjectable({ factory: function BaseGeolocator_Factory() { return new BaseGeolocator(i0.inject(i1.Logger)); }, token: BaseGeolocator, providedIn: "root" });
    return BaseGeolocator;
}());
export { BaseGeolocator };
if (false) {
    /** @type {?} */
    BaseGeolocator.prototype.capabilities;
    /** @type {?} */
    BaseGeolocator.prototype.accuracyUnit;
    /**
     * @type {?}
     * @protected
     */
    BaseGeolocator.prototype.log;
    /**
     * @abstract
     * @return {?}
     */
    BaseGeolocator.prototype.name = function () { };
    /**
     * @abstract
     * @return {?}
     */
    BaseGeolocator.prototype.description = function () { };
    /**
     * Get the device's current position.
     * @abstract
     * @param {?=} options
     * @return {?}
     */
    BaseGeolocator.prototype.getCurrentPosition = function (options) { };
    /**
     * Watch the current device's position.  Clear the watch by unsubscribing from
     * Observable changes.
     *
     * ```typescript
     * const subscription = this.geolocation.watchPosition()
     *                               .filter(p => Boolean(p.coords)) //Filter Out Errors
     *                               .subscribe(position =>
     *   console.log(position.coords.longitude + ' ' + position.coords.latitude)
     * );
     *
     * // To stop notifications
     * subscription.unsubscribe();
     * ```
     * @abstract
     * @param {?=} options
     * @return {?}
     */
    BaseGeolocator.prototype.watchPosition = function (options) { };
    /**
     * @abstract
     * @return {?}
     */
    BaseGeolocator.prototype.isAvailable = function () { };
}
