// @tss/sdk/esm5/lib/geolocation/geolocator/geolocator-source.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable, Injector } from '@angular/core';
// 3rd party.
import { forkJoin, of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { GEOLOCATOR_REGISTRY } from '../geolocation.constants';
import * as i0 from "@angular/core";
var GeolocatorSourceService = /** @class */ (function () {
    function GeolocatorSourceService(injector) {
        this.injector = injector;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    GeolocatorSourceService.prototype.getGeolocator = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var registry = GEOLOCATOR_REGISTRY.find((/**
         * @param {?} registedSource
         * @return {?}
         */
        function (registedSource) { return registedSource.id === id; }));
        return this.getLocatorFromRegistry(registry);
    };
    /**
     * @return {?}
     */
    GeolocatorSourceService.prototype.getGeolocators = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return GEOLOCATOR_REGISTRY
            .map((/**
         * @param {?} registry
         * @return {?}
         */
        function (registry) { return _this.getLocatorFromRegistry(registry); }));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    GeolocatorSourceService.prototype.getGeolocatorSource = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.getGeolocatorSources()
            .pipe(map((/**
         * @param {?} devices
         * @return {?}
         */
        function (devices) { return devices.find((/**
         * @param {?} device
         * @return {?}
         */
        function (device) { return device.id === id; })); })));
    };
    /**
     * @return {?}
     */
    GeolocatorSourceService.prototype.getAvailableGeolocatorSources = /**
     * @return {?}
     */
    function () {
        return this.getGeolocatorSources()
            .pipe(map((/**
         * @param {?} devices
         * @return {?}
         */
        function (devices) { return devices.filter((/**
         * @param {?} device
         * @return {?}
         */
        function (device) { return device.isAvailable; })); })));
    };
    /**
     * @return {?}
     */
    GeolocatorSourceService.prototype.getGeolocatorSources = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var availableObservables = GEOLOCATOR_REGISTRY
            .map((/**
         * @param {?} registry
         * @return {?}
         */
        function (registry) {
            /** @type {?} */
            var locator = _this.getLocatorFromRegistry(registry);
            /** @type {?} */
            var source = {
                id: registry.id,
                name: locator.name,
                description: locator.description,
                accuracyUnit: locator.accuracyUnit,
                capabilities: locator.capabilities,
                isAvailable: false,
            };
            return locator.isAvailable()
                .pipe(tap((/**
             * @param {?} available
             * @return {?}
             */
            function (available) { return source.isAvailable = available; })), map((/**
             * @return {?}
             */
            function () { return source; })));
        }));
        return availableObservables.length === 0
            ? of([])
            : forkJoin(availableObservables);
    };
    /**
     * @private
     * @param {?} registry
     * @return {?}
     */
    GeolocatorSourceService.prototype.getLocatorFromRegistry = /**
     * @private
     * @param {?} registry
     * @return {?}
     */
    function (registry) {
        /** @type {?} */
        var instance = this.injector.get(registry.constructor);
        return instance;
    };
    GeolocatorSourceService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    GeolocatorSourceService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    /** @nocollapse */ GeolocatorSourceService.ngInjectableDef = i0.defineInjectable({ factory: function GeolocatorSourceService_Factory() { return new GeolocatorSourceService(i0.inject(i0.INJECTOR)); }, token: GeolocatorSourceService, providedIn: "root" });
    return GeolocatorSourceService;
}());
export { GeolocatorSourceService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GeolocatorSourceService.prototype.injector;
}
