// @tss/sdk/esm5/lib/data-source/data-source.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataSourceProviderType } from './data-source-provider-type';
/** @type {?} */
export var SELECT_ALL_WHERE_CLAUSE = '1=1';
/** @type {?} */
export var DATA_SOURCE_PROVIDERS = [
    { alias: 'SQL Server', provider: DataSourceProviderType.SqlServer, canUseEF: true, schedulingCompatible: true },
    { alias: 'SQLite', provider: DataSourceProviderType.SqlLite, canUseEF: false, schedulingCompatible: true },
    {
        alias: 'SQL Server Compact Edition',
        provider: DataSourceProviderType.SqlCompactEdition,
        canUseEF: true,
        schedulingCompatible: false
    },
    { alias: 'Oracle', provider: DataSourceProviderType.Oracle, canUseEF: false, schedulingCompatible: false },
    { alias: 'Oracle', provider: DataSourceProviderType.OracleDevArt, canUseEF: true, efOnly: true, schedulingCompatible: false }
];
/** @type {?} */
export var DATA_SOURCE_EXAMPLES = [
    {
        name: 'SQL Server Connection',
        providerName: DataSourceProviderType.SqlServer,
        connectionString: 'data source=YOUR_HOST, 1433;initial catalog=YOUR_DB;persist security info=True;'
            + 'user id=YOUR_USER;password=YOUR_PASSWORD;MultipleActiveResultSets=true;'
    },
    {
        name: 'SQL Server Compact (SQL CE) Connection',
        providerName: DataSourceProviderType.SqlCompactEdition,
        connectionString: 'Data Source=|DataDirectory|\\YOUR_DB_NAME.sdf;Max Database Size=4091'
    },
    {
        name: 'SQLite',
        providerName: DataSourceProviderType.SqlLite,
        connectionString: 'Data Source=|DataDirectory|\\YOUR_DB_NAME.sqlite;Version=3'
    },
    {
        name: 'Oracle Connection (without TNS names entry)',
        providerName: DataSourceProviderType.Oracle,
        connectionString: 'Data Source=(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)'
            + '(HOST=YOUR_HOST)(PORT=1522)))(CONNECT_DATA=(SERVER=DEDICATED)(SERVICE_NAME=YOUR_SERVICE)));'
            + 'User Id=YOUR_USER;Password=YOUR_PASSWORD;'
    },
    {
        name: 'Oracle Connection (with TNS names entry)',
        providerName: DataSourceProviderType.Oracle,
        connectionString: 'Data Source=YOUR_SOURCE;User Id=YOUR_USER;Password=YOUR_PASSWORD;'
    },
    {
        name: 'Oracle Connection (without TNS names entry)',
        providerName: DataSourceProviderType.OracleDevArt,
        connectionString: 'Data Source=(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)'
            + '(HOST=YOUR_HOST)(PORT=1522)))(CONNECT_DATA=(SERVER=DEDICATED)(SERVICE_NAME=YOUR_SERVICE)));'
            + 'User Id=YOUR_USER;Password=YOUR_PASSWORD;'
    },
    {
        name: 'Oracle Connection (with TNS names entry)',
        providerName: DataSourceProviderType.OracleDevArt,
        connectionString: 'Data Source=YOUR_SOURCE;User Id=YOUR_USER;Password=YOUR_PASSWORD;'
    }
];
