// @tss/sdk/esm5/lib/queuing/queue-worker-token.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Guid } from '../core/uuid/guid';
import { getElapsedSeconds } from '../core/date-time/get-elapsed-seconds.function';
import { calculateRemainingSeconds } from '../core/stats/calculate-remaining-seconds.function';
/**
 * Unique token created for each queue worker
 */
var /**
 * Unique token created for each queue worker
 */
QueueWorkerToken = /** @class */ (function () {
    function QueueWorkerToken() {
        this.isWorking = false;
        this.isCanceled = false;
        this.id = Guid.newGuid().value;
    }
    Object.defineProperty(QueueWorkerToken.prototype, "isComplete", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.isWorking;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(QueueWorkerToken.prototype, "ellapsedSeconds", {
        get: /**
         * @return {?}
         */
        function () {
            return getElapsedSeconds(new Date(), this.startTime);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(QueueWorkerToken.prototype, "totalProcessed", {
        get: /**
         * @return {?}
         */
        function () {
            return this.totalComplete + this.totalFailed;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(QueueWorkerToken.prototype, "recordsProcessedPerSecond", {
        get: /**
         * @return {?}
         */
        function () {
            return this.totalComplete / this.ellapsedSeconds;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(QueueWorkerToken.prototype, "recordsProcessedPerMinute", {
        get: /**
         * @return {?}
         */
        function () {
            return this.recordsProcessedPerSecond * 60;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    QueueWorkerToken.prototype.start = /**
     * @return {?}
     */
    function () {
        this.reset();
        this.startTime = new Date();
        this.isWorking = true;
        return this;
    };
    /**
     * @return {?}
     */
    QueueWorkerToken.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.isWorking = false;
        this.isCanceled = false;
        this.totalComplete = 0;
        this.totalFailed = 0;
        this.startTime = null;
        this.endTime = null;
        return this;
    };
    /**
     * @param {?} remainingCount
     * @return {?}
     */
    QueueWorkerToken.prototype.calculateRemainingSeconds = /**
     * @param {?} remainingCount
     * @return {?}
     */
    function (remainingCount) {
        return this.isComplete
            ? 0
            : calculateRemainingSeconds(this.totalProcessed, this.ellapsedSeconds, remainingCount);
    };
    /**
     * @return {?}
     */
    QueueWorkerToken.prototype.complete = /**
     * @return {?}
     */
    function () {
        this.endTime = new Date();
        this.isWorking = false;
        return this;
    };
    /**
     * @return {?}
     */
    QueueWorkerToken.prototype.cancel = /**
     * @return {?}
     */
    function () {
        this.isCanceled = true;
        this.isWorking = false;
        return this;
    };
    return QueueWorkerToken;
}());
/**
 * Unique token created for each queue worker
 */
export { QueueWorkerToken };
if (false) {
    /** @type {?} */
    QueueWorkerToken.prototype.id;
    /** @type {?} */
    QueueWorkerToken.prototype.isWorking;
    /** @type {?} */
    QueueWorkerToken.prototype.isCanceled;
    /** @type {?} */
    QueueWorkerToken.prototype.totalComplete;
    /** @type {?} */
    QueueWorkerToken.prototype.totalFailed;
    /** @type {?} */
    QueueWorkerToken.prototype.startTime;
    /** @type {?} */
    QueueWorkerToken.prototype.endTime;
}
