// @tss/sdk/esm5/lib/queuing/work-queue-status-event-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var WorkQueueStatusEventType = {
    UnitOfWorkStarted: 0,
    UnitOfWorkComplete: 1,
    UnitOfWorkStatusUpdate: 2,
    QueueWorkerStarted: 3,
    QueueWorkerComplete: 4,
    QueueStarted: 5,
    QueueComplete: 6,
};
export { WorkQueueStatusEventType };
WorkQueueStatusEventType[WorkQueueStatusEventType.UnitOfWorkStarted] = 'UnitOfWorkStarted';
WorkQueueStatusEventType[WorkQueueStatusEventType.UnitOfWorkComplete] = 'UnitOfWorkComplete';
WorkQueueStatusEventType[WorkQueueStatusEventType.UnitOfWorkStatusUpdate] = 'UnitOfWorkStatusUpdate';
WorkQueueStatusEventType[WorkQueueStatusEventType.QueueWorkerStarted] = 'QueueWorkerStarted';
WorkQueueStatusEventType[WorkQueueStatusEventType.QueueWorkerComplete] = 'QueueWorkerComplete';
WorkQueueStatusEventType[WorkQueueStatusEventType.QueueStarted] = 'QueueStarted';
WorkQueueStatusEventType[WorkQueueStatusEventType.QueueComplete] = 'QueueComplete';
