// @tss/sdk/esm2015/lib/data-source/attachments/attachment.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { DataSourceConfig } from '../data-source-config';
import { toFormDataString } from '../../http/to-form-data-string.function';
import { AttachmentType } from './attachment-type';
import { base64toBlob } from '../../core/media/base64-to-blob.function';
import { ATTACHMENT_FILE_FORM_KEY, ATTACHMENT_DATA_FORM_KEY } from './attachments.constants';
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { isString } from '../../core/type-check/is-string.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../data-source-config";
export class AttachmentService extends BaseResourceService {
    /**
     * @param {?} http
     * @param {?} dsConfig
     */
    constructor(http, dsConfig) {
        super(http, {
            url: dsConfig.serviceUrl,
            queryEndpoint: dsConfig.attachmentsEndpoint,
            getEndpoint: dsConfig.attachmentsEndpoint,
            updateEndpoint: dsConfig.attachmentsEndpoint,
            deleteEndpoint: dsConfig.attachmentsEndpoint,
            createEndpoint: dsConfig.attachmentsEndpoint,
        });
        this.http = http;
    }
    /**
     * @param {?} attachment
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    create(attachment, routeTokens, params) {
        /** @type {?} */
        const url = this.getUrl(this.config.createEndpoint, routeTokens);
        /** @type {?} */
        const formData = new FormData();
        /** @type {?} */
        let fileData;
        if (attachment.type === AttachmentType.Image && isString(attachment.data)) {
            fileData = base64toBlob((/** @type {?} */ (attachment.data)));
        }
        else if (attachment.type === AttachmentType.Text) {
            fileData = new Blob([attachment.data]);
        }
        else {
            fileData = (/** @type {?} */ (attachment.data));
        }
        delete attachment.data;
        formData.append(ATTACHMENT_FILE_FORM_KEY, fileData);
        formData.append(ATTACHMENT_DATA_FORM_KEY, toFormDataString(attachment));
        return this.http.post(url, formData, { params, reportProgress: true });
    }
    /**
     * @param {?} attachmentId
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    download(attachmentId, routeTokens, params) {
        // TODO: Complete - RG.
        // const url = this.getUrl(this.config.getEndpoint, Object.assign({ action: 'download' }, routeTokens));
        throw new Error('Not implemented yet');
    }
}
AttachmentService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AttachmentService.ctorParameters = () => [
    { type: HttpClient },
    { type: DataSourceConfig }
];
/** @nocollapse */ AttachmentService.ngInjectableDef = i0.defineInjectable({ factory: function AttachmentService_Factory() { return new AttachmentService(i0.inject(i1.HttpClient), i0.inject(i2.DataSourceConfig)); }, token: AttachmentService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AttachmentService.prototype.http;
}
