// @tss/sdk/esm2015/lib/core/uuid/guid.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { uuid } from './uuid.function';
import { INVALID_GUID_ERROR_MSG } from './guid.constants';
/**
 * Represents a Global Unique Identifier (GUID). GUID creation ported from:
 * https://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
 */
export class Guid {
    /**
     * @param {?=} value
     */
    constructor(value) {
        this.value = value || Guid.empty;
        if (!Guid.isValid(value)) {
            throw new Error(INVALID_GUID_ERROR_MSG);
        }
    }
    /**
     * @return {?}
     */
    get isEmpty() {
        return this.value === Guid.empty;
    }
    /**
     * @return {?}
     */
    static newGuid() {
        return new Guid(uuid());
    }
    /**
     * @param {?} str
     * @return {?}
     */
    static isValid(str) {
        /** @type {?} */
        const validRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
        return validRegex.test(str);
    }
    /**
     * @param {?} guid
     * @return {?}
     */
    equals(guid) {
        return guid.toString() === this.toString();
    }
    /**
     * @return {?}
     */
    toString() {
        return this.value;
    }
    /**
     * @return {?}
     */
    toJSON() {
        return this.value;
    }
}
Guid.empty = '00000000-0000-0000-0000-000000000000';
if (false) {
    /** @type {?} */
    Guid.empty;
    /** @type {?} */
    Guid.prototype.value;
}
