// @tss/sdk/esm5/lib/data-source/attachments/attachment.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { DataSourceConfig } from '../data-source-config';
import { toFormDataString } from '../../http/to-form-data-string.function';
import { AttachmentType } from './attachment-type';
import { base64toBlob } from '../../core/media/base64-to-blob.function';
import { ATTACHMENT_FILE_FORM_KEY, ATTACHMENT_DATA_FORM_KEY } from './attachments.constants';
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { isString } from '../../core/type-check/is-string.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../data-source-config";
var AttachmentService = /** @class */ (function (_super) {
    tslib_1.__extends(AttachmentService, _super);
    function AttachmentService(http, dsConfig) {
        var _this = _super.call(this, http, {
            url: dsConfig.serviceUrl,
            queryEndpoint: dsConfig.attachmentsEndpoint,
            getEndpoint: dsConfig.attachmentsEndpoint,
            updateEndpoint: dsConfig.attachmentsEndpoint,
            deleteEndpoint: dsConfig.attachmentsEndpoint,
            createEndpoint: dsConfig.attachmentsEndpoint,
        }) || this;
        _this.http = http;
        return _this;
    }
    /**
     * @param {?} attachment
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    AttachmentService.prototype.create = /**
     * @param {?} attachment
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    function (attachment, routeTokens, params) {
        /** @type {?} */
        var url = this.getUrl(this.config.createEndpoint, routeTokens);
        /** @type {?} */
        var formData = new FormData();
        /** @type {?} */
        var fileData;
        if (attachment.type === AttachmentType.Image && isString(attachment.data)) {
            fileData = base64toBlob((/** @type {?} */ (attachment.data)));
        }
        else if (attachment.type === AttachmentType.Text) {
            fileData = new Blob([attachment.data]);
        }
        else {
            fileData = (/** @type {?} */ (attachment.data));
        }
        delete attachment.data;
        formData.append(ATTACHMENT_FILE_FORM_KEY, fileData);
        formData.append(ATTACHMENT_DATA_FORM_KEY, toFormDataString(attachment));
        return this.http.post(url, formData, { params: params, reportProgress: true });
    };
    /**
     * @param {?} attachmentId
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    AttachmentService.prototype.download = /**
     * @param {?} attachmentId
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    function (attachmentId, routeTokens, params) {
        // TODO: Complete - RG.
        // const url = this.getUrl(this.config.getEndpoint, Object.assign({ action: 'download' }, routeTokens));
        throw new Error('Not implemented yet');
    };
    AttachmentService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AttachmentService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: DataSourceConfig }
    ]; };
    /** @nocollapse */ AttachmentService.ngInjectableDef = i0.defineInjectable({ factory: function AttachmentService_Factory() { return new AttachmentService(i0.inject(i1.HttpClient), i0.inject(i2.DataSourceConfig)); }, token: AttachmentService, providedIn: "root" });
    return AttachmentService;
}(BaseResourceService));
export { AttachmentService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AttachmentService.prototype.http;
}
