// @tss/sdk/esm2015/lib/http/http.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { AutoConfig } from '../core/auto-config.decorator';
import { isRouteMatch } from './is-route-match.function';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing global data caching settings.
 */
let HttpConfig = /**
 * Configuration class for storing global data caching settings.
 */
class HttpConfig {
    constructor() {
        this.paramInjectionsEnabled = true;
        this.paramInjectionDefitions = [];
    }
    /**
     * @return {?}
     */
    get hasInjections() {
        return Boolean(this.paramInjectionDefitions.length);
    }
    /**
     * @param {?} route
     * @param {?} paramName
     * @param {?} paramValue
     * @return {?}
     */
    addOrUpdateParameterInjection(route, paramName, paramValue) {
        /** @type {?} */
        const definition = {
            route,
            paramName,
            paramValue
        };
        /** @type {?} */
        const existingDefinition = this.paramInjectionDefitions.find((/**
         * @param {?} def
         * @return {?}
         */
        def => def.route === route && def.paramName === paramName));
        if (existingDefinition) {
            Object.assign(existingDefinition, definition);
            return existingDefinition;
        }
        else {
            this.paramInjectionDefitions.push(definition);
            return definition;
        }
    }
    /**
     * @param {?} route
     * @param {?} paramName
     * @param {?} paramValue
     * @return {?}
     */
    addParameterInjection(route, paramName, paramValue) {
        /** @type {?} */
        const definition = {
            route,
            paramName,
            paramValue
        };
        this.paramInjectionDefitions.push(definition);
        return definition;
    }
    /**
     * @param {?} route
     * @return {?}
     */
    getInjectionDefinitions(route) {
        return this.paramInjectionDefitions.filter((/**
         * @param {?} def
         * @return {?}
         */
        def => isRouteMatch(route, def.route)));
    }
    /**
     * @param {?} route
     * @return {?}
     */
    getFirstInjectionDefinition(route) {
        return this.paramInjectionDefitions.find((/**
         * @param {?} def
         * @return {?}
         */
        def => isRouteMatch(route, def.route)));
    }
};
HttpConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ HttpConfig.ngInjectableDef = i0.defineInjectable({ factory: function HttpConfig_Factory() { return new HttpConfig(); }, token: HttpConfig, providedIn: "root" });
/**
 * Configuration class for storing global data caching settings.
 */
HttpConfig = tslib_1.__decorate([
    AutoConfig('HttpConfig')
], HttpConfig);
export { HttpConfig };
if (false) {
    /** @type {?} */
    HttpConfig.prototype.paramInjectionsEnabled;
    /** @type {?} */
    HttpConfig.prototype.paramInjectionDefitions;
}
