// @tss/sdk/esm2015/lib/http/http-param.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { Logger } from '../core/logging/logger.service';
import { HttpConfig } from './http.config';
import * as i0 from "@angular/core";
import * as i1 from "./http.config";
import * as i2 from "../core/logging/logger.service";
export class HttpParamInterceptor {
    /**
     * @param {?} config
     * @param {?} log
     */
    constructor(config, log) {
        this.config = config;
        this.log = log;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        /** @type {?} */
        const httpParamInjections = this.config.paramInjectionsEnabled
            ? this.config.getInjectionDefinitions(request.urlWithParams)
            : [];
        if (httpParamInjections.length) {
            request = this.generateRequestWithParams(request, httpParamInjections);
        }
        return next.handle(request);
    }
    /**
     * @param {?} request
     * @param {?} httpParamInjections
     * @return {?}
     */
    generateRequestWithParams(request, httpParamInjections) {
        /** @type {?} */
        let params = request.params;
        httpParamInjections
            .forEach((/**
         * @param {?} paramInjection
         * @return {?}
         */
        paramInjection => {
            this.log.debug(`Injecting URL '${request.url}' with parameter value (name: '${paramInjection.paramName}')`);
            params = params.set(paramInjection.paramName, paramInjection.paramValue);
        }));
        return request.clone({ params });
    }
}
HttpParamInterceptor.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
HttpParamInterceptor.ctorParameters = () => [
    { type: HttpConfig },
    { type: Logger }
];
/** @nocollapse */ HttpParamInterceptor.ngInjectableDef = i0.defineInjectable({ factory: function HttpParamInterceptor_Factory() { return new HttpParamInterceptor(i0.inject(i1.HttpConfig), i0.inject(i2.Logger)); }, token: HttpParamInterceptor, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    HttpParamInterceptor.prototype.config;
    /**
     * @type {?}
     * @private
     */
    HttpParamInterceptor.prototype.log;
}
