// @tss/sdk/esm5/lib/core/notify/notification-subscription-tracker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { tap, publishReplay } from 'rxjs/operators';
import { BaseNotificationTracker } from './base-notification-tracker';
/**
 * @template T
 */
var /**
 * @template T
 */
NotificationSubscriptionTracker = /** @class */ (function (_super) {
    tslib_1.__extends(NotificationSubscriptionTracker, _super);
    function NotificationSubscriptionTracker(source, notify, log, config) {
        if (config === void 0) { config = {}; }
        var _this = _super.call(this, notify, log, config) || this;
        _this.source = source;
        // We need to do a timeout here so that the chained events coming in after this method will be fired
        // before this subscription is added - this is necessary for any synchronous/immediate observables.
        _this.source = _this.source.pipe(tap((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            setTimeout((/**
             * @return {?}
             */
            function () { return _this.onSuccess(x); }));
        }), (/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            setTimeout((/**
             * @return {?}
             */
            function () { return _this.onFail(error); }));
        })));
        if (_this.config.useSingletonSubscriptionEmission) {
            _this.source = _this.source.pipe(publishReplay());
        }
        return _this;
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.wait = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        return (/** @type {?} */ (_super.prototype.wait.call(this, message, title, config)));
    };
    /**
     * @param {?} predicate
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.waitIf = /**
     * @param {?} predicate
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (predicate, message, title, config) {
        return (/** @type {?} */ (_super.prototype.waitIf.call(this, predicate, message, title, config)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.success = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        return (/** @type {?} */ (_super.prototype.success.call(this, message, title, config)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.error = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        return (/** @type {?} */ (_super.prototype.error.call(this, message, title, config)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.warn = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        return (/** @type {?} */ (_super.prototype.warn.call(this, message, title, config)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.info = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        return (/** @type {?} */ (_super.prototype.info.call(this, message, title, config)));
    };
    /**
     * @param {?} onDone
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.finally = /**
     * @param {?} onDone
     * @return {?}
     */
    function (onDone) {
        return (/** @type {?} */ (_super.prototype.finally.call(this, onDone)));
    };
    /**
     * @param {?=} next
     * @param {?=} error
     * @param {?=} complete
     * @return {?}
     */
    NotificationSubscriptionTracker.prototype.subscribe = /**
     * @param {?=} next
     * @param {?=} error
     * @param {?=} complete
     * @return {?}
     */
    function (next, error, complete) {
        return this.source.subscribe(next, error, complete);
    };
    return NotificationSubscriptionTracker;
}(BaseNotificationTracker));
/**
 * @template T
 */
export { NotificationSubscriptionTracker };
if (false) {
    /** @type {?} */
    NotificationSubscriptionTracker.prototype.source;
}
