// @tss/sdk/esm5/lib/core/notify/notification-promise-tracker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BaseNotificationTracker } from './base-notification-tracker';
/**
 * @template T
 */
var /**
 * @template T
 */
NotificationPromiseTracker = /** @class */ (function (_super) {
    tslib_1.__extends(NotificationPromiseTracker, _super);
    function NotificationPromiseTracker(source, notify, log, config) {
        if (config === void 0) { config = {}; }
        var _this = _super.call(this, notify, log, config) || this;
        _this.source = source;
        return _this;
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    NotificationPromiseTracker.prototype.wait = /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    function (message, title) {
        var _this = this;
        _super.prototype.wait.call(this, message, title);
        this.source = this.source
            .then((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            _this.onSuccess(result);
            return result;
        }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.onFail(error);
            throw error;
        }));
        return this;
    };
    /**
     * @param {?} predicate
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    NotificationPromiseTracker.prototype.waitIf = /**
     * @param {?} predicate
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    function (predicate, message, title) {
        return (/** @type {?} */ (_super.prototype.waitIf.call(this, predicate, message, title)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    NotificationPromiseTracker.prototype.success = /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    function (message, title) {
        return (/** @type {?} */ (_super.prototype.success.call(this, message, title)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    NotificationPromiseTracker.prototype.error = /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    function (message, title) {
        return (/** @type {?} */ (_super.prototype.error.call(this, message, title)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    NotificationPromiseTracker.prototype.warn = /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    function (message, title) {
        return (/** @type {?} */ (_super.prototype.warn.call(this, message, title)));
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    NotificationPromiseTracker.prototype.info = /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    function (message, title) {
        return (/** @type {?} */ (_super.prototype.info.call(this, message, title)));
    };
    /**
     * @param {?} onDone
     * @return {?}
     */
    NotificationPromiseTracker.prototype.finally = /**
     * @param {?} onDone
     * @return {?}
     */
    function (onDone) {
        return (/** @type {?} */ (_super.prototype.finally.call(this, onDone)));
    };
    /**
     * @param {?=} onfulfilled
     * @return {?}
     */
    NotificationPromiseTracker.prototype.then = /**
     * @param {?=} onfulfilled
     * @return {?}
     */
    function (onfulfilled) {
        return this.source.then(onfulfilled);
    };
    /**
     * @param {?=} onrejected
     * @return {?}
     */
    NotificationPromiseTracker.prototype.catch = /**
     * @param {?=} onrejected
     * @return {?}
     */
    function (onrejected) {
        return this.source.catch(onrejected);
    };
    return NotificationPromiseTracker;
}(BaseNotificationTracker));
/**
 * @template T
 */
export { NotificationPromiseTracker };
if (false) {
    /** @type {?} */
    NotificationPromiseTracker.prototype.source;
}
