// @tss/sdk/esm5/lib/core/notify/notification-tracker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 * @template T
 */
export function NotificationTracker() { }
if (false) {
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationTracker.prototype.wait = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationTracker.prototype.success = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationTracker.prototype.error = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationTracker.prototype.warn = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotificationTracker.prototype.info = function (message, title, config) { };
    /**
     * @param {?} onDone
     * @return {?}
     */
    NotificationTracker.prototype.finally = function (onDone) { };
}
