// @tss/sdk/esm5/lib/core/notify/notify.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { CoreConfig } from '../core.config';
import { NotifyType } from './notify-type';
import { Logger } from '../logging/logger.service';
import { NotificationSubscriptionTracker } from './notification-subscription-tracker';
import { NotificationPromiseTracker } from './notification-promise-tracker';
import { format } from '../formatting/format.function';
import * as i0 from "@angular/core";
import * as i1 from "../core.config";
import * as i2 from "../logging/logger.service";
/**
 * Notification service.
 */
var NotifyService = /** @class */ (function () {
    function NotifyService(config, log) {
        this.config = config;
        this.log = log;
    }
    /**
     * @param {?} type
     * @param {?} message
     * @param {?} title
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.notify = /**
     * @param {?} type
     * @param {?} message
     * @param {?} title
     * @param {?=} config
     * @return {?}
     */
    function (type, message, title, config) {
        // Add actual notification implementation here.
        return {};
    };
    /**
     * @param {?=} notifyId
     * @param {?=} containerId
     * @return {?}
     */
    NotifyService.prototype.clear = /**
     * @param {?=} notifyId
     * @param {?=} containerId
     * @return {?}
     */
    function (notifyId, containerId) {
    };
    /**
     * @template T
     * @param {?} source
     * @return {?}
     */
    NotifyService.prototype.with = /**
     * @template T
     * @param {?} source
     * @return {?}
     */
    function (source) {
        return new NotificationSubscriptionTracker(source, this, this.log);
    };
    /**
     * @template T
     * @param {?} source
     * @return {?}
     */
    NotifyService.prototype.promise = /**
     * @template T
     * @param {?} source
     * @return {?}
     */
    function (source) {
        return new NotificationPromiseTracker(source, this, this.log);
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.wait = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        if (title === void 0) { title = this.config.defaultNotifyWaitMsg; }
        return this.notify(NotifyType.Wait, message, title, config);
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.success = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        if (title === void 0) { title = this.config.defaultNotifySuccessMsg; }
        return this.notify(NotifyType.Success, message, title, config);
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} error
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.error = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} error
     * @param {?=} config
     * @return {?}
     */
    function (message, title, error, config) {
        if (title === void 0) { title = this.config.defaultNotifyErrorMsg; }
        if (error === void 0) { error = null; }
        return this.notify(NotifyType.Error, format(message, { error: error }), title, config);
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} error
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.warn = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} error
     * @param {?=} config
     * @return {?}
     */
    function (message, title, error, config) {
        if (title === void 0) { title = this.config.defaultNotifyWarnMsg; }
        if (error === void 0) { error = null; }
        return this.notify(NotifyType.Warn, format(message, { error: error }), title, config);
    };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyService.prototype.info = /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    function (message, title, config) {
        if (title === void 0) { title = this.config.defaultNotifyInfoMsg; }
        return this.notify(NotifyType.Info, message, title, config);
    };
    NotifyService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    NotifyService.ctorParameters = function () { return [
        { type: CoreConfig },
        { type: Logger }
    ]; };
    /** @nocollapse */ NotifyService.ngInjectableDef = i0.defineInjectable({ factory: function NotifyService_Factory() { return new NotifyService(i0.inject(i1.CoreConfig), i0.inject(i2.Logger)); }, token: NotifyService, providedIn: "root" });
    return NotifyService;
}());
export { NotifyService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    NotifyService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    NotifyService.prototype.log;
}
