// @tss/sdk/esm5/lib/orientation/device-orientation.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// 3rd Party.
import { Subject, fromEvent, merge } from 'rxjs';
import { getHeadingFromOrientation } from './heading/get-heading-from-orientation.function';
import { OrientationConfig } from './orientation.config';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./orientation.config";
/**
 * Geolocator service. Provides a single point of GPS interaction.
 */
var DeviceOrientationService = /** @class */ (function () {
    function DeviceOrientationService(config) {
        var _this = this;
        this.config = config;
        this.internalDeviceNeedsCalibrated = false;
        this.headingChangeSource = new Subject();
        this.orientationChangeSource = new Subject();
        this.orientationChange = this.orientationChangeSource.asObservable();
        this.headingChange = merge(this.headingChangeSource, this.orientationChange.pipe(map((/**
         * @param {?} orientation
         * @return {?}
         */
        function (orientation) { return orientation.heading; }))));
        fromEvent(window, 'deviceorientation').subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return _this.emitOrientationChange(event); }));
        fromEvent(window, 'oncompassneedscalibration').subscribe((/**
         * @return {?}
         */
        function () { return _this.internalDeviceNeedsCalibrated = true; }));
        fromEvent(window, 'compassneedscalibration').subscribe((/**
         * @return {?}
         */
        function () { return _this.internalDeviceNeedsCalibrated = true; }));
    }
    Object.defineProperty(DeviceOrientationService.prototype, "deviceNeedsCalibrated", {
        get: /**
         * @return {?}
         */
        function () {
            return this.internalDeviceNeedsCalibrated;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} orientation
     * @return {THIS}
     */
    DeviceOrientationService.prototype.mockOrientation = /**
     * @template THIS
     * @this {THIS}
     * @param {?} orientation
     * @return {THIS}
     */
    function (orientation) {
        if (!(/** @type {?} */ (this)).config.allowOrientationMocking) {
            throw new Error('Permission denied to mock device orientation');
        }
        (/** @type {?} */ (this)).orientationChangeSource.next(orientation);
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} heading
     * @return {THIS}
     */
    DeviceOrientationService.prototype.mockHeading = /**
     * @template THIS
     * @this {THIS}
     * @param {?} heading
     * @return {THIS}
     */
    function (heading) {
        if (!(/** @type {?} */ (this)).config.allowHeadingMocking) {
            throw new Error('Permission denied to mock heading');
        }
        (/** @type {?} */ (this)).headingChangeSource.next(heading);
        return (/** @type {?} */ (this));
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    DeviceOrientationService.prototype.emitOrientationChange = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.orientationChangeSource.next({
            alpha: event.alpha,
            beta: event.beta,
            gamma: event.gamma,
            heading: getHeadingFromOrientation(event)
        });
    };
    DeviceOrientationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DeviceOrientationService.ctorParameters = function () { return [
        { type: OrientationConfig }
    ]; };
    /** @nocollapse */ DeviceOrientationService.ngInjectableDef = i0.defineInjectable({ factory: function DeviceOrientationService_Factory() { return new DeviceOrientationService(i0.inject(i1.OrientationConfig)); }, token: DeviceOrientationService, providedIn: "root" });
    return DeviceOrientationService;
}());
export { DeviceOrientationService };
if (false) {
    /** @type {?} */
    DeviceOrientationService.prototype.orientationChange;
    /** @type {?} */
    DeviceOrientationService.prototype.headingChange;
    /** @type {?} */
    DeviceOrientationService.prototype.orientationChangeSource;
    /** @type {?} */
    DeviceOrientationService.prototype.headingChangeSource;
    /**
     * @type {?}
     * @private
     */
    DeviceOrientationService.prototype.internalDeviceNeedsCalibrated;
    /**
     * @type {?}
     * @private
     */
    DeviceOrientationService.prototype.config;
}
