// @tss/sdk/esm5/lib/app/settings/app-setting.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
// TSS.
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { ALL_APP_SETTINGS_QUERY_ID } from './app-setting';
import { AppConfig } from '../app.config';
import { join } from '../../core/url/join.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../app.config";
var AppSettingService = /** @class */ (function (_super) {
    tslib_1.__extends(AppSettingService, _super);
    function AppSettingService(http, config) {
        var _this = _super.call(this, http, {
            url: join(config.serviceUrl, config.appSettingsEndpoint),
            defaultRouteTokens: {
                appId: config.appId
            }
        }) || this;
        _this.http = http;
        return _this;
    }
    /**
     * Retrieves a list of application settings.
     */
    /**
     * Retrieves a list of application settings.
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    AppSettingService.prototype.queryAllApps = /**
     * Retrieves a list of application settings.
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    function (routeTokens, params) {
        routeTokens = routeTokens || {};
        routeTokens.appId = ALL_APP_SETTINGS_QUERY_ID;
        return _super.prototype.query.call(this, routeTokens, params);
    };
    AppSettingService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AppSettingService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: AppConfig }
    ]; };
    /** @nocollapse */ AppSettingService.ngInjectableDef = i0.defineInjectable({ factory: function AppSettingService_Factory() { return new AppSettingService(i0.inject(i1.HttpClient), i0.inject(i2.AppConfig)); }, token: AppSettingService, providedIn: "root" });
    return AppSettingService;
}(BaseResourceService));
export { AppSettingService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AppSettingService.prototype.http;
}
