// @tss/sdk/esm5/lib/data-source/apply-edits-seed/data-source-edit.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { DataSourceConfig } from '../data-source-config';
import { BaseEmptyResourceService } from '../../core/resource/base-empty-resource.service';
import { EMPTY_PARAM_VALUE } from '../../core/resource/resource.constants';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../data-source-config";
var DataSourceEditService = /** @class */ (function (_super) {
    tslib_1.__extends(DataSourceEditService, _super);
    function DataSourceEditService(http, dsConfig) {
        var _this = _super.call(this, http, {
            url: dsConfig.serviceUrl
        }) || this;
        _this.http = http;
        _this.dsConfig = dsConfig;
        return _this;
    }
    /**
     * @param {?} data
     * @param {?=} params
     * @return {?}
     */
    DataSourceEditService.prototype.applyEdits = /**
     * @param {?} data
     * @param {?=} params
     * @return {?}
     */
    function (data, params) {
        if (!data) {
            throw new Error('Missing or invalid apply edits seed data provided');
        }
        else if (!data.dataSourceId) {
            throw new Error("Missing 'dataSourceId' from the apply edits seed data");
        }
        /** @type {?} */
        var url = this.getUrlWithId(data.dataSourceId, this.dsConfig.applyEditsEndpoint, {});
        return this.http.post(url, data, { params: params });
    };
    /**
     * @param {?} dataSourceId
     * @param {?} table
     * @param {?} feature
     * @param {?=} version
     * @param {?=} params
     * @return {?}
     */
    DataSourceEditService.prototype.createRecord = /**
     * @param {?} dataSourceId
     * @param {?} table
     * @param {?} feature
     * @param {?=} version
     * @param {?=} params
     * @return {?}
     */
    function (dataSourceId, table, feature, version, params) {
        /** @type {?} */
        var url = this.getUrlWithId(dataSourceId, this.dsConfig.editDataEndpoint, { table: table });
        params = (params || new HttpParams())
            .set('version', version || EMPTY_PARAM_VALUE);
        return this.http.post(url, feature, { params: params });
    };
    DataSourceEditService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    DataSourceEditService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: DataSourceConfig }
    ]; };
    /** @nocollapse */ DataSourceEditService.ngInjectableDef = i0.defineInjectable({ factory: function DataSourceEditService_Factory() { return new DataSourceEditService(i0.inject(i1.HttpClient), i0.inject(i2.DataSourceConfig)); }, token: DataSourceEditService, providedIn: "root" });
    return DataSourceEditService;
}(BaseEmptyResourceService));
export { DataSourceEditService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    DataSourceEditService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    DataSourceEditService.prototype.dsConfig;
}
