// @tss/sdk/esm2015/lib/core/resource/base-resource.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TSS.
import { BaseEmptyResourceService } from './base-empty-resource.service';
/**
 * @abstract
 * @template T
 */
export class BaseResourceService extends BaseEmptyResourceService {
    /**
     * @param {?} http
     * @param {?} paramsOrUrl
     */
    constructor(http, paramsOrUrl) {
        super(http, paramsOrUrl);
        this.http = http;
    }
    /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    query(routeTokens, params) {
        /** @type {?} */
        const url = this.getUrl(this.config.queryEndpoint, routeTokens);
        return this.http.get(url, { params });
    }
    /**
     * @param {?} id
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    get(id, routeTokens, params) {
        /** @type {?} */
        const url = this.getUrlWithId(id, this.config.getEndpoint, routeTokens);
        return this.http.get(url, { params });
    }
    /**
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    create(entity, routeTokens, params) {
        /** @type {?} */
        const url = this.getUrl(this.config.createEndpoint, routeTokens);
        return this.http.post(url, entity, { params });
    }
    /**
     * @param {?} entity
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    update(entity, routeTokens, params) {
        /** @type {?} */
        const url = this.getUrlWithEntity(entity, this.config.updateEndpoint, routeTokens);
        return this.http.put(url, entity, { params });
    }
    /**
     * @param {?} id
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    delete(id, routeTokens, params) {
        /** @type {?} */
        const url = this.getUrlWithId(id, this.config.deleteEndpoint, routeTokens);
        return this.http.delete(url, { params });
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseResourceService.prototype.http;
}
