// @tss/sdk/esm2015/lib/core/resource/base-empty-resource.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TSS.
import { join } from '../url/join.function';
import { format } from '../formatting/format.function';
import { clean } from '../url/clean.function';
import { isString } from '../type-check/is-string.function';
import { NO_URL_ERROR_MSG } from './resource.constants';
/**
 * @abstract
 * @template T
 */
export class BaseEmptyResourceService {
    /**
     * @param {?} http
     * @param {?} paramsOrUrl
     */
    constructor(http, paramsOrUrl) {
        this.http = http;
        this.config = {
            url: null,
            idFieldName: 'id',
            cleanUrl: true,
        };
        if (isString(paramsOrUrl)) {
            this.config.url = (/** @type {?} */ (paramsOrUrl));
        }
        else {
            Object.assign(this.config, paramsOrUrl);
        }
        if (!this.config.url) {
            throw NO_URL_ERROR_MSG;
        }
    }
    /**
     * @protected
     * @param {?} entity
     * @param {?} endpoint
     * @param {?} tokens
     * @return {?}
     */
    getUrlWithEntity(entity, endpoint, tokens) {
        return this.getUrlWithId(entity ? entity[this.config.idFieldName] : null, endpoint, tokens);
    }
    /**
     * @protected
     * @param {?} id
     * @param {?} endpoint
     * @param {?} tokens
     * @return {?}
     */
    getUrlWithId(id, endpoint, tokens) {
        /** @type {?} */
        const inferredRouteTokens = {};
        inferredRouteTokens[this.config.idFieldName] = id;
        return this.getUrl(endpoint, Object.assign(inferredRouteTokens, tokens));
    }
    /**
     * @protected
     * @param {?} endpoint
     * @param {?} tokens
     * @return {?}
     */
    getUrl(endpoint, tokens) {
        /** @type {?} */
        const url = endpoint
            ? join(this.config.url, endpoint)
            : this.config.url;
        return this.replaceTokens(url, tokens);
    }
    /**
     * @protected
     * @param {?} url
     * @param {?} tokens
     * @return {?}
     */
    replaceTokens(url, tokens) {
        tokens = tokens || {};
        if (this.config.defaultRouteTokens) {
            tokens = Object.assign({}, this.config.defaultRouteTokens, tokens);
        }
        url = format(url, tokens, true);
        if (this.config.cleanUrl) {
            url = clean(url);
        }
        return url;
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseEmptyResourceService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    BaseEmptyResourceService.prototype.http;
}
