// @tss/sdk/esm5/lib/maps/drawing/layer-drawing-info.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { LayerDrawingInfoRenderer } from './layer-drawing-info-renderer';
import { transparencyToOpacity } from '../../core/color/transparency-to-opacity.function';
/**
 * @record
 */
export function DrawingInfoProperties() { }
if (false) {
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.renderer;
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.transparency;
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.positionRenderer;
    /** @type {?|undefined} */
    DrawingInfoProperties.prototype.labelingInfo;
}
var LayerDrawingInfo = /** @class */ (function () {
    function LayerDrawingInfo() {
        var baseDrawingInfos = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            baseDrawingInfos[_i] = arguments[_i];
        }
        this.merge.apply(this, tslib_1.__spread(baseDrawingInfos));
        this.transparency = this.transparency || 0;
        this.zIndex = this.zIndex || null;
        this.renderer = new LayerDrawingInfoRenderer(this.renderer);
        this.positionRenderer = new LayerDrawingInfoRenderer(this.positionRenderer);
        this.labelingInfo = this.labelingInfo || null;
    }
    Object.defineProperty(LayerDrawingInfo.prototype, "isEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.transparency
                && !this.zIndex
                && this.renderer.isEmpty
                && this.positionRenderer.isEmpty
                && !Object.keys(this.labelingInfo || {}).length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerDrawingInfo.prototype, "opacity", {
        get: /**
         * @return {?}
         */
        function () {
            return transparencyToOpacity(this.transparency);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @template THIS
     * @this {THIS}
     * @param {...?} targetDrawingInfos
     * @return {THIS}
     */
    LayerDrawingInfo.prototype.merge = /**
     * @template THIS
     * @this {THIS}
     * @param {...?} targetDrawingInfos
     * @return {THIS}
     */
    function () {
        var targetDrawingInfos = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            targetDrawingInfos[_i] = arguments[_i];
        }
        /** @type {?} */
        var properties = targetDrawingInfos.map((/**
         * @param {?} info
         * @return {?}
         */
        function (info) {
            return info && ((/** @type {?} */ (info))).toProperties
                ? ((/** @type {?} */ (info))).toProperties()
                : info;
        }));
        Object.assign.apply(Object, tslib_1.__spread([(/** @type {?} */ (this))], properties));
        return (/** @type {?} */ (this));
    };
    /**
     * @param {?=} value
     * @return {?}
     */
    LayerDrawingInfo.prototype.getRendererSymbol = /**
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        return this.renderer.getSymbol(value);
    };
    /**
     * @param {?=} value
     * @return {?}
     */
    LayerDrawingInfo.prototype.getPositionRendererSymbol = /**
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        return this.positionRenderer.getSymbol(value);
    };
    /**
     * @param {?=} value
     * @return {?}
     */
    LayerDrawingInfo.prototype.getBestRepresentativeColor = /**
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        return this.renderer.getBestRepresentativeColor(value);
    };
    /**
     * @return {?}
     */
    LayerDrawingInfo.prototype.toProperties = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var infoProperties = {};
        if (this.labelingInfo) {
            infoProperties.labelingInfo = this.labelingInfo;
        }
        if (this.positionRenderer && !this.positionRenderer.isEmpty) {
            infoProperties.positionRenderer = this.positionRenderer;
        }
        if (this.renderer && !this.renderer.isEmpty) {
            infoProperties.renderer = this.renderer;
        }
        if (this.transparency) {
            infoProperties.transparency = this.transparency;
        }
        return infoProperties;
    };
    return LayerDrawingInfo;
}());
export { LayerDrawingInfo };
if (false) {
    /** @type {?} */
    LayerDrawingInfo.prototype.renderer;
    /** @type {?} */
    LayerDrawingInfo.prototype.transparency;
    /** @type {?} */
    LayerDrawingInfo.prototype.zIndex;
    /** @type {?} */
    LayerDrawingInfo.prototype.positionRenderer;
    /** @type {?} */
    LayerDrawingInfo.prototype.labelingInfo;
}
