// @tss/sdk/esm5/lib/maps/drawing/layer-drawing-info-renderer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { isClassBreakInfoEmpty } from './class-break-info';
import { isUniqueValueInfoEmpty } from './unique-value-info';
import { isArray } from '../../core/type-check/is-array.function';
import { rgbToHex } from '../../core/color/rgb-to-hex.function';
import { removeWhere } from '../../core/array/remove-where.function';
/** @type {?} */
export var DEFAULT_UNIQUE_VALUE_INFO = null;
/** @type {?} */
export var DEFAULT_CLASS_BREAK_VALUE_INFO = null;
/** @type {?} */
export var DEFAULT_RENDERER_SYMBOL = {
    type: 'esriSLS',
    style: 'esriSLSSolid',
    color: '#000000',
    width: 2
};
/**
 * @param {?} symbol
 * @return {?}
 */
export function isRendererSymbolDefault(symbol) {
    return Object.keys(symbol).length === Object.keys(DEFAULT_RENDERER_SYMBOL).length
        && symbol.type === DEFAULT_RENDERER_SYMBOL.type
        && symbol.style === DEFAULT_RENDERER_SYMBOL.style
        && symbol.color === DEFAULT_RENDERER_SYMBOL.color
        && symbol.width === DEFAULT_RENDERER_SYMBOL.width;
}
var LayerDrawingInfoRenderer = /** @class */ (function () {
    function LayerDrawingInfoRenderer(base) {
        Object.assign(this, base);
        this.type = this.type || 'simple';
        this.minValue = this.minValue || 0;
        this.classBreakInfos = this.classBreakInfos || [];
        this.uniqueValueInfos = this.uniqueValueInfos || [];
        this.symbol = Object.assign({}, DEFAULT_RENDERER_SYMBOL, this.symbol);
        this.label = this.label || '';
        this.description = this.description || '';
        this.standardizeRenderSymbols();
    }
    Object.defineProperty(LayerDrawingInfoRenderer.prototype, "isSimple", {
        get: /**
         * @return {?}
         */
        function () {
            return this.type === 'simple';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerDrawingInfoRenderer.prototype, "isUniqueValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this.type === 'uniqueValue';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerDrawingInfoRenderer.prototype, "isClassBreak", {
        get: /**
         * @return {?}
         */
        function () {
            return this.type === 'classBreaks';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerDrawingInfoRenderer.prototype, "isEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.minValue
                && !this.classBreakInfos.length
                && !this.uniqueValueInfos.length
                && isRendererSymbolDefault(this.symbol);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayerDrawingInfoRenderer.prototype, "field", {
        get: /**
         * @return {?}
         */
        function () {
            return this.field1;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.field1 = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.getSymbol = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var symbol = this.symbol || this.defaultSymbol;
        if (this.isUniqueValue) {
            /** @type {?} */
            var uniqueValueInfo = this.getUniqueValueInfoOrDefault(value);
            symbol = isUniqueValueInfoEmpty(uniqueValueInfo) ? null : uniqueValueInfo.symbol;
        }
        else if (this.isClassBreak) {
            /** @type {?} */
            var classBreakInfo = this.getClassBreakInfoOrDefault(value);
            symbol = isClassBreakInfoEmpty(classBreakInfo) ? null : classBreakInfo.symbol;
        }
        return symbol || this.defaultSymbol || null;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.getUniqueValueInfo = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (!this.uniqueValueInfos || !this.uniqueValueInfos.length) {
            return null;
        }
        return this.uniqueValueInfos.find((/**
         * @param {?} info
         * @return {?}
         */
        function (info) {
            return info.value === value || String(info.value) === String(value);
        }));
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.getUniqueValueInfoOrDefault = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.getUniqueValueInfo(value) || DEFAULT_UNIQUE_VALUE_INFO;
    };
    /**
     * @param {?} classValue
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.getClassBreakInfo = /**
     * @param {?} classValue
     * @return {?}
     */
    function (classValue) {
        var _this = this;
        if (!this.classBreakInfos || !this.classBreakInfos.length) {
            return null;
        }
        return this.classBreakInfos.find((/**
         * @param {?} value
         * @param {?} index
         * @return {?}
         */
        function (value, index) { return _this.isValueInClassBreakRange(value, classValue, index); }));
    };
    /**
     * @param {?} value
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.getClassBreakInfoOrDefault = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.getClassBreakInfo(value) || DEFAULT_CLASS_BREAK_VALUE_INFO;
    };
    /**
     * @param {?} classBreakInfo
     * @param {?} value
     * @param {?} index
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.isValueInClassBreakRange = /**
     * @param {?} classBreakInfo
     * @param {?} value
     * @param {?} index
     * @return {?}
     */
    function (classBreakInfo, value, index) {
        /** @type {?} */
        var rangeMin = index === 0 ? this.minValue : this.classBreakInfos[index - 1].classMaxValue;
        /** @type {?} */
        var rangeMax = classBreakInfo.classMaxValue;
        return rangeMin < value && rangeMax > value;
    };
    /**
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.clearEmptyInfos = /**
     * @return {?}
     */
    function () {
        removeWhere(this.uniqueValueInfos, (/**
         * @param {?} info
         * @return {?}
         */
        function (info) { return isUniqueValueInfoEmpty(info); }));
        removeWhere(this.classBreakInfos, (/**
         * @param {?} info
         * @return {?}
         */
        function (info) { return isClassBreakInfoEmpty(info); }));
    };
    /**
     * @param {?=} value
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.getBestRepresentativeColor = /**
     * @param {?=} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var symbol = this.defaultSymbol || this.symbol;
        if (this.isUniqueValue && this.uniqueValueInfos.length) {
            /** @type {?} */
            var foundInfo = this.getUniqueValueInfo(value);
            symbol = foundInfo ? foundInfo.symbol : this.uniqueValueInfos[0].symbol;
        }
        else if (this.isClassBreak && this.classBreakInfos.length) {
            /** @type {?} */
            var foundInfo = this.getClassBreakInfo(value);
            symbol = foundInfo ? foundInfo.symbol : this.classBreakInfos[0].symbol;
        }
        return (/** @type {?} */ (symbol.color));
    };
    /**
     * @return {?}
     */
    LayerDrawingInfoRenderer.prototype.standardizeRenderSymbols = /**
     * @return {?}
     */
    function () {
        standardizeRenderSymbol(this.defaultSymbol);
        standardizeRenderSymbol(this.symbol);
        this.uniqueValueInfos.forEach((/**
         * @param {?} info
         * @return {?}
         */
        function (info) { return standardizeRenderSymbol(info.symbol); }));
        this.classBreakInfos.forEach((/**
         * @param {?} info
         * @return {?}
         */
        function (info) { return standardizeRenderSymbol(info.symbol); }));
    };
    return LayerDrawingInfoRenderer;
}());
export { LayerDrawingInfoRenderer };
if (false) {
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.type;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.field1;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.field2;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.field3;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.fieldDelimiter;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.valueExpression;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.classificationMethod;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.minValue;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.classBreakInfos;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.uniqueValueInfos;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.label;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.defaultLabel;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.description;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.defaultSymbol;
    /** @type {?} */
    LayerDrawingInfoRenderer.prototype.symbol;
}
/**
 * @param {?} symbol
 * @return {?}
 */
function standardizeRenderSymbol(symbol) {
    if (!symbol) {
        return;
    }
    if (isArray(symbol.color)) {
        symbol.color = rgbToHex((/** @type {?} */ (symbol.color)));
    }
    if (symbol.outline && isArray(symbol.outline.color)) {
        symbol.outline.color = rgbToHex((/** @type {?} */ (symbol.outline.color)));
    }
}
