// @tss/sdk/esm2015/lib/data-source/apply-edits-seed/data-source-edit.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { DataSourceConfig } from '../data-source-config';
import { BaseEmptyResourceService } from '../../core/resource/base-empty-resource.service';
import { EMPTY_PARAM_VALUE } from '../../core/resource/resource.constants';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../data-source-config";
export class DataSourceEditService extends BaseEmptyResourceService {
    /**
     * @param {?} http
     * @param {?} dsConfig
     */
    constructor(http, dsConfig) {
        super(http, {
            url: dsConfig.serviceUrl
        });
        this.http = http;
        this.dsConfig = dsConfig;
    }
    /**
     * @param {?} data
     * @param {?=} params
     * @return {?}
     */
    applyEdits(data, params) {
        if (!data) {
            throw new Error('Missing or invalid apply edits seed data provided');
        }
        else if (!data.dataSourceId) {
            throw new Error(`Missing 'dataSourceId' from the apply edits seed data`);
        }
        /** @type {?} */
        const url = this.getUrlWithId(data.dataSourceId, this.dsConfig.applyEditsEndpoint, {});
        return this.http.post(url, data, { params });
    }
    /**
     * @param {?} dataSourceId
     * @param {?} table
     * @param {?} feature
     * @param {?=} version
     * @param {?=} params
     * @return {?}
     */
    createRecord(dataSourceId, table, feature, version, params) {
        /** @type {?} */
        const url = this.getUrlWithId(dataSourceId, this.dsConfig.editDataEndpoint, { table });
        params = (params || new HttpParams())
            .set('version', version || EMPTY_PARAM_VALUE);
        return this.http.post(url, feature, { params });
    }
}
DataSourceEditService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
DataSourceEditService.ctorParameters = () => [
    { type: HttpClient },
    { type: DataSourceConfig }
];
/** @nocollapse */ DataSourceEditService.ngInjectableDef = i0.defineInjectable({ factory: function DataSourceEditService_Factory() { return new DataSourceEditService(i0.inject(i1.HttpClient), i0.inject(i2.DataSourceConfig)); }, token: DataSourceEditService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    DataSourceEditService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    DataSourceEditService.prototype.dsConfig;
}
