// @tss/sdk/esm2015/lib/core/environment/online-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// 3rd party.
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class OnlineOfflineService {
    constructor() {
        this.internalConnectionChanged = new Subject();
        this.connectionChanged = this.internalConnectionChanged.asObservable();
        window.addEventListener('online', (/**
         * @return {?}
         */
        () => this.updateOnlineStatus()));
        window.addEventListener('offline', (/**
         * @return {?}
         */
        () => this.updateOnlineStatus()));
    }
    /**
     * @return {?}
     */
    get isOnline() {
        return !!window.navigator.onLine;
    }
    /**
     * @private
     * @return {?}
     */
    updateOnlineStatus() {
        this.internalConnectionChanged.next(window.navigator.onLine);
    }
}
OnlineOfflineService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
OnlineOfflineService.ctorParameters = () => [];
/** @nocollapse */ OnlineOfflineService.ngInjectableDef = i0.defineInjectable({ factory: function OnlineOfflineService_Factory() { return new OnlineOfflineService(); }, token: OnlineOfflineService, providedIn: "root" });
if (false) {
    /** @type {?} */
    OnlineOfflineService.prototype.connectionChanged;
    /**
     * @type {?}
     * @private
     */
    OnlineOfflineService.prototype.internalConnectionChanged;
}
