// @tss/sdk/esm5/lib/http/http-param.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { Logger } from '../core/logging/logger.service';
import { HttpConfig } from './http.config';
import * as i0 from "@angular/core";
import * as i1 from "./http.config";
import * as i2 from "../core/logging/logger.service";
var HttpParamInterceptor = /** @class */ (function () {
    function HttpParamInterceptor(config, log) {
        this.config = config;
        this.log = log;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    HttpParamInterceptor.prototype.intercept = /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    function (request, next) {
        /** @type {?} */
        var httpParamInjections = this.config.paramInjectionsEnabled
            ? this.config.getInjectionDefinitions(request.urlWithParams)
            : [];
        if (httpParamInjections.length) {
            request = this.generateRequestWithParams(request, httpParamInjections);
        }
        return next.handle(request);
    };
    /**
     * @param {?} request
     * @param {?} httpParamInjections
     * @return {?}
     */
    HttpParamInterceptor.prototype.generateRequestWithParams = /**
     * @param {?} request
     * @param {?} httpParamInjections
     * @return {?}
     */
    function (request, httpParamInjections) {
        var _this = this;
        /** @type {?} */
        var params = request.params;
        httpParamInjections
            .forEach((/**
         * @param {?} paramInjection
         * @return {?}
         */
        function (paramInjection) {
            _this.log.debug("Injecting URL '" + request.url + "' with parameter value (name: '" + paramInjection.paramName + "')");
            params = params.set(paramInjection.paramName, paramInjection.paramValue);
        }));
        return request.clone({ params: params });
    };
    HttpParamInterceptor.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    HttpParamInterceptor.ctorParameters = function () { return [
        { type: HttpConfig },
        { type: Logger }
    ]; };
    /** @nocollapse */ HttpParamInterceptor.ngInjectableDef = i0.defineInjectable({ factory: function HttpParamInterceptor_Factory() { return new HttpParamInterceptor(i0.inject(i1.HttpConfig), i0.inject(i2.Logger)); }, token: HttpParamInterceptor, providedIn: "root" });
    return HttpParamInterceptor;
}());
export { HttpParamInterceptor };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HttpParamInterceptor.prototype.config;
    /**
     * @type {?}
     * @private
     */
    HttpParamInterceptor.prototype.log;
}
