// @tss/sdk/esm5/lib/http/http.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { AutoConfig } from '../core/auto-config.decorator';
import { isRouteMatch } from './is-route-match.function';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing global data caching settings.
 */
var HttpConfig = /** @class */ (function () {
    function HttpConfig() {
        this.paramInjectionsEnabled = true;
        this.paramInjectionDefitions = [];
    }
    Object.defineProperty(HttpConfig.prototype, "hasInjections", {
        get: /**
         * @return {?}
         */
        function () {
            return Boolean(this.paramInjectionDefitions.length);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} route
     * @param {?} paramName
     * @param {?} paramValue
     * @return {?}
     */
    HttpConfig.prototype.addOrUpdateParameterInjection = /**
     * @param {?} route
     * @param {?} paramName
     * @param {?} paramValue
     * @return {?}
     */
    function (route, paramName, paramValue) {
        /** @type {?} */
        var definition = {
            route: route,
            paramName: paramName,
            paramValue: paramValue
        };
        /** @type {?} */
        var existingDefinition = this.paramInjectionDefitions.find((/**
         * @param {?} def
         * @return {?}
         */
        function (def) { return def.route === route && def.paramName === paramName; }));
        if (existingDefinition) {
            Object.assign(existingDefinition, definition);
            return existingDefinition;
        }
        else {
            this.paramInjectionDefitions.push(definition);
            return definition;
        }
    };
    /**
     * @param {?} route
     * @param {?} paramName
     * @param {?} paramValue
     * @return {?}
     */
    HttpConfig.prototype.addParameterInjection = /**
     * @param {?} route
     * @param {?} paramName
     * @param {?} paramValue
     * @return {?}
     */
    function (route, paramName, paramValue) {
        /** @type {?} */
        var definition = {
            route: route,
            paramName: paramName,
            paramValue: paramValue
        };
        this.paramInjectionDefitions.push(definition);
        return definition;
    };
    /**
     * @param {?} route
     * @return {?}
     */
    HttpConfig.prototype.getInjectionDefinitions = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return this.paramInjectionDefitions.filter((/**
         * @param {?} def
         * @return {?}
         */
        function (def) { return isRouteMatch(route, def.route); }));
    };
    /**
     * @param {?} route
     * @return {?}
     */
    HttpConfig.prototype.getFirstInjectionDefinition = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return this.paramInjectionDefitions.find((/**
         * @param {?} def
         * @return {?}
         */
        function (def) { return isRouteMatch(route, def.route); }));
    };
    HttpConfig.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ HttpConfig.ngInjectableDef = i0.defineInjectable({ factory: function HttpConfig_Factory() { return new HttpConfig(); }, token: HttpConfig, providedIn: "root" });
    /**
     * Configuration class for storing global data caching settings.
     */
    HttpConfig = tslib_1.__decorate([
        AutoConfig('HttpConfig')
    ], HttpConfig);
    return HttpConfig;
}());
export { HttpConfig };
if (false) {
    /** @type {?} */
    HttpConfig.prototype.paramInjectionsEnabled;
    /** @type {?} */
    HttpConfig.prototype.paramInjectionDefitions;
}
