// @tss/sdk/esm5/lib/http/http.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { CommonModule } from '@angular/common';
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
// TSS
import { TssCoreModule } from '../core/core.module';
import { HttpParamInterceptor } from './http-param.interceptor';
var TssHttpModule = /** @class */ (function () {
    function TssHttpModule() {
    }
    TssHttpModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        HttpClientModule,
                        TssCoreModule,
                    ],
                    declarations: [],
                    exports: [],
                    providers: [
                        {
                            provide: HTTP_INTERCEPTORS,
                            useClass: HttpParamInterceptor,
                            multi: true
                        }
                    ],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                },] }
    ];
    return TssHttpModule;
}());
export { TssHttpModule };
