// @tss/sdk/esm5/lib/geometry/point/get-distance-between-points.function.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { toSinglePartPointArray } from './to-single-part-point.function';
import { EARTH_RADIUS_IN_KILOMETERS } from '../geometry.constants';
import { degreesToRadians } from '../geometry/degrees-to-radians.function';
import { convertLengthUnits } from '../../core/units/convert-length-units.function';
import { LengthUnit } from '../../core/units/length-unit';
/**
 * @param {?} point1
 * @param {?} point2
 * @return {?}
 */
export function getDistanceBetweenPoints(point1, point2) {
    /** @type {?} */
    var coordinate1 = toSinglePartPointArray(point1);
    /** @type {?} */
    var coordinate2 = toSinglePartPointArray(point2);
    /** @type {?} */
    var x1 = coordinate1[0];
    /** @type {?} */
    var y1 = coordinate1[1];
    /** @type {?} */
    var x2 = coordinate2[0];
    /** @type {?} */
    var y2 = coordinate2[1];
    return getDistanceBetweenPointValues(x1, y1, x2, y2);
}
/**
 * @param {?} x1
 * @param {?} y1
 * @param {?} x2
 * @param {?} y2
 * @return {?}
 */
export function getDistanceBetweenPointValues(x1, y1, x2, y2) {
    // NOTE: This is using the pythagorean theorem (a^2 + b^2 = c^2).
    /** @type {?} */
    var xLength = Math.pow(x2 - x1, 2);
    /** @type {?} */
    var yLength = Math.pow(y2 - y1, 2);
    /** @type {?} */
    var distance = Math.sqrt(xLength + yLength);
    if (isNaN(distance)) {
        throw new Error("Unable to calculate distance between points (" + x1 + "," + x2 + " and " + y1 + "," + y2 + "). Invalid or missing data was provided");
    }
    return distance;
}
/**
 * @param {?} lat1
 * @param {?} lng1
 * @param {?} lat2
 * @param {?} lng2
 * @return {?}
 */
export function getKilometersBetweenLatLng(lat1, lng1, lat2, lng2) {
    // Modified from: https://stackoverflow.com/questions/365826/calculate-distance-between-2-gps-coordinates - RG.
    /** @type {?} */
    var dLat = degreesToRadians(lat2 - lat1);
    /** @type {?} */
    var dLon = degreesToRadians(lng2 - lng1);
    lat1 = degreesToRadians(lat1);
    lat2 = degreesToRadians(lat2);
    /** @type {?} */
    var a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
        Math.sin(dLon / 2) * Math.sin(dLon / 2) * Math.cos(lat1) * Math.cos(lat2);
    /** @type {?} */
    var circumference = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return EARTH_RADIUS_IN_KILOMETERS * circumference;
}
/**
 * @param {?} lat1
 * @param {?} lng1
 * @param {?} lat2
 * @param {?} lng2
 * @return {?}
 */
export function getMetersBetweenLatLng(lat1, lng1, lat2, lng2) {
    return getDistanceBetweenLatLng(lat1, lng1, lat2, lng2, LengthUnit.Meter);
}
/**
 * @param {?} lat1
 * @param {?} lng1
 * @param {?} lat2
 * @param {?} lng2
 * @return {?}
 */
export function getFeetBetweenLatLng(lat1, lng1, lat2, lng2) {
    return getDistanceBetweenLatLng(lat1, lng1, lat2, lng2, LengthUnit.Foot);
}
/**
 * @param {?} lat1
 * @param {?} lng1
 * @param {?} lat2
 * @param {?} lng2
 * @return {?}
 */
export function getMilesBetweenLatLng(lat1, lng1, lat2, lng2) {
    return getDistanceBetweenLatLng(lat1, lng1, lat2, lng2, LengthUnit.Mile);
}
/**
 * @param {?} lat1
 * @param {?} lon1
 * @param {?} lat2
 * @param {?} lon2
 * @param {?} outputUnit
 * @return {?}
 */
export function getDistanceBetweenLatLng(lat1, lon1, lat2, lon2, outputUnit) {
    /** @type {?} */
    var distanceInKilometers = getKilometersBetweenLatLng(lat1, lon1, lat2, lon2);
    return convertLengthUnits(distanceInKilometers, LengthUnit.Kilometer, outputUnit);
}
