// @tss/sdk/esm5/lib/geolocation/geolocation.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { AutoConfig } from '../core/auto-config.decorator';
import { DEFAULT_POINT_EQUALITY_PRECISION } from '../core/equality/equality.constants';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing geolocation-related settings.
 */
var GeolocationConfig = /** @class */ (function () {
    function GeolocationConfig() {
        this.gpsDisplayDecimalPlaces = 3;
        this.gpsWarnableAccuracyMeters = 5;
        this.gpsRequiredAccuracyMeters = 10;
        this.gpsWarnableDilutionOfPrecision = 2;
        this.gpsRequiredDilutionOfPrecision = 5;
        this.gpsEnableHighAccuracy = true;
        this.gpsNoAccuracyReading = '--';
        this.allowGpsMocking = true;
        this.emitDuplicateGpsPositions = false;
        this.gpsPositionDuplicatePrecision = DEFAULT_POINT_EQUALITY_PRECISION;
        this.gpsAntennaHeight = 0;
        this.adjustForGpsAntennaHeight = true;
        this.calculateMissingPositionValues = true;
        this.geolocationGeolocatorName = 'Built-in (default)';
        this.geolocationGeolocatorDescription = "Uses the device's built-in GPS sensor within the hardware. " +
            "Devices without a GPS sensor will defer the location retrieval to the Inernet Service Provider (ISP).";
    }
    GeolocationConfig.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ GeolocationConfig.ngInjectableDef = i0.defineInjectable({ factory: function GeolocationConfig_Factory() { return new GeolocationConfig(); }, token: GeolocationConfig, providedIn: "root" });
    /**
     * Configuration class for storing geolocation-related settings.
     */
    GeolocationConfig = tslib_1.__decorate([
        AutoConfig('GeolocationConfig')
    ], GeolocationConfig);
    return GeolocationConfig;
}());
export { GeolocationConfig };
if (false) {
    /** @type {?} */
    GeolocationConfig.prototype.gpsDisplayDecimalPlaces;
    /** @type {?} */
    GeolocationConfig.prototype.gpsWarnableAccuracyMeters;
    /** @type {?} */
    GeolocationConfig.prototype.gpsRequiredAccuracyMeters;
    /** @type {?} */
    GeolocationConfig.prototype.gpsWarnableDilutionOfPrecision;
    /** @type {?} */
    GeolocationConfig.prototype.gpsRequiredDilutionOfPrecision;
    /** @type {?} */
    GeolocationConfig.prototype.gpsEnableHighAccuracy;
    /** @type {?} */
    GeolocationConfig.prototype.gpsNoAccuracyReading;
    /** @type {?} */
    GeolocationConfig.prototype.allowGpsMocking;
    /** @type {?} */
    GeolocationConfig.prototype.emitDuplicateGpsPositions;
    /** @type {?} */
    GeolocationConfig.prototype.gpsPositionDuplicatePrecision;
    /** @type {?} */
    GeolocationConfig.prototype.gpsAntennaHeight;
    /** @type {?} */
    GeolocationConfig.prototype.adjustForGpsAntennaHeight;
    /** @type {?} */
    GeolocationConfig.prototype.calculateMissingPositionValues;
    /** @type {?} */
    GeolocationConfig.prototype.geolocationGeolocatorName;
    /** @type {?} */
    GeolocationConfig.prototype.geolocationGeolocatorDescription;
}
