// @tss/sdk/esm2015/lib/app/settings/app-setting.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
// TSS.
import { BaseResourceService } from '../../core/resource/base-resource.service';
import { ALL_APP_SETTINGS_QUERY_ID } from './app-setting';
import { AppConfig } from '../app.config';
import { join } from '../../core/url/join.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "../app.config";
export class AppSettingService extends BaseResourceService {
    /**
     * @param {?} http
     * @param {?} config
     */
    constructor(http, config) {
        super(http, {
            url: join(config.serviceUrl, config.appSettingsEndpoint),
            defaultRouteTokens: {
                appId: config.appId
            }
        });
        this.http = http;
    }
    /**
     * Retrieves a list of application settings.
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    queryAllApps(routeTokens, params) {
        routeTokens = routeTokens || {};
        routeTokens.appId = ALL_APP_SETTINGS_QUERY_ID;
        return super.query(routeTokens, params);
    }
}
AppSettingService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AppSettingService.ctorParameters = () => [
    { type: HttpClient },
    { type: AppConfig }
];
/** @nocollapse */ AppSettingService.ngInjectableDef = i0.defineInjectable({ factory: function AppSettingService_Factory() { return new AppSettingService(i0.inject(i1.HttpClient), i0.inject(i2.AppConfig)); }, token: AppSettingService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AppSettingService.prototype.http;
}
