// @tss/sdk/esm5/lib/core/storage/base-storage.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { getByteSize } from './get-byte-size.function';
import { formatByteSize } from './format-byte-size.function';
import * as i0 from "@angular/core";
/**
 * @abstract
 */
var BaseStorage = /** @class */ (function () {
    function BaseStorage() {
        this.keyPrefix = '';
        this.keySuffix = '';
    }
    /**
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.has = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return Boolean(this.get(key));
    };
    /**
     * Returns the number of bytes for this storage key.
     */
    /**
     * Returns the number of bytes for this storage key.
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.size = /**
     * Returns the number of bytes for this storage key.
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var value = this.get(key);
        return getByteSize(value) + getByteSize(key);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.sizeFormatted = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return formatByteSize(this.size(key));
    };
    /**
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.getObject = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var value = this.get(key);
        return value ? JSON.parse(value) : null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    BaseStorage.prototype.setObject = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.set(key, JSON.stringify(value));
    };
    /**
     * @protected
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.formatKey = /**
     * @protected
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return "" + this.keyPrefix + key + this.keySuffix;
    };
    BaseStorage.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ BaseStorage.ngInjectableDef = i0.defineInjectable({ factory: function BaseStorage_Factory() { return new BaseStorage(); }, token: BaseStorage, providedIn: "root" });
    return BaseStorage;
}());
export { BaseStorage };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseStorage.prototype.keyPrefix;
    /**
     * @type {?}
     * @protected
     */
    BaseStorage.prototype.keySuffix;
    /** @type {?} */
    BaseStorage.prototype.length;
    /**
     * @abstract
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.get = function (key) { };
    /**
     * @abstract
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    BaseStorage.prototype.set = function (key, value) { };
    /**
     * @abstract
     * @param {?} key
     * @return {?}
     */
    BaseStorage.prototype.remove = function (key) { };
    /**
     * @abstract
     * @return {?}
     */
    BaseStorage.prototype.clear = function () { };
}
