// @tss/sdk/esm2015/lib/core/logging/logger.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
import { LogLevel, typeToLevel } from './log-level';
import { CoreConfig } from '../core.config';
import { format } from '../formatting/format.function';
import * as i0 from "@angular/core";
import * as i1 from "../core.config";
/**
 * Console logger.
 */
export class Logger {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.config = config;
        this.logs = [];
    }
    /**
     * @return {?}
     */
    get level() {
        return this.config.logLevel;
    }
    /**
     * @return {?}
     */
    get enabled() {
        return this.config.logEnabled;
    }
    /**
     * @return {?}
     */
    get isTrace() {
        return this.level >= LogLevel.Trace;
    }
    /**
     * @return {?}
     */
    get isDebug() {
        return this.level >= LogLevel.Debug;
    }
    /**
     * @return {?}
     */
    get isWarn() {
        return this.level >= LogLevel.Warn;
    }
    /**
     * @return {?}
     */
    get isError() {
        return this.level >= LogLevel.Error;
    }
    /**
     * @return {?}
     */
    get isFatal() {
        return this.level >= LogLevel.Fatal;
    }
    /**
     * @return {?}
     */
    get traceCount() {
        return this.count('trace');
    }
    /**
     * @return {?}
     */
    get debugCount() {
        return this.count('debug');
    }
    /**
     * @return {?}
     */
    get warnCount() {
        return this.count('warn');
    }
    /**
     * @return {?}
     */
    get errorCount() {
        return this.count('error');
    }
    /**
     * @return {?}
     */
    get fatalCount() {
        return this.count('fatal');
    }
    /**
     * @param {?} logType
     * @return {?}
     */
    count(logType) {
        return this.logsByType(logType).length;
    }
    /**
     * @param {?} logType
     * @return {?}
     */
    logsByType(logType) {
        return this.logsByLevel(typeToLevel(logType));
    }
    /**
     * @param {?} logLevel
     * @return {?}
     */
    logsByLevel(logLevel) {
        return this.logs.filter((/**
         * @param {?} l
         * @return {?}
         */
        l => l.level === logLevel));
    }
    /**
     * @param {?} log
     * @return {?}
     */
    logEntry(log) {
        if (this.config.logEnabled && log.level >= this.config.logLevel) {
            this.logs.push(log);
            if (this.isDebug) {
                console.log(`${LogLevel[log.level]}: ${log.msg} ${log.error || ''}`);
            }
        }
        return log;
    }
    /**
     * @param {?} msg
     * @param {?} level
     * @param {?=} error
     * @return {?}
     */
    log(msg, level, error = null) {
        return this.logEntry({
            msg: format(msg, { error }),
            level,
            error,
            created: new Date()
        });
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    trace(msg) {
        return this.log(msg, LogLevel.Trace);
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    debug(msg) {
        return this.log(msg, LogLevel.Debug);
    }
    /**
     * @param {?} msg
     * @return {?}
     */
    info(msg) {
        return this.log(msg, LogLevel.Info);
    }
    /**
     * @param {?} msg
     * @param {?=} error
     * @return {?}
     */
    warn(msg, error = null) {
        return this.log(msg, LogLevel.Warn, error);
    }
    /**
     * @param {?} msg
     * @param {?=} error
     * @return {?}
     */
    error(msg, error = null) {
        return this.log(msg, LogLevel.Error, error);
    }
    /**
     * @param {?} msg
     * @param {?=} error
     * @return {?}
     */
    fatal(msg, error = null) {
        return this.log(msg, LogLevel.Fatal, error);
    }
    /**
     * @return {?}
     */
    clear() {
        this.logs.length = 0;
    }
}
Logger.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
Logger.ctorParameters = () => [
    { type: CoreConfig }
];
/** @nocollapse */ Logger.ngInjectableDef = i0.defineInjectable({ factory: function Logger_Factory() { return new Logger(i0.inject(i1.CoreConfig)); }, token: Logger, providedIn: "root" });
if (false) {
    /** @type {?} */
    Logger.prototype.logs;
    /**
     * @type {?}
     * @private
     */
    Logger.prototype.config;
}
