// @tss/sdk/esm5/lib/core/media/media.constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var BASE_64_METADATA_REGEX = /data:([a-zA-Z0-9]+\/[a-zA-Z0-9-.+]+).*,.*/;
/** @type {?} */
export var JSON_MIME_TYPE = 'application/json';
// Text files.
/** @type {?} */
export var TEXT_MIME_TYPE = 'text/plain';
/** @type {?} */
export var TEXT_FILE_EXTENSION = '.txt';
// Audio.
/** @type {?} */
export var AUDIO_WEBM_MIME_TYPE = 'audio/webm';
/** @type {?} */
export var AUDIO_MP3_MIME_TYPE = 'audio/mpeg-3';
/** @type {?} */
export var AUDIO_MP3_FILE_EXTENSION = '.mp3';
// Video.
/** @type {?} */
export var VIDEO_WEBM_MIME_TYPE = 'video/webm;codecs=h264';
/** @type {?} */
export var VIDEO_MP4_MIME_TYPE = 'video/mp4';
/** @type {?} */
export var VIDEO_MP4_FILE_EXTENSION = '.mp4';
// Images.
/** @type {?} */
export var IMAGE_PNG_FILE_EXTENSION = '.png';
/** @type {?} */
export var IMAGE_PNG_MIME_TYPE = 'image/png';
/** @type {?} */
export var IMAGE_JPEG_FILE_EXTENSION = '.jpeg';
/** @type {?} */
export var IMAGE_JPEG_MIME_TYPE = 'image/jpeg';
