// @tss/sdk/esm2015/lib/cache/cache.config.js
import * as tslib_1 from "tslib";
var CacheConfig_1;
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { AutoConfig } from '../core/auto-config.decorator';
import { isRouteMatch } from '../http/is-route-match.function';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing global data caching settings.
 */
let CacheConfig = CacheConfig_1 = /**
 * Configuration class for storing global data caching settings.
 */
class CacheConfig {
    constructor() {
        /**
         * Determines whether caching should be used or not. When disabled, all
         * caching-related functionality is ignored (not used).
         */
        this.cacheEnabled = false;
        /**
         * List of configurations that describe what to cache and how to cache it.
         * The main parts of this configuration include the "route" (how to match what
         * to cache) and the "strategy" (whether to look for content in the cache first
         * or try to make the HTTP request first).
         */
        this.cacheDefitions = [];
        /**
         * When set, this strategy will trump all strategies defined in the "Cache Defintions".
         * WARNING: This is a global setting and can have adverse affects when used improperly.
         * The main use case for this setting is to allow for the global cache strategy of
         * 'cacheOnly', which will allow an application to force an "offline" state (forcing
         * requests to pull from cache only).
         */
        this.forcedGlobalCacheStrategy = null;
    }
    /**
     * @param {?} config
     * @return {?}
     */
    static create(config) {
        return Object.assign(new CacheConfig_1(), config);
    }
    /**
     * @param {...?} defs
     * @return {?}
     */
    static enable(...defs) {
        return CacheConfig_1.create({
            cacheEnabled: true,
            cacheDefitions: defs,
        });
    }
    /**
     * @param {?} route
     * @return {?}
     */
    getCacheDefinition(route) {
        return this.cacheDefitions.find((/**
         * @param {?} def
         * @return {?}
         */
        def => isRouteMatch(route, def.route)));
    }
};
CacheConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ CacheConfig.ngInjectableDef = i0.defineInjectable({ factory: function CacheConfig_Factory() { return new CacheConfig(); }, token: CacheConfig, providedIn: "root" });
/**
 * Configuration class for storing global data caching settings.
 */
CacheConfig = CacheConfig_1 = tslib_1.__decorate([
    AutoConfig('CacheConfig')
], CacheConfig);
export { CacheConfig };
if (false) {
    /**
     * Determines whether caching should be used or not. When disabled, all
     * caching-related functionality is ignored (not used).
     * @type {?}
     */
    CacheConfig.prototype.cacheEnabled;
    /**
     * List of configurations that describe what to cache and how to cache it.
     * The main parts of this configuration include the "route" (how to match what
     * to cache) and the "strategy" (whether to look for content in the cache first
     * or try to make the HTTP request first).
     * @type {?}
     */
    CacheConfig.prototype.cacheDefitions;
    /**
     * When set, this strategy will trump all strategies defined in the "Cache Defintions".
     * WARNING: This is a global setting and can have adverse affects when used improperly.
     * The main use case for this setting is to allow for the global cache strategy of
     * 'cacheOnly', which will allow an application to force an "offline" state (forcing
     * requests to pull from cache only).
     * @type {?}
     */
    CacheConfig.prototype.forcedGlobalCacheStrategy;
}
