// @tss/sdk/esm2015/lib/cache/cache.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { CommonModule } from '@angular/common';
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
// TSS
import { CacheService } from './cache.service';
import { CacheInterceptor } from './cache.interceptor';
import { CacheConfig } from './cache.config';
import { PersistentCacheStorage } from './cache.storage';
import { TssCoreModule } from '../core/core.module';
export class TssCacheModule {
}
TssCacheModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    HttpClientModule,
                    TssCoreModule,
                ],
                declarations: [],
                exports: [],
                providers: [
                    CacheConfig,
                    PersistentCacheStorage,
                    CacheService,
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: CacheInterceptor,
                        multi: true
                    }
                ],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            },] }
];
