// @tss/sdk/esm2015/lib/lrs/events/event-feature.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TSS.
import { LrsFeature } from '../lrs-feature';
import { toFeature } from '../feature';
import { WGS_84_SRID } from '../../geometry/geometry.constants';
import { clone } from '../../core/clone/clone.function';
export class EventFeature extends LrsFeature {
    /**
     * @param {?} feature
     * @param {?} layer
     * @param {?=} spatialReferenceId
     */
    constructor(feature, layer, spatialReferenceId = WGS_84_SRID) {
        super(feature, layer, spatialReferenceId);
        this.routeId = this.getValue(layer.routeIdFieldName);
        this.fromMeasure = this.getValue(layer.fromMeasureFieldName);
        this.toMeasure = this.getValue(layer.toMeasureFieldName);
    }
    /**
     * @param {?} layer
     * @param {?=} properties
     * @param {?=} geometry
     * @return {?}
     */
    static create(layer, properties = {}, geometry = null) {
        /** @type {?} */
        const feature = toFeature(properties, (/** @type {?} */ (geometry)));
        return new EventFeature(feature, layer);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} routeId
     * @return {THIS}
     */
    setRouteId(routeId) {
        (/** @type {?} */ (this)).routeId = routeId;
        return (/** @type {?} */ (this)).setValue(routeId, (/** @type {?} */ (this)).layer.routeIdFieldName);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    setFromMeasure(measure) {
        (/** @type {?} */ (this)).fromMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.fromMeasureFieldName, measure);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} measure
     * @return {THIS}
     */
    setToMeasure(measure) {
        if (!(/** @type {?} */ (this)).isLinear) {
            throw new Error('Cannot set to measure for a point-based feature');
        }
        (/** @type {?} */ (this)).toMeasure = measure;
        return (/** @type {?} */ (this)).setValue((/** @type {?} */ (this)).layer.toMeasureFieldName, measure);
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @return {THIS}
     */
    setMeasureExtent(fromMeasure, toMeasure) {
        return (/** @type {?} */ (this)).setFromMeasure(fromMeasure)
            .setToMeasure(toMeasure);
    }
    /**
     * @return {?}
     */
    clone() {
        /** @type {?} */
        const copyFeature = clone(this.toFeature());
        return new EventFeature(copyFeature, this.layer);
    }
    /**
     * @protected
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    setGeometryTypeProperties() {
        (/** @type {?} */ (this)).isLinear = Boolean((/** @type {?} */ (this)).layer.toMeasureFieldName);
        (/** @type {?} */ (this)).isPoint = !(/** @type {?} */ (this)).isLinear;
        (/** @type {?} */ (this)).isPolygon = false;
        return (/** @type {?} */ (this));
    }
}
if (false) {
    /** @type {?} */
    EventFeature.prototype.layer;
}
