// @tss/sdk/esm5/lib/esri/base-esri.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LrsWhereClauseBuilder } from '../lrs/lrs-where-clause-builder';
/**
 * @abstract
 */
var /**
 * @abstract
 */
BaseEsriService = /** @class */ (function () {
    function BaseEsriService(http, config, dataConfig, log) {
        this.http = http;
        this.config = config;
        this.dataConfig = dataConfig;
        this.log = log;
    }
    /**
     * @protected
     * @param {?} payload
     * @return {?}
     */
    BaseEsriService.prototype.shouldUsePostMethod = /**
     * @protected
     * @param {?} payload
     * @return {?}
     */
    function (payload) {
        return this.config.arcgisQueryMethodStrategy === 'force-post'
            || JSON.stringify(payload || '').length > this.config.arcgisMaxWhereLengthGetThreshold;
    };
    /**
     * @protected
     * @template T
     * @param {?} response
     * @param {?} url
     * @return {?}
     */
    BaseEsriService.prototype.handleEsriResponseError = /**
     * @protected
     * @template T
     * @param {?} response
     * @param {?} url
     * @return {?}
     */
    function (response, url) {
        /** @type {?} */
        var error = (/** @type {?} */ (response));
        if (error && error.error) {
            /** @type {?} */
            var returnError = new Error("Failed to query map service: " + error.error.message);
            this.log.error("Failed to execute map service request: " + url, returnError);
            throw returnError;
        }
    };
    /**
     * @protected
     * @return {?}
     */
    BaseEsriService.prototype.getWhereBuilder = /**
     * @protected
     * @return {?}
     */
    function () {
        return new LrsWhereClauseBuilder({ quoteFields: !this.dataConfig.skipColumnQuotes });
    };
    return BaseEsriService;
}());
/**
 * @abstract
 */
export { BaseEsriService };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.http;
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.dataConfig;
    /**
     * @type {?}
     * @protected
     */
    BaseEsriService.prototype.log;
}
