// @tss/sdk/esm5/lib/esri/esri-map-service-request.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpHeaders, HttpParams } from '@angular/common/http';
import { toFormDataString } from '../http/to-form-data-string.function';
import { isValue } from '../core/type-check/is-value.function';
/**
 * @record
 */
export function AngularHttpOption() { }
if (false) {
    /** @type {?|undefined} */
    AngularHttpOption.prototype.headers;
    /** @type {?|undefined} */
    AngularHttpOption.prototype.observe;
    /** @type {?|undefined} */
    AngularHttpOption.prototype.params;
    /** @type {?|undefined} */
    AngularHttpOption.prototype.reportProgress;
    /** @type {?|undefined} */
    AngularHttpOption.prototype.responseType;
    /** @type {?|undefined} */
    AngularHttpOption.prototype.withCredentials;
}
/**
 * @param {?=} params
 * @param {?=} format
 * @return {?}
 */
export function getEsriMapServiceHttOptions(params, format) {
    if (format === void 0) { format = 'json'; }
    /** @type {?} */
    var headers = new HttpHeaders();
    params = params || new HttpParams();
    if (format) {
        params = params.set('f', format);
    }
    headers.append('Content-Type', 'application/json');
    headers.append('Access-Control-Allow-Origin', '*');
    /** @type {?} */
    var httpOptions = {
        withCredentials: true,
        headers: headers,
        params: params,
    };
    return httpOptions;
}
/**
 * @param {?} httpOption
 * @param {?} params
 * @return {?}
 */
export function addParamsToHttpOptions(httpOption, params) {
    Object
        .keys(params)
        .filter((/**
     * @param {?} key
     * @return {?}
     */
    function (key) { return isValue(params[key]); }))
        .forEach((/**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var value = params[key];
        httpOption.params = httpOption.params.set(key, toFormDataString(value, false));
    }));
}
