// @tss/sdk/esm5/lib/esri/esri-map-server.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map, tap } from 'rxjs/operators';
// TSS.
import { Logger } from '../core/logging/logger.service';
import { NetworkFeature } from '../lrs/networks/network-feature';
import { EventFeatureCollection } from '../lrs/events/event-feature-collection';
import { NetworkFeatureCollection } from '../lrs/networks/network-feature-collection';
import { CriteriaOperator } from '../maps/where-clause-builder';
import { EsriConfig } from './esri-config';
import { getEsriMapServiceHttOptions, addParamsToHttpOptions } from './esri-map-service-request';
import { toFieldName } from '../maps/fields/layer-field';
import { DataSourceConfig } from '../data-source/data-source-config';
import { toFormData } from '../http/to-form-data.function';
import { MAP_LAYERS_SERVICE_ENDPOINT, LRS_LAYERS_SERVICE_ENDPOINT, MAP_LAYER_QUERY_SERVICE_ENDPOINT, LRS_SERVER_META_ENDPOINT, LRS_SERVER_APPLY_EDITS_ENDPOINT, LRS_GEOMETRY_TO_MEASURE, MAP_LAYER_SERVICE_ENDPOINT } from './esri.constants';
import { getGeometryToMeasureOptions } from './geometry-to-measure/get-geometry-to-measure-options.function';
import { getMapQueryOptions } from './query/get-map-query-options.function';
import { toHttpQueryParams } from './query/to-http-query-params.function';
import { tryStripGdbVersion } from './try-strip-gdb-version.function';
import { toLayerId } from '../maps/to-layer-id.function';
import { BaseEsriService } from './base-esri.service';
import { joinAndFomatWithLayer } from './join-and-format-with-layer.function';
import { join } from '../core/url/join.function';
import { format } from '../core/formatting/format.function';
import { FeatureClassLayer } from '../maps/feature-class-layer';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./esri-config";
import * as i3 from "../data-source/data-source-config";
import * as i4 from "../core/logging/logger.service";
var EsriMapServerService = /** @class */ (function (_super) {
    tslib_1.__extends(EsriMapServerService, _super);
    function EsriMapServerService(http, config, dataConfig, log) {
        return _super.call(this, http, config, dataConfig, log) || this;
    }
    /**
     * Retrieves the map service metadata for all layers.
     */
    /**
     * Retrieves the map service metadata for all layers.
     * @param {?} serviceUrl
     * @return {?}
     */
    EsriMapServerService.prototype.getMapLayers = /**
     * Retrieves the map service metadata for all layers.
     * @param {?} serviceUrl
     * @return {?}
     */
    function (serviceUrl) {
        var _this = this;
        /** @type {?} */
        var url = join(serviceUrl, MAP_LAYERS_SERVICE_ENDPOINT);
        return this.http
            .get(url, getEsriMapServiceHttOptions())
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })));
    };
    /**
     * Retrieves the map service metadata for a given layer (by it's ID).
     */
    /**
     * Retrieves the map service metadata for a given layer (by it's ID).
     * @param {?} serviceUrl
     * @param {?} layerId
     * @return {?}
     */
    EsriMapServerService.prototype.getMapLayer = /**
     * Retrieves the map service metadata for a given layer (by it's ID).
     * @param {?} serviceUrl
     * @param {?} layerId
     * @return {?}
     */
    function (serviceUrl, layerId) {
        var _this = this;
        /** @type {?} */
        var url = joinAndFomatWithLayer(layerId, {}, serviceUrl, MAP_LAYER_SERVICE_ENDPOINT);
        return this.http
            .get(url, getEsriMapServiceHttOptions())
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })), map((/**
         * @param {?} mapLayerMetadata
         * @return {?}
         */
        function (mapLayerMetadata) { return new FeatureClassLayer(mapLayerMetadata); })));
    };
    /**
     * Retrieves the LRS Layers Metadata information from the map service
     */
    /**
     * Retrieves the LRS Layers Metadata information from the map service
     * @param {?} serviceUrl
     * @return {?}
     */
    EsriMapServerService.prototype.getLrsLayers = /**
     * Retrieves the LRS Layers Metadata information from the map service
     * @param {?} serviceUrl
     * @return {?}
     */
    function (serviceUrl) {
        var _this = this;
        /** @type {?} */
        var url = join(serviceUrl, LRS_LAYERS_SERVICE_ENDPOINT);
        return this.http
            .get(url, getEsriMapServiceHttOptions())
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })));
    };
    /**
     * Retrieves the LRS Server Metadata information from the map service
     */
    /**
     * Retrieves the LRS Server Metadata information from the map service
     * @param {?} serviceUrl
     * @return {?}
     */
    EsriMapServerService.prototype.getLrsServer = /**
     * Retrieves the LRS Server Metadata information from the map service
     * @param {?} serviceUrl
     * @return {?}
     */
    function (serviceUrl) {
        var _this = this;
        /** @type {?} */
        var url = join(serviceUrl, LRS_SERVER_META_ENDPOINT);
        return this.http
            .get(url, getEsriMapServiceHttOptions())
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?=} lrsId
     * @return {?}
     */
    EsriMapServerService.prototype.getVersions = /**
     * @param {?} serviceUrl
     * @param {?=} lrsId
     * @return {?}
     */
    function (serviceUrl, lrsId) {
        return this.getLrsServer(serviceUrl)
            .pipe(map((/**
         * @param {?} metadata
         * @return {?}
         */
        function (metadata) {
            /** @type {?} */
            var lrs = metadata ? metadata.lrs[0] : null;
            if (!lrs) {
                throw new Error("No LRS infos found");
            }
            if (lrsId) {
                lrs = metadata.lrs.find((/**
                 * @param {?} lrsInfo
                 * @return {?}
                 */
                function (lrsInfo) { return lrsInfo.id === lrsId; }));
                if (!lrs) {
                    throw new Error("No LRS info found with ID '" + lrsId + "'");
                }
            }
            return lrs.versions;
        })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} networkLayerId
     * @param {?} routeId
     * @param {?} x
     * @param {?} y
     * @param {?=} viewDate
     * @param {?=} inSpatialReferenceId
     * @param {?=} outputSpatialReferenceId
     * @return {?}
     */
    EsriMapServerService.prototype.singleGeometryToMeasure = /**
     * @param {?} serviceUrl
     * @param {?} networkLayerId
     * @param {?} routeId
     * @param {?} x
     * @param {?} y
     * @param {?=} viewDate
     * @param {?=} inSpatialReferenceId
     * @param {?=} outputSpatialReferenceId
     * @return {?}
     */
    function (serviceUrl, networkLayerId, routeId, x, y, viewDate, inSpatialReferenceId, outputSpatialReferenceId) {
        if (inSpatialReferenceId === void 0) { inSpatialReferenceId = 4326; }
        if (outputSpatialReferenceId === void 0) { outputSpatialReferenceId = 4326; }
        /** @type {?} */
        var locations = [
            { routeId: routeId, geometry: { x: x, y: y } }
        ];
        return this.geometryToMeasure(serviceUrl, networkLayerId, {
            locations: locations,
            temporalViewDate: viewDate,
            inSR: inSpatialReferenceId,
            outSR: outputSpatialReferenceId,
        })
            .pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            /** @type {?} */
            var match = { m: null, x: null, y: null, z: null };
            if (result.locations.length && result.locations[0].results.length) {
                var _a = tslib_1.__read(result.locations[0].results, 1), found = _a[0];
                match.m = found.measure;
                match.x = found.geometry.x;
                match.y = found.geometry.y;
            }
            return match;
        })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} networkLayerId
     * @param {?} options
     * @return {?}
     */
    EsriMapServerService.prototype.geometryToMeasure = /**
     * @param {?} serviceUrl
     * @param {?} networkLayerId
     * @param {?} options
     * @return {?}
     */
    function (serviceUrl, networkLayerId, options) {
        var _this = this;
        /** @type {?} */
        var endpoint = format(LRS_GEOMETRY_TO_MEASURE, { layerId: networkLayerId });
        /** @type {?} */
        var url = join(serviceUrl, endpoint);
        /** @type {?} */
        var httpOption = getEsriMapServiceHttOptions();
        /** @type {?} */
        var serviceOptions = getGeometryToMeasureOptions(options);
        /** @type {?} */
        var useGetMethod = !this.shouldUsePostMethod(serviceOptions);
        if (useGetMethod) {
            addParamsToHttpOptions(httpOption, serviceOptions);
        }
        return (useGetMethod
            ? this.http.get(url, httpOption)
            : this.http.post(url, toFormData(serviceOptions), httpOption))
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} edits
     * @param {?=} gdbVersion
     * @return {?}
     */
    EsriMapServerService.prototype.applyEdits = /**
     * @param {?} serviceUrl
     * @param {?} edits
     * @param {?=} gdbVersion
     * @return {?}
     */
    function (serviceUrl, edits, gdbVersion) {
        var _this = this;
        /** @type {?} */
        var url = join(serviceUrl, LRS_SERVER_APPLY_EDITS_ENDPOINT);
        /** @type {?} */
        var options = getEsriMapServiceHttOptions();
        /** @type {?} */
        var useGetMethod = !this.shouldUsePostMethod(edits);
        if (useGetMethod) {
            options.params = options.params
                .set('edits', JSON.stringify(edits))
                .set('gdbVersion', gdbVersion);
        }
        return (useGetMethod
            ? this.http.get(url, options)
            : this.http.post(url, toFormData({ edits: edits, gdbVersion: gdbVersion }), options))
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })));
    };
    /**
     * @template T
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} field
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.getUniqueValues = /**
     * @template T
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} field
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, field, options) {
        /** @type {?} */
        var fieldName = toFieldName(field);
        /** @type {?} */
        var queryOptions = getMapQueryOptions({
            outFields: fieldName,
            returnDistinctValues: true,
            returnGeometry: false,
        }, options);
        return this.query(serviceUrl, toLayerId(layer), queryOptions)
            .pipe(map((/**
         * @param {?} collection
         * @return {?}
         */
        function (collection) {
            return collection.features.map((/**
             * @param {?} f
             * @return {?}
             */
            function (f) { return f.properties[fieldName]; }));
        })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} whereClause
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.getQueryCount = /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} whereClause
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, whereClause, options) {
        /** @type {?} */
        var queryOptions = getMapQueryOptions({
            returnCountOnly: true,
            returnGeometry: false,
            where: whereClause,
        }, options);
        return this.query(serviceUrl, toLayerId(layer), queryOptions)
            .pipe(map((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            return (result || { count: 0 }).count;
        })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} eventId
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.getEvent = /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} eventId
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, eventId, options) {
        tryStripGdbVersion(layer, options);
        throw new Error('Not implemented yet');
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} routeId
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @param {?=} viewDate
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.queryEvents = /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} routeId
     * @param {?} fromMeasure
     * @param {?} toMeasure
     * @param {?=} viewDate
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, routeId, fromMeasure, toMeasure, viewDate, options) {
        tryStripGdbVersion(layer, options);
        /** @type {?} */
        var where = this.getWhereBuilder()
            .withEventLayer(layer)
            .withRouteId(routeId)
            .withViewDate(viewDate)
            .betweenMeasures(fromMeasure, toMeasure)
            .toWhereClause();
        /** @type {?} */
        var queryOptions = getMapQueryOptions({ where: where, returnGeometry: false }, options);
        return this
            .query(serviceUrl, layer.id, queryOptions)
            .pipe(map((/**
         * @param {?} featureCollection
         * @return {?}
         */
        function (featureCollection) { return new EventFeatureCollection(featureCollection, layer); })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} searchText
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.searchEventsByText = /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} searchText
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, searchText, options) {
        var _this = this;
        return this.searchByText(serviceUrl, layer.id, searchText, options)
            .pipe(tap((/**
         * @param {?} response
         * @return {?}
         */
        function (response) { return _this.handleEsriResponseError(response, serviceUrl); })), map((/**
         * @param {?} featureCollection
         * @return {?}
         */
        function (featureCollection) { return new EventFeatureCollection(featureCollection, layer); })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} routeId
     * @param {?} layer
     * @param {?=} viewDate
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.getRoute = /**
     * @param {?} serviceUrl
     * @param {?} routeId
     * @param {?} layer
     * @param {?=} viewDate
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, routeId, layer, viewDate, options) {
        tryStripGdbVersion(layer, options);
        /** @type {?} */
        var where = this.getWhereBuilder()
            .withNetworkLayer(layer)
            .where(layer.compositeRouteIdFieldName, CriteriaOperator.Equals, routeId)
            .withViewDate(viewDate)
            .toWhereClause();
        /** @type {?} */
        var queryOptions = getMapQueryOptions({
            where: where
        }, options);
        return this
            .query(serviceUrl, layer.id, queryOptions)
            .pipe(map((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            return x && x.features && x.features.length
                ? new NetworkFeature(x.features[0], layer)
                : null;
        })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} whereClause
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.queryRoutes = /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} whereClause
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, whereClause, options) {
        /** @type {?} */
        var queryOptions = getMapQueryOptions({
            where: whereClause
        }, options);
        return this.query(serviceUrl, layer.id, queryOptions)
            .pipe(map((/**
         * @param {?} featureCollection
         * @return {?}
         */
        function (featureCollection) { return new NetworkFeatureCollection(featureCollection, layer, options.outSR); })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} searchText
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.searchRoutesByText = /**
     * @param {?} serviceUrl
     * @param {?} layer
     * @param {?} searchText
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layer, searchText, options) {
        return this.searchByText(serviceUrl, layer.id, searchText, options)
            .pipe(map((/**
         * @param {?} featureCollection
         * @return {?}
         */
        function (featureCollection) { return new NetworkFeatureCollection(featureCollection, layer); })));
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layerId
     * @param {?} searchText
     * @param {?=} options
     * @return {?}
     */
    EsriMapServerService.prototype.searchByText = /**
     * @param {?} serviceUrl
     * @param {?} layerId
     * @param {?} searchText
     * @param {?=} options
     * @return {?}
     */
    function (serviceUrl, layerId, searchText, options) {
        /** @type {?} */
        var queryOptions = getMapQueryOptions({
            text: searchText
        }, options);
        // The where clause must be deleted because if a where clause is specified it will override the search text.
        delete queryOptions.where;
        return this.query(serviceUrl, layerId, queryOptions);
    };
    /**
     * @param {?} serviceUrl
     * @param {?} layerId
     * @param {?} options
     * @return {?}
     */
    EsriMapServerService.prototype.query = /**
     * @param {?} serviceUrl
     * @param {?} layerId
     * @param {?} options
     * @return {?}
     */
    function (serviceUrl, layerId, options) {
        var _this = this;
        /** @type {?} */
        var url = joinAndFomatWithLayer(layerId, options, serviceUrl, MAP_LAYER_QUERY_SERVICE_ENDPOINT);
        /** @type {?} */
        var httpOption = getEsriMapServiceHttOptions();
        /** @type {?} */
        var usePostMethod = this.shouldUsePostMethod(options.where);
        if (!usePostMethod) {
            httpOption.params = toHttpQueryParams(options, httpOption.params);
        }
        /** @type {?} */
        var source = usePostMethod
            ? this.http.post(url, options, httpOption)
            : this.http.get(url, httpOption);
        return source.pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            _this.handleEsriResponseError(response, url);
            return (/** @type {?} */ (response));
        })));
    };
    EsriMapServerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    EsriMapServerService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: EsriConfig },
        { type: DataSourceConfig },
        { type: Logger }
    ]; };
    /** @nocollapse */ EsriMapServerService.ngInjectableDef = i0.defineInjectable({ factory: function EsriMapServerService_Factory() { return new EsriMapServerService(i0.inject(i1.HttpClient), i0.inject(i2.EsriConfig), i0.inject(i3.DataSourceConfig), i0.inject(i4.Logger)); }, token: EsriMapServerService, providedIn: "root" });
    return EsriMapServerService;
}(BaseEsriService));
export { EsriMapServerService };
