// @tss/sdk/esm2015/lib/core/notify/notify.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable } from '@angular/core';
// TSS.
import { CoreConfig } from '../core.config';
import { NotifyType } from './notify-type';
import { Logger } from '../logging/logger.service';
import { NotificationSubscriptionTracker } from './notification-subscription-tracker';
import { NotificationPromiseTracker } from './notification-promise-tracker';
import { format } from '../formatting/format.function';
import * as i0 from "@angular/core";
import * as i1 from "../core.config";
import * as i2 from "../logging/logger.service";
/**
 * Notification service.
 */
export class NotifyService {
    /**
     * @param {?} config
     * @param {?} log
     */
    constructor(config, log) {
        this.config = config;
        this.log = log;
    }
    /**
     * @param {?} type
     * @param {?} message
     * @param {?} title
     * @param {?=} config
     * @return {?}
     */
    notify(type, message, title, config) {
        // Add actual notification implementation here.
        return {};
    }
    /**
     * @param {?=} notifyId
     * @param {?=} containerId
     * @return {?}
     */
    clear(notifyId, containerId) {
    }
    /**
     * @template T
     * @param {?} source
     * @return {?}
     */
    with(source) {
        return new NotificationSubscriptionTracker(source, this, this.log);
    }
    /**
     * @template T
     * @param {?} source
     * @return {?}
     */
    promise(source) {
        return new NotificationPromiseTracker(source, this, this.log);
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    wait(message, title = this.config.defaultNotifyWaitMsg, config) {
        return this.notify(NotifyType.Wait, message, title, config);
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    success(message, title = this.config.defaultNotifySuccessMsg, config) {
        return this.notify(NotifyType.Success, message, title, config);
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} error
     * @param {?=} config
     * @return {?}
     */
    error(message, title = this.config.defaultNotifyErrorMsg, error = null, config) {
        return this.notify(NotifyType.Error, format(message, { error }), title, config);
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} error
     * @param {?=} config
     * @return {?}
     */
    warn(message, title = this.config.defaultNotifyWarnMsg, error = null, config) {
        return this.notify(NotifyType.Warn, format(message, { error }), title, config);
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    info(message, title = this.config.defaultNotifyInfoMsg, config) {
        return this.notify(NotifyType.Info, message, title, config);
    }
}
NotifyService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
NotifyService.ctorParameters = () => [
    { type: CoreConfig },
    { type: Logger }
];
/** @nocollapse */ NotifyService.ngInjectableDef = i0.defineInjectable({ factory: function NotifyService_Factory() { return new NotifyService(i0.inject(i1.CoreConfig), i0.inject(i2.Logger)); }, token: NotifyService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    NotifyService.prototype.config;
    /**
     * @type {?}
     * @protected
     */
    NotifyService.prototype.log;
}
