// @tss/sdk/esm2015/lib/core/notify/notification-promise-tracker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BaseNotificationTracker } from './base-notification-tracker';
/**
 * @template T
 */
export class NotificationPromiseTracker extends BaseNotificationTracker {
    /**
     * @param {?} source
     * @param {?} notify
     * @param {?=} log
     * @param {?=} config
     */
    constructor(source, notify, log, config = {}) {
        super(notify, log, config);
        this.source = source;
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    wait(message, title) {
        super.wait(message, title);
        this.source = this.source
            .then((/**
         * @param {?} result
         * @return {?}
         */
        (result) => {
            this.onSuccess(result);
            return result;
        }))
            .catch((/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            this.onFail(error);
            throw error;
        }));
        return this;
    }
    /**
     * @param {?} predicate
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    waitIf(predicate, message, title) {
        return (/** @type {?} */ (super.waitIf(predicate, message, title)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    success(message, title) {
        return (/** @type {?} */ (super.success(message, title)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    error(message, title) {
        return (/** @type {?} */ (super.error(message, title)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    warn(message, title) {
        return (/** @type {?} */ (super.warn(message, title)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @return {?}
     */
    info(message, title) {
        return (/** @type {?} */ (super.info(message, title)));
    }
    /**
     * @param {?} onDone
     * @return {?}
     */
    finally(onDone) {
        return (/** @type {?} */ (super.finally(onDone)));
    }
    /**
     * @param {?=} onfulfilled
     * @return {?}
     */
    then(onfulfilled) {
        return this.source.then(onfulfilled);
    }
    /**
     * @param {?=} onrejected
     * @return {?}
     */
    catch(onrejected) {
        return this.source.catch(onrejected);
    }
}
if (false) {
    /** @type {?} */
    NotificationPromiseTracker.prototype.source;
}
