// @tss/sdk/esm2015/lib/core/notify/notification-subscription-tracker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { tap, publishReplay } from 'rxjs/operators';
import { BaseNotificationTracker } from './base-notification-tracker';
/**
 * @template T
 */
export class NotificationSubscriptionTracker extends BaseNotificationTracker {
    /**
     * @param {?} source
     * @param {?} notify
     * @param {?=} log
     * @param {?=} config
     */
    constructor(source, notify, log, config = {}) {
        super(notify, log, config);
        this.source = source;
        // We need to do a timeout here so that the chained events coming in after this method will be fired
        // before this subscription is added - this is necessary for any synchronous/immediate observables.
        this.source = this.source.pipe(tap((/**
         * @param {?} x
         * @return {?}
         */
        (x) => {
            setTimeout((/**
             * @return {?}
             */
            () => this.onSuccess(x)));
        }), (/**
         * @param {?} error
         * @return {?}
         */
        (error) => {
            setTimeout((/**
             * @return {?}
             */
            () => this.onFail(error)));
        })));
        if (this.config.useSingletonSubscriptionEmission) {
            this.source = this.source.pipe(publishReplay());
        }
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    wait(message, title, config) {
        return (/** @type {?} */ (super.wait(message, title, config)));
    }
    /**
     * @param {?} predicate
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    waitIf(predicate, message, title, config) {
        return (/** @type {?} */ (super.waitIf(predicate, message, title, config)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    success(message, title, config) {
        return (/** @type {?} */ (super.success(message, title, config)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    error(message, title, config) {
        return (/** @type {?} */ (super.error(message, title, config)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    warn(message, title, config) {
        return (/** @type {?} */ (super.warn(message, title, config)));
    }
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    info(message, title, config) {
        return (/** @type {?} */ (super.info(message, title, config)));
    }
    /**
     * @param {?} onDone
     * @return {?}
     */
    finally(onDone) {
        return (/** @type {?} */ (super.finally(onDone)));
    }
    /**
     * @param {?=} next
     * @param {?=} error
     * @param {?=} complete
     * @return {?}
     */
    subscribe(next, error, complete) {
        return this.source.subscribe(next, error, complete);
    }
}
if (false) {
    /** @type {?} */
    NotificationSubscriptionTracker.prototype.source;
}
