// @tss/sdk/esm2015/lib/core/notify/notify-chain.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function NotifyChain() { }
if (false) {
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyChain.prototype.wait = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyChain.prototype.success = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyChain.prototype.error = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyChain.prototype.warn = function (message, title, config) { };
    /**
     * @param {?} message
     * @param {?=} title
     * @param {?=} config
     * @return {?}
     */
    NotifyChain.prototype.info = function (message, title, config) { };
}
