// @tss/sdk/esm2015/lib/esri/portal/to-esri-portal-access-token.function.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { getPortalUrl } from './get-portal-url.function';
import { addSeconds } from '../../core/date-time/add/add-seconds.function';
/**
 * @param {?} token
 * @param {?=} portalUrl
 * @return {?}
 */
export function toEsriPortalAccessToken(token, portalUrl) {
    /** @type {?} */
    const portalToken = {
        accessToken: token.access_token || token.accessToken || token.token || null,
        refreshToken: token.refresh_token || token.refreshToken || null,
        expiresIn: token.expires_in || token.expiresIn || 0,
        isSecure: token.ssl || token.isSecure || false,
        server: token.server || portalUrl || null,
        username: token.username || null,
        expires: token.expires || null
    };
    if (portalToken.expires) {
        portalToken.expires = new Date(portalToken.expires);
    }
    if (portalToken.server) {
        getPortalUrl(portalToken.server);
    }
    if (!portalToken.expires) {
        portalToken.expires = addSeconds(portalToken.expiresIn, new Date());
    }
    return portalToken;
}
