// @tss/sdk/esm2015/lib/core/core.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { KeysPipe } from './keys.pipe';
import { FileSizePipe } from './storage/file-size.pipe';
import { ReadableTimePipe } from './date-time/readable/readable-time.pipe';
import { SafeUrlPipe } from './safe-url.pipe';
export class TssCoreModule {
}
TssCoreModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [
                    FileSizePipe,
                    KeysPipe,
                    ReadableTimePipe,
                    SafeUrlPipe,
                ],
                exports: [
                    FileSizePipe,
                    KeysPipe,
                    ReadableTimePipe,
                    SafeUrlPipe
                ],
                providers: [],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            },] }
];
