// @tss/sdk/esm2015/lib/core/core.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LogLevel } from './logging/log-level';
import { MeasurementUnit } from './units/measurement-unit';
import * as i0 from "@angular/core";
/**
 * A simple settings/config class for storing key/value pairs with persistence.
 */
export class CoreConfig {
    constructor() {
        this.autoConfig = true;
        this.logEnabled = true;
        this.logLevel = LogLevel.Info;
        this.defaultNotifyInfoMsg = 'Info';
        this.defaultNotifyWarnMsg = 'Warning';
        this.defaultNotifyErrorMsg = 'Error';
        this.defaultNotifySuccessMsg = 'Success';
        this.defaultNotifyWaitMsg = 'Wait';
        this.measurementUnit = MeasurementUnit.Imperial;
    }
}
CoreConfig.moduleConfigs = [{ className: 'CoreConfig', injectorToken: CoreConfig }];
CoreConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ CoreConfig.ngInjectableDef = i0.defineInjectable({ factory: function CoreConfig_Factory() { return new CoreConfig(); }, token: CoreConfig, providedIn: "root" });
if (false) {
    /** @type {?} */
    CoreConfig.moduleConfigs;
    /** @type {?} */
    CoreConfig.prototype.autoConfig;
    /** @type {?} */
    CoreConfig.prototype.logEnabled;
    /** @type {?} */
    CoreConfig.prototype.logLevel;
    /** @type {?} */
    CoreConfig.prototype.defaultNotifyInfoMsg;
    /** @type {?} */
    CoreConfig.prototype.defaultNotifyWarnMsg;
    /** @type {?} */
    CoreConfig.prototype.defaultNotifyErrorMsg;
    /** @type {?} */
    CoreConfig.prototype.defaultNotifySuccessMsg;
    /** @type {?} */
    CoreConfig.prototype.defaultNotifyWaitMsg;
    /** @type {?} */
    CoreConfig.prototype.measurementUnit;
}
