// @tss/sdk/esm2015/lib/geolocation/geolocator/geolocation-gps-locator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
// Angular.
import { Injectable } from '@angular/core';
// 3rd Party.
import { Subject, of } from 'rxjs';
import { finalize } from 'rxjs/operators';
// TSS.
import { Geolocatable } from './geolocatable.decorator';
import { BaseGeolocator } from './base-geolocator.service';
import { positionToGpsPosition } from '../gps-position/position-to-gps-position.function';
import { GeolocatorCapability } from './geolocator-capability';
import { Logger } from '../../core/logging/logger.service';
import { GeolocationConfig } from '../geolocation.config';
import { GEOLOCATION_GEOLOCATOR_ID } from '../geolocation.constants';
import * as i0 from "@angular/core";
import * as i1 from "../../core/logging/logger.service";
import * as i2 from "../geolocation.config";
/**
 * Wrapper for geolocation service. The wrapper primarily exists to abstract the direct interaction with
 * the "navigator" object and to allow for injecting other implementations of geolocation functionality.
 */
let GeolocationGeolocator = /**
 * Wrapper for geolocation service. The wrapper primarily exists to abstract the direct interaction with
 * the "navigator" object and to allow for injecting other implementations of geolocation functionality.
 */
class GeolocationGeolocator extends BaseGeolocator {
    /**
     * @param {?} log
     * @param {?} config
     */
    constructor(log, config) {
        super(log);
        this.config = config;
        this.capabilities = [
            GeolocatorCapability.Accuracy,
            GeolocatorCapability.Altitude,
            GeolocatorCapability.AltitudeAccuracy,
            GeolocatorCapability.Heading,
            GeolocatorCapability.Speed,
        ];
    }
    /**
     * @return {?}
     */
    get name() {
        return this.config.geolocationGeolocatorName;
    }
    /**
     * @return {?}
     */
    get description() {
        return this.config.geolocationGeolocatorDescription;
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    getCurrentPosition(options) {
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        (resolve, reject) => navigator.geolocation
            .getCurrentPosition((/**
         * @param {?} position
         * @return {?}
         */
        (position) => {
            this.log.trace(`Received current position for ${this.name}`);
            resolve(positionToGpsPosition(position));
        }), reject, options)));
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    watchPosition(options) {
        /** @type {?} */
        const subject = new Subject();
        /** @type {?} */
        const watchId = navigator.geolocation.watchPosition((/**
         * @param {?} position
         * @return {?}
         */
        position => {
            this.log.trace(`Received watch position for ${this.name}`);
            subject.next(positionToGpsPosition(position));
        }), (/**
         * @param {?} error
         * @return {?}
         */
        error => subject.error(error)), options);
        return subject.pipe(finalize((/**
         * @return {?}
         */
        () => navigator.geolocation.clearWatch(watchId))));
    }
    /**
     * @return {?}
     */
    isAvailable() {
        return of('geolocation' in navigator);
    }
};
GeolocationGeolocator.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
GeolocationGeolocator.ctorParameters = () => [
    { type: Logger },
    { type: GeolocationConfig }
];
/** @nocollapse */ GeolocationGeolocator.ngInjectableDef = i0.defineInjectable({ factory: function GeolocationGeolocator_Factory() { return new GeolocationGeolocator(i0.inject(i1.Logger), i0.inject(i2.GeolocationConfig)); }, token: GeolocationGeolocator, providedIn: "root" });
/**
 * Wrapper for geolocation service. The wrapper primarily exists to abstract the direct interaction with
 * the "navigator" object and to allow for injecting other implementations of geolocation functionality.
 */
GeolocationGeolocator = tslib_1.__decorate([
    Geolocatable(GEOLOCATION_GEOLOCATOR_ID),
    tslib_1.__metadata("design:paramtypes", [Logger, GeolocationConfig])
], GeolocationGeolocator);
export { GeolocationGeolocator };
if (false) {
    /** @type {?} */
    GeolocationGeolocator.prototype.capabilities;
    /**
     * @type {?}
     * @private
     */
    GeolocationGeolocator.prototype.config;
}
