// @tss/sdk/esm2015/lib/geolocation/geolocator/geolocator-source.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { Injectable, Injector } from '@angular/core';
// 3rd party.
import { forkJoin, of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { GEOLOCATOR_REGISTRY } from '../geolocation.constants';
import * as i0 from "@angular/core";
export class GeolocatorSourceService {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getGeolocator(id) {
        /** @type {?} */
        const registry = GEOLOCATOR_REGISTRY.find((/**
         * @param {?} registedSource
         * @return {?}
         */
        registedSource => registedSource.id === id));
        return this.getLocatorFromRegistry(registry);
    }
    /**
     * @return {?}
     */
    getGeolocators() {
        return GEOLOCATOR_REGISTRY
            .map((/**
         * @param {?} registry
         * @return {?}
         */
        registry => this.getLocatorFromRegistry(registry)));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getGeolocatorSource(id) {
        return this.getGeolocatorSources()
            .pipe(map((/**
         * @param {?} devices
         * @return {?}
         */
        devices => devices.find((/**
         * @param {?} device
         * @return {?}
         */
        device => device.id === id)))));
    }
    /**
     * @return {?}
     */
    getAvailableGeolocatorSources() {
        return this.getGeolocatorSources()
            .pipe(map((/**
         * @param {?} devices
         * @return {?}
         */
        devices => devices.filter((/**
         * @param {?} device
         * @return {?}
         */
        device => device.isAvailable)))));
    }
    /**
     * @return {?}
     */
    getGeolocatorSources() {
        /** @type {?} */
        const availableObservables = GEOLOCATOR_REGISTRY
            .map((/**
         * @param {?} registry
         * @return {?}
         */
        registry => {
            /** @type {?} */
            const locator = this.getLocatorFromRegistry(registry);
            /** @type {?} */
            const source = {
                id: registry.id,
                name: locator.name,
                description: locator.description,
                accuracyUnit: locator.accuracyUnit,
                capabilities: locator.capabilities,
                isAvailable: false,
            };
            return locator.isAvailable()
                .pipe(tap((/**
             * @param {?} available
             * @return {?}
             */
            available => source.isAvailable = available)), map((/**
             * @return {?}
             */
            () => source)));
        }));
        return availableObservables.length === 0
            ? of([])
            : forkJoin(availableObservables);
    }
    /**
     * @private
     * @param {?} registry
     * @return {?}
     */
    getLocatorFromRegistry(registry) {
        /** @type {?} */
        const instance = this.injector.get(registry.constructor);
        return instance;
    }
}
GeolocatorSourceService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
GeolocatorSourceService.ctorParameters = () => [
    { type: Injector }
];
/** @nocollapse */ GeolocatorSourceService.ngInjectableDef = i0.defineInjectable({ factory: function GeolocatorSourceService_Factory() { return new GeolocatorSourceService(i0.inject(i0.INJECTOR)); }, token: GeolocatorSourceService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    GeolocatorSourceService.prototype.injector;
}
