// @tss/sdk/esm2015/lib/data-source/data-source-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ResourceConfig } from '../core/resource.config';
import { AutoConfig } from '../core/auto-config.decorator';
import * as i0 from "@angular/core";
/**
 * Configuration class for storing global ESRI-related settings.
 */
let DataSourceConfig = /**
 * Configuration class for storing global ESRI-related settings.
 */
class DataSourceConfig extends ResourceConfig {
    constructor() {
        super(...arguments);
        this.dataSourcesEndpoint = 'api/data/sources/{id}';
        this.tablesEndpoint = 'api/data/sources/{id}/tables';
        this.columnsEndpoint = 'api/data/sources/{id}/tables/{table}/columns/{column}/{action}';
        this.tableActionEndpoint = 'api/data/sources/{id}/tables/{table}/{action}/{subaction}';
        this.editDataEndpoint = 'api/data/sources/{id}/tables/{table}/edit';
        this.applyEditsEndpoint = 'api/data/sources/{id}/applyedits';
        this.attachmentsEndpoint = 'api/attachments/{id}/{action}';
        this.skipColumnQuotes = false;
    }
};
DataSourceConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ DataSourceConfig.ngInjectableDef = i0.defineInjectable({ factory: function DataSourceConfig_Factory() { return new DataSourceConfig(); }, token: DataSourceConfig, providedIn: "root" });
/**
 * Configuration class for storing global ESRI-related settings.
 */
DataSourceConfig = tslib_1.__decorate([
    AutoConfig('DataSourceConfig')
], DataSourceConfig);
export { DataSourceConfig };
if (false) {
    /** @type {?} */
    DataSourceConfig.prototype.dataSourcesEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.tablesEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.columnsEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.tableActionEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.editDataEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.applyEditsEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.attachmentsEndpoint;
    /** @type {?} */
    DataSourceConfig.prototype.skipColumnQuotes;
}
