// @tss/sdk/esm2015/lib/app/api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// Angular.
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
// TSS.
import { AppConfig } from './app.config';
import { BaseEmptyResourceService } from '../core/resource/base-empty-resource.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./app.config";
/**
 * @record
 */
export function ApiStatus() { }
if (false) {
    /** @type {?} */
    ApiStatus.prototype.up;
    /** @type {?} */
    ApiStatus.prototype.down;
    /** @type {?} */
    ApiStatus.prototype.configStoreUp;
    /** @type {?} */
    ApiStatus.prototype.configStoreDown;
    /** @type {?} */
    ApiStatus.prototype.upSince;
    /** @type {?} */
    ApiStatus.prototype.daysUp;
    /** @type {?} */
    ApiStatus.prototype.appCount;
}
export class ApiService extends BaseEmptyResourceService {
    /**
     * @param {?} http
     * @param {?} appConfig
     */
    constructor(http, appConfig) {
        super(http, appConfig.serviceUrl);
        this.appConfig = appConfig;
    }
    /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    ping(routeTokens, params) {
        /** @type {?} */
        const url = this.getUrl(this.appConfig.pingEndpoint, routeTokens);
        return this.http.get(url, { params });
    }
    /**
     * @param {?=} routeTokens
     * @param {?=} params
     * @return {?}
     */
    status(routeTokens, params) {
        /** @type {?} */
        const url = this.getUrl(this.appConfig.pingEndpoint, routeTokens);
        return this.http.get(url, { params });
    }
}
ApiService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ApiService.ctorParameters = () => [
    { type: HttpClient },
    { type: AppConfig }
];
/** @nocollapse */ ApiService.ngInjectableDef = i0.defineInjectable({ factory: function ApiService_Factory() { return new ApiService(i0.inject(i1.HttpClient), i0.inject(i2.AppConfig)); }, token: ApiService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ApiService.prototype.appConfig;
}
