// @tss/sdk/esm2015/lib/app/app-profile.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { AppConfig } from './app.config';
import { BaseResourceService } from '../core/resource/base-resource.service';
import { join } from '../core/url/join.function';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./app.config";
export class AppProfileService extends BaseResourceService {
    /**
     * @param {?} http
     * @param {?} config
     */
    constructor(http, config) {
        super(http, {
            url: join(config.serviceUrl, config.appProfileEndpoint),
            defaultRouteTokens: {
                appId: config.appId
            }
        });
        this.http = http;
    }
}
AppProfileService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AppProfileService.ctorParameters = () => [
    { type: HttpClient },
    { type: AppConfig }
];
/** @nocollapse */ AppProfileService.ngInjectableDef = i0.defineInjectable({ factory: function AppProfileService_Factory() { return new AppProfileService(i0.inject(i1.HttpClient), i0.inject(i2.AppConfig)); }, token: AppProfileService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @protected
     */
    AppProfileService.prototype.http;
}
