// @tss/sdk/esm2015/lib/app/app.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Injector, CUSTOM_ELEMENTS_SCHEMA, APP_INITIALIZER } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { RouterModule } from '@angular/router';
import { TssCoreModule } from '../core/core.module';
import { TssSecurityModule } from '../security/security.module';
import { TssDataSourceModule } from '../data-source/data-source.module';
import { TssPolyfillModule } from '../polyfill/polyfill.module';
import { AppInitializerFactory, AppInitializerService } from './app-initializer.service';
import { AppProfileService } from './app-profile.service';
import { ApiService } from './api.service';
import { TssQueueModule } from '../queuing/queue.module';
import { TssHttpModule } from '../http/http.module';
const ɵ0 = AppInitializerFactory;
export class TssAppModule {
}
TssAppModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpClientModule,
                    RouterModule,
                    TssCoreModule,
                    TssPolyfillModule,
                    TssHttpModule,
                    TssDataSourceModule,
                    TssQueueModule,
                    TssSecurityModule,
                ],
                declarations: [],
                exports: [],
                providers: [
                    AppProfileService,
                    ApiService,
                    AppInitializerService,
                    {
                        provide: APP_INITIALIZER,
                        useFactory: ɵ0,
                        multi: true,
                        deps: [Injector]
                    }
                ],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            },] }
];
export { ɵ0 };
