// @tss/sdk/esm2015/lib/app/app.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ResourceConfig } from '../core/resource.config';
import { AutoConfig } from '../core/auto-config.decorator';
import * as i0 from "@angular/core";
/** @type {?} */
export const DEFAULT_APP_ID = 'all';
/** @type {?} */
export const DEFAULT_PROFILE_ID = 'default';
/**
 * Application related configuration options. These settings are used to
 * drive all general "global" application related behaviour. The settings can be
 * overridden directly in the code, application module, or from the TDS server.
 */
let AppConfig = /**
 * Application related configuration options. These settings are used to
 * drive all general "global" application related behaviour. The settings can be
 * overridden directly in the code, application module, or from the TDS server.
 */
class AppConfig extends ResourceConfig {
    constructor() {
        super(...arguments);
        this.appId = DEFAULT_APP_ID;
        this.queryProfile = true;
        this.profileId = null;
        this.profile = null;
        this.defaultProfileId = DEFAULT_PROFILE_ID;
        this.allowQueryParamInjections = true;
        this.continueOnAppBootstrapError = false;
        this.appSettingsEndpoint = 'api/apps/{appId}/settings/{id}';
        this.appProfileEndpoint = 'api/apps/{appId}/profiles/{id}';
        this.pingEndpoint = 'api/ping';
        this.statusEndpoint = 'api/status';
    }
};
AppConfig.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ AppConfig.ngInjectableDef = i0.defineInjectable({ factory: function AppConfig_Factory() { return new AppConfig(); }, token: AppConfig, providedIn: "root" });
/**
 * Application related configuration options. These settings are used to
 * drive all general "global" application related behaviour. The settings can be
 * overridden directly in the code, application module, or from the TDS server.
 */
AppConfig = tslib_1.__decorate([
    AutoConfig('AppConfig')
], AppConfig);
export { AppConfig };
if (false) {
    /** @type {?} */
    AppConfig.prototype.appId;
    /** @type {?} */
    AppConfig.prototype.queryProfile;
    /** @type {?} */
    AppConfig.prototype.profileId;
    /** @type {?} */
    AppConfig.prototype.profile;
    /** @type {?} */
    AppConfig.prototype.defaultProfileId;
    /** @type {?} */
    AppConfig.prototype.allowQueryParamInjections;
    /** @type {?} */
    AppConfig.prototype.continueOnAppBootstrapError;
    /** @type {?} */
    AppConfig.prototype.appSettingsEndpoint;
    /** @type {?} */
    AppConfig.prototype.appProfileEndpoint;
    /** @type {?} */
    AppConfig.prototype.pingEndpoint;
    /** @type {?} */
    AppConfig.prototype.statusEndpoint;
}
